{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n",
    "import math\n",
    "import imageio\n",
    "import json\n",
    "import time\n",
    "\n",
    "from collections import Counter\n",
    "from matplotlib.pyplot import figure\n",
    "from matplotlib import animation\n",
    "\n",
    "from Bio.PDB import *\n",
    "import nglview as nv\n",
    "import math\n",
    "import warnings\n",
    "from Bio.PDB.StructureBuilder import PDBConstructionWarning\n",
    "warnings.filterwarnings(\"ignore\", category=PDBConstructionWarning)\n",
    "\n",
    "# this is dangerous of course but uncomment it when you want to run the code\n",
    "# warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "\n",
    "\n",
    "\n",
    "# seg_dir = \"C:\\Users\\Kevin\\Documents\\TDA\\afrl\\Segmentations\"\n",
    "# seg_dir = \"C:/Users/Kevin/Documents/TDA/afrl/Segmentations\"\n",
    "seg_dir = \"../Batch_protein_all_5_3_no_na.csv_2019.07.08 16.10/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# aaron's code to make sure the folders are all in their proper place\n",
    "\n",
    "dirlocal = os.path.curdir\n",
    "data_path = os.path.join(dirlocal, 'Data/')\n",
    "protein_path = os.path.join(data_path, 'Proteins/')\n",
    "pdb_path = os.path.join(data_path, 'PDB/')\n",
    "fasta_path = os.path.join(data_path, 'fasta/')\n",
    "dssp_path = os.path.join(data_path, 'DSSP/')\n",
    "\n",
    "if not os.path.exists(data_path):\n",
    "    os.makedirs(data_path)\n",
    "if not os.path.exists(protein_path):\n",
    "    os.makedirs(protein_path)\n",
    "if not os.path.exists(pdb_path):\n",
    "    os.makedirs(pdb_path)\n",
    "ss_path = os.path.join(data_path, 'ss.txt')\n",
    "feature_path = os.path.join(data_path, 'sample-input-features.npy')\n",
    "distance_path = os.path.join(data_path, 'sample-distance-maps-cb.npy')\n",
    "full_feature_path = os.path.join(data_path, 'full-input-features.npy')\n",
    "full_distance_path = os.path.join(data_path, 'full-distance-maps-cb.npy')\n",
    "test_feature_path = os.path.join(data_path, 'testset-input-features.npy')\n",
    "test_distance_path = os.path.join(data_path, 'testset-distance-maps-cb.npy')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## aaron's code to get raw features from pdb\n",
    "\n",
    "import numpy as np\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from matplotlib.pyplot import figure\n",
    "%matplotlib inline\n",
    "import time\n",
    "import requests\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "# ! pip install biopython\n",
    "# ! pip install nglview\n",
    "# ! jupyter-nbextension enable nglview --py --sys-prefix\n",
    "\n",
    "from Bio.PDB import *\n",
    "import nglview as nv\n",
    "import math\n",
    "import warnings\n",
    "from Bio.PDB.StructureBuilder import PDBConstructionWarning\n",
    "warnings.filterwarnings(\"ignore\", category=PDBConstructionWarning)\n",
    "\n",
    "# this is dangerous of course but uncomment it when you want to run the code\n",
    "# warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_chip_start_stop(df, step_size, chip_size):\n",
    "\n",
    "    ch=[int(x.split(\"_\")[-1]) for x in df['chip_id']]\n",
    "\n",
    "    start=[c*step_size for c in ch]\n",
    "    stop=[chip_size+c*step_size for c in ch]\n",
    "\n",
    "    df['start']=start\n",
    "    df['stop']=stop\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "###  aaron's functions: \n",
    "\n",
    "# gets dataframe containing torsion angles, peptides, chain, etc\n",
    "# from pdb file\n",
    "def get_torsion_angles(pdb_id, degrees=False):\n",
    "    pdb_struct = get_pdb_structure(pdb_id)\n",
    "    torsion_angles = []\n",
    "    for model in pdb_struct:\n",
    "        for chain in model:\n",
    "            polypeptides = PPBuilder().build_peptides(chain)\n",
    "            for poly_index, poly in enumerate(polypeptides):\n",
    "                #print(\"Model %s Chain %s\" % (str(model.id), str(chain.id)))\n",
    "                #print(\"(part %i of %i)\" % (poly_index+1, len(polypeptides)))\n",
    "                #print(\"length %i\" % (len(poly)))\n",
    "                #print(\"from %s%i\" % (poly[0].resname, poly[0].id[1]))\n",
    "                #print(\"to %s%i\" % (poly[-1].resname, poly[-1].id[1]))\n",
    "                phi_psi = poly.get_phi_psi_list()\n",
    "                for res_index, residue in enumerate(poly):\n",
    "                    res_name = \"%s%i\" % (residue.resname, residue.id[1])\n",
    "                    #print(res_name, tuple(math.degrees(b) for b in phi_psi[res_index] if b))\n",
    "                    deg = phi_psi[res_index]\n",
    "                    if degrees:\n",
    "                        deg = tuple(math.degrees(b) if b else None for b in deg)\n",
    "                    phi, psi = deg\n",
    "                    model_name, model_id = model.full_id\n",
    "                    torsion_angles.append([model_name, model_id, chain.id, residue.resname, residue.id[1], phi, psi])\n",
    "    return pd.DataFrame(torsion_angles, columns=['Model_Name', 'Model_ID', 'Chain', 'Residue_Name', 'Residue_ID', 'Phi', 'Psi'])\n",
    "\n",
    "\n",
    "# generates ramachandran plot from a given pdb file\n",
    "def ramachandran_plot(pdb_id, degrees=True):\n",
    "    df = get_torsion_angles(pdb_id, degrees)\n",
    "    x = df['Phi']\n",
    "    y = df['Psi']\n",
    "    # Generate plot\n",
    "    plt.plot(x, y, \".\")\n",
    "    plt.title('Ramachandran Plot')\n",
    "    if degrees:\n",
    "        plt.xlabel(f'$\\Phi$ Angle (Degrees)')\n",
    "        plt.xlim(-180, 180)\n",
    "        plt.ylabel(f'$\\Psi$ Angle (Radians)')\n",
    "        plt.ylim(-180, 180)\n",
    "        plt.show()\n",
    "    else:\n",
    "        plt.xlabel(f'$\\Phi$ Angle (Radians)')\n",
    "        plt.xlim(-math.pi, math.pi)\n",
    "        plt.ylabel(f'$\\Psi$ Angle (Radians)')\n",
    "        plt.ylim(-math.pi, math.pi)\n",
    "        plt.show()\n",
    "\n",
    "# helper function which generates a biopython structure\n",
    "# for display in display_protein()\n",
    "def get_pdb_structure(pdb_id):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    parser = PDBParser()\n",
    "    file_path = os.path.join(pdb_path, f\"{pdb_id}.pdb\")\n",
    "    try:\n",
    "        struct = parser.get_structure(pdb_id, file_path)\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "        return None\n",
    "    return struct\n",
    "\n",
    "\n",
    "#  downloads the protein in question from pdb\n",
    "def get_pdb_file(pdb_id, replace=False):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    file_path = os.path.join(pdb_path, f'{pdb_id}.pdb')\n",
    "    if os.path.isfile(file_path) and not replace:\n",
    "        return True\n",
    "    parser = PDBParser()\n",
    "    url = f'https://files.rcsb.org/download/{pdb_id}.pdb'\n",
    "    resp = requests.get(url)\n",
    "    try:\n",
    "        #print(file_path)\n",
    "        file = open(file_path, \"wb\")\n",
    "        file.write(resp.content)\n",
    "        file.close()\n",
    "    except Exception as e: \n",
    "        print(e)\n",
    "        return False\n",
    "    return True\n",
    "\n",
    "# neato 3d viewer of a protein's structure\n",
    "def display_protein(pdb_id):\n",
    "    pdb_struct = get_pdb_structure(pdb_id)\n",
    "    view = nv.show_biopython(pdb_struct)\n",
    "    return view\n",
    "\n",
    "\n",
    "#  returns a dataframe with secondary structure, amino acid code, chain id, etc\n",
    "# deprecated\n",
    "def get_secondary_structure(pdb_id):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    indexes = get_pdb_ss_seq(pdb_id)\n",
    "    chain_df = []\n",
    "    for i, chain_ss in enumerate(indexes):\n",
    "        chain_id, seq, ss = chain_ss\n",
    "        #unknown_list = [i for i, c in enumerate(seq) if c == 'X']\n",
    "        #print(unknown_list)\n",
    "        seq_list = [c for c in seq]\n",
    "        # [f(x) if condition else g(x) for x in sequence]\n",
    "        ss_list = ['L' if c is ' ' else c for c in ss]\n",
    "        #print(seq_list + ss_list)\n",
    "        chain_df.append(pd.DataFrame({'Chain': chain_id, 'Amino_Acid':seq_list, 'Secondary_Structure':ss_list}))\n",
    "    return pd.concat(chain_df)\n",
    "\n",
    "# get matrix of pairwise carbon beta distances\n",
    "def get_cb_distances(full_pdb_id, partial=False):\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    get_pdb_file(pdb_id)\n",
    "    pdb_struct = get_pdb_structure(pdb_id)  # Returns the BioPython structure of the PDB file\n",
    "    ppb = PPBuilder()\n",
    "    chain_dist_matrix = {}\n",
    "    for model in pdb_struct:\n",
    "        for chain in model:\n",
    "            if chain.id == full_pdb_id[4] and partial:\n",
    "                residues = []\n",
    "                for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                    for residue in pp:\n",
    "                        residues.append(residue)\n",
    "                lng = len(residues)\n",
    "                dist_matrix = np.zeros((lng, lng), np.float)\n",
    "                for i in range(0, lng):\n",
    "                    for j in range(i, lng):\n",
    "                        try:\n",
    "                            diff_vector  = residues[i]['C'].coord - residues[j]['C'].coord\n",
    "                        except Exception:\n",
    "                            # This just gets the average value around the unknown residue distance\n",
    "                            count = 0\n",
    "                            sums = 0\n",
    "                            if j - 1 > 0:\n",
    "                                count += 1\n",
    "                                sums += dist_matrix[i][j - 1]\n",
    "                            if i - 1 > 0:\n",
    "                                count += 1\n",
    "                                sums += dist_matrix[i - 1][j]\n",
    "                            if count > 0:\n",
    "                                diff_vector = sums / count\n",
    "                            else:\n",
    "                                diff_vector = 0\n",
    "                        dist = np.sqrt(np.sum(diff_vector * diff_vector))\n",
    "                        dist_matrix[i][j] = dist\n",
    "                        dist_matrix[j][i] = dist\n",
    "                    chain_dist_matrix = dist_matrix\n",
    "        \n",
    "    if partial:\n",
    "        start_pos, end_pos, bool_arr = get_seq_pos(full_pdb_id)\n",
    "        if start_pos == -1:\n",
    "            return pd.DataFrame(chain_dist_matrix[bool_arr]).iloc[:, bool_arr]\n",
    "        return pd.DataFrame(chain_dist_matrix[start_pos:end_pos]).iloc[:, start_pos:end_pos]\n",
    "    else:\n",
    "        return chain_dist_matrix\n",
    "    \n",
    "    \n",
    "# gets alignment of given sequence on top of the original sequence\n",
    "# from a fasta file downloaded from the cloud repository\n",
    "def get_seq_alignment(full_pdb_id):\n",
    "    seq = ''\n",
    "    for record in SeqIO.parse(os.path.join(fasta_path, f'{full_pdb_id}.fasta'), \"fasta\"):\n",
    "        seq = record.seq\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    #print(pdb_id)\n",
    "    get_pdb_file(pdb_id)\n",
    "    struct = get_pdb_structure(pdb_id)\n",
    "    ppb = PPBuilder()\n",
    "    for model in struct:\n",
    "        for chain in model:\n",
    "            if chain.id == full_pdb_id[4]:\n",
    "                ground_seq = ''\n",
    "                for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                    ground_seq += pp.get_sequence()\n",
    "                alignments = pairwise2.align.globalxx(ground_seq, str(seq))\n",
    "                al_seq = alignments[0][1]\n",
    "                al_seq_bool = [False if a is '-' else True for a in al_seq]\n",
    "                return al_seq_bool\n",
    "\n",
    "            \n",
    "def get_seq_pos(full_pdb_id):\n",
    "    seq = ''\n",
    "    for record in SeqIO.parse(os.path.join(fasta_path, f'{full_pdb_id}.fasta'), \"fasta\"):\n",
    "        seq = str(record.seq)\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    get_pdb_file(pdb_id)\n",
    "    struct = get_pdb_structure(pdb_id)\n",
    "    ppb = PPBuilder()\n",
    "    for model in struct:\n",
    "        for chain in model:\n",
    "            if chain.id == full_pdb_id[4]:\n",
    "                ground_seq = ''\n",
    "                for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                    ground_seq += pp.get_sequence()\n",
    "                start_pos = ground_seq.find(seq)\n",
    "                if start_pos == -1:\n",
    "                    alignments = pairwise2.align.globalxx(ground_seq, seq)\n",
    "                    al_seq = alignments[0][1]\n",
    "                    bool_seq = [False if a == '-' else True for a in al_seq]\n",
    "                    #start_pos = largest_substring_pos(ground_seq, seq)\n",
    "                    #start_pos = int(alignments[0][2] - 1)\n",
    "                    #end_pos = start_pos + len(al_seq)\n",
    "                    return (-1, -1, bool_seq)\n",
    "                else:\n",
    "                    end_pos = start_pos + len(seq)\n",
    "                    return (start_pos, end_pos, '')\n",
    "\n",
    "\n",
    "            \n",
    "            \n",
    "# calls dssp api to get dssp file\n",
    "def pdb_id_to_dssp_file(pdb_id, replace=False):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    rest_url = 'http://www.cmbi.umcn.nl/xssp/'\n",
    "    # Read the pdb id data into a variable\n",
    "    data = {'data': pdb_id}\n",
    "\n",
    "    # Send a request to the server to create hssp data from the pdb file data.\n",
    "    # If an error occurs, an exception is raised and the program exits. If the\n",
    "    # request is successful, the id of the job running on the server is\n",
    "    # returned.\n",
    "    url_create = f'{rest_url}api/create/pdb_id/dssp/'\n",
    "    r = requests.post(url_create, data=data)\n",
    "    r.raise_for_status()\n",
    "\n",
    "    job_id = json.loads(r.text)['id']\n",
    "    #print(f'Job submitted successfully. Id is: {job_id}')\n",
    "\n",
    "    # Loop until the job running on the server has finished, either successfully\n",
    "    # or due to an error.\n",
    "    ready = False\n",
    "    while not ready:\n",
    "        # Check the status of the running job. If an error occurs an exception\n",
    "        # is raised and the program exits. If the request is successful, the\n",
    "        # status is returned.\n",
    "        url_status = f'{rest_url}api/status/pdb_id/dssp/{job_id}/'\n",
    "        r = requests.get(url_status)\n",
    "        r.raise_for_status()\n",
    "\n",
    "        status = json.loads(r.text)['status']\n",
    "        #print(f'Job status is: {status}')\n",
    "\n",
    "        # If the status equals SUCCESS, exit out of the loop by changing the\n",
    "        # condition ready. This causes the code to drop into the `else` block\n",
    "        # below.\n",
    "        #\n",
    "        # If the status equals either FAILURE or REVOKED, an exception is raised\n",
    "        # containing the error message. The program exits.\n",
    "        #\n",
    "        # Otherwise, wait for five seconds and start at the beginning of the\n",
    "        # loop again.\n",
    "        if status == 'SUCCESS':\n",
    "            ready = True\n",
    "        elif status in ['FAILURE', 'REVOKED']:\n",
    "            raise Exception(json.loads(r.text)['message'])\n",
    "        else:\n",
    "            time.sleep(5)\n",
    "    else:\n",
    "        # Requests the result of the job. If an error occurs an exception is\n",
    "        # raised and the program exits. If the request is successful, the result\n",
    "        # is returned.\n",
    "        url_result = f'{rest_url}api/result/pdb_id/dssp/{job_id}/'\n",
    "        r = requests.get(url_result)\n",
    "        r.raise_for_status()\n",
    "        result = json.loads(r.text)['result']\n",
    "        try:\n",
    "            file_path = os.path.join(dssp_path, f'{pdb_id}.dssp')\n",
    "            if os.path.isfile(file_path) and not replace:\n",
    "                return True\n",
    "            #print(file_path)\n",
    "            file = open(file_path, \"w\")\n",
    "            file.write(result)\n",
    "            file.close()\n",
    "        except Exception as e: \n",
    "            print(e)\n",
    "            return False\n",
    "        # Return the result to the caller, which prints it to the screen.\n",
    "        return True\n",
    "\n",
    "# uses api call file to generate secondary structure, phi/psi, solvent-accessibility stuff\n",
    "# because of indexing stuff, it crashes if we use partial = True\n",
    "# ask aaron about that\n",
    "def get_ground_truth_api(full_pdb_id, partial=False):\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    get_pdb_file(pdb_id)\n",
    "    struct = get_pdb_structure(pdb_id)\n",
    "    ppb = PPBuilder()\n",
    "    struct_info = []\n",
    "    for model in struct:\n",
    "        file = os.path.join(pdb_path, f'{pdb_id}.pdb')\n",
    "        pdb_id_to_dssp_file(pdb_id)\n",
    "        dssp = DSSP(model=model, in_file=os.path.join(dssp_path, f'{pdb_id}.dssp'), file_type='DSSP')\n",
    "        seq_count = 0\n",
    "        #dssp = DSSP(model=model, in_file=file)\n",
    "        for chain in model:\n",
    "            if chain.id == full_pdb_id[4] or not partial:\n",
    "                residues = []\n",
    "                seq = ''\n",
    "                for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                    seq += pp.get_sequence()\n",
    "                    for residue in pp:\n",
    "                        residues.append(residue)\n",
    "                for i, residue in enumerate(residues):\n",
    "                    try:\n",
    "                        key = list(dssp.keys())[seq_count]   \n",
    "                        dssp_info = dssp[key]\n",
    "                        amino_acid = dssp_info[1]\n",
    "                        sec_struct = dssp_info[2]\n",
    "                        solv_acc = dssp_info[3]\n",
    "                        phi = dssp_info[4]\n",
    "                        psi = dssp_info[5]\n",
    "                    except Exception:\n",
    "                        # DSSP didn't capture this amino acid\n",
    "                        #print(full_pdb_id, seq_count, len(dssp), len(seq), residue)\n",
    "                        amino_acid = seq[i]\n",
    "                        sec_struct = '-'\n",
    "                        solv_acc = 0\n",
    "                        phi = 360\n",
    "                        psi = 360\n",
    "                    seq_count += 1\n",
    "                    # Keys 6 through 13 is bonding energy / relidx (no clue what this is)\n",
    "                    struct_info.append([model.full_id[0], model.full_id[1], chain.id,\n",
    "                        residue.resname, residue.id[1], amino_acid, sec_struct, solv_acc, phi, psi])\n",
    "    info_df = pd.DataFrame(struct_info, \n",
    "        columns=['Model_Name', 'Model_ID', 'Chain', 'Residue_Name',\n",
    "        'Residue_ID', 'Amino_Acid', 'Secondary_Structure', 'Solvent_Accessability', \n",
    "        'Phi', 'Psi'])\n",
    "    if partial:    \n",
    "        start_pos, end_pos, bool_arr = get_seq_pos(full_pdb_id)\n",
    "        if start_pos == -1:\n",
    "            return info_df[bool_arr]\n",
    "        return info_df[start_pos:end_pos]\n",
    "    else:\n",
    "        return info_df\n",
    "\n",
    "\n",
    "    \n",
    "# the following two functions download the files from the api\n",
    "# to create a cache on the disk\n",
    "# speeds things up immensely, only need to run once though\n",
    "def get_ground_truth_files(full_pdb_id, replace=False, partial=False):\n",
    "    bad_list = ['4osnA0']\n",
    "    if full_pdb_id in bad_list:\n",
    "        return 'Bad File'\n",
    "    if partial:\n",
    "        file_path = os.path.join(protein_path, full_pdb_id + '.csv')\n",
    "    else:\n",
    "        file_path = os.path.join(protein_path, full_pdb_id[0:4] + '.csv')\n",
    "        \n",
    "    if os.path.isfile(file_path) and not replace:\n",
    "        return True\n",
    "    \n",
    "    df = get_ground_truth(full_pdb_id, partial)\n",
    "    file_path = ''\n",
    "    try:\n",
    "        df.to_csv(file_path)\n",
    "    except Exception:\n",
    "        return False\n",
    "    return True\n",
    "    \n",
    "def get_cb_distance_files(full_pdb_id, replace=False, partial=False):\n",
    "    bad_list = ['4osnA0']\n",
    "    if full_pdb_id in bad_list:\n",
    "        return 'Bad File'\n",
    "    file_path = ''\n",
    "    if partial:\n",
    "        file_path = os.path.join(distances_path, full_pdb_id + '.csv')\n",
    "        if os.path.isfile(file_path) and not replace:\n",
    "            return True\n",
    "        try:\n",
    "            df = get_cb_distances(full_pdb_id, partial)\n",
    "            df.to_csv(file_path)\n",
    "        except Exception:\n",
    "            return False\n",
    "        return True\n",
    "    else:\n",
    "        df = get_cb_distances(full_pdb_id, partial)\n",
    "        keys = df.keys\n",
    "        for key in keys:\n",
    "            file_path = os.path.join(distances_path, full_pdb_id[0:4] + key + '0.csv')\n",
    "            if os.path.isfile(file_path) and not replace:\n",
    "                return True\n",
    "            try:\n",
    "                df[key].to_csv(file_path)\n",
    "            except Exception:\n",
    "                return False\n",
    "        return True\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_25=pandas.read_csv(\"datasets/protein_all_25_12_no_na.csv\")\n",
    "dataset_10=pandas.read_csv(\"datasets/protein_all_10_5_no_na.csv\")\n",
    "dataset_5=pandas.read_csv(\"datasets/protein_all_5_3_no_na.csv\")\n",
    "\n",
    "fix_chip_start_stop(dataset_25,12,25)\n",
    "fix_chip_start_stop(dataset_10,5,10)\n",
    "fix_chip_start_stop(dataset_5,3,5)\n",
    "\n",
    "\n",
    "pdb_list, length_dict, aa_seq, input_features = np.load(\"datasets/testset-input-features.npy\",allow_pickle=True)\n",
    "pdb_list_y, distance_maps_cb = np.load(\"datasets/testset-distance-maps-cb.npy\",encoding=\"latin1\",allow_pickle=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions\n",
    "def get_rows_from_segmentation(seg, dataset):\n",
    "    with open(seg_dir+seg) as jsonf:\n",
    "        a=json.load(jsonf)\n",
    "        df=dataset.iloc[[int(m) for m in a['rows']],:]\n",
    "        return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def draw_ss_predicts(index,ss):\n",
    "    \n",
    "#     H, G and I to H; E to E; the rest to C\n",
    "    ss_dict={\n",
    "        \"H\":0,\n",
    "        \"G\":0,\n",
    "        \"I\":0,\n",
    "        \"E\":1,\n",
    "        \"B\":2,\n",
    "        \"T\":2,\n",
    "        \"S\":2             \n",
    "            }\n",
    "    \n",
    "    aa_length=length_dict[pdb_list[index]]\n",
    "    \n",
    "    plt.figure(figsize=(15,10))\n",
    "\n",
    "    df25=dataset_25[dataset_25[\"protein_id\"]==dataset_25[\"protein_id\"].unique()[index]] \n",
    "    df10=dataset_10[dataset_10[\"protein_id\"]==dataset_10[\"protein_id\"].unique()[index]] \n",
    "    df5=dataset_5[dataset_5[\"protein_id\"]==dataset_5[\"protein_id\"].unique()[index]] \n",
    "\n",
    "    plt.subplot(4,1,1)\n",
    "    y=ground_truth_window_avg(df5)[\"ground_truth_\"+ss]\n",
    "    x=range(0,len(y))\n",
    "    plt.bar(x,y,1)\n",
    "    plt.xlim((0,aa_length))    \n",
    "\n",
    "    plt.subplot(4,1,2)\n",
    "    y=ground_truth_window_avg(df10)[\"ground_truth_\"+ss]\n",
    "    x=range(0,len(y))\n",
    "    plt.bar(x,y,1)\n",
    "    plt.xlim((0,aa_length))\n",
    "\n",
    "    plt.subplot(4,1,3)\n",
    "    y=ground_truth_window_avg(df25)[\"ground_truth_\"+ss]\n",
    "    x=range(0,len(y))\n",
    "    plt.bar(x,y,1)\n",
    "    plt.xlim((0,aa_length))\n",
    "\n",
    "    plt.subplot(4,1,4)\n",
    "    y=[1-n for n in input_features[pdb_list[index]][ss_dict[ss]].astype(np.float)]\n",
    "    x=range(0,len(y))\n",
    "    plt.bar(x,y,1)\n",
    "    plt.xlim((0,aa_length))\n",
    "\n",
    "    \n",
    "    \n",
    "def get_inv_psipred_from_ss(protein):\n",
    "#     H to H; E to E; the rest to C\n",
    "    ss_dict={\n",
    "        \"H\":0,\n",
    "        \"G\":2,\n",
    "        \"I\":2,\n",
    "        \"E\":1,\n",
    "        \"B\":2,\n",
    "        \"T\":2,\n",
    "        \"S\":2             \n",
    "            }\n",
    "    \n",
    "    # I subtract it from 1 to make it on the same scale as the other predictions\n",
    "    # (i.e. high means more likely to be X)\n",
    "    \n",
    "    return [1-n for n in input_features[protein][ss_dict[ss]].astype(np.float)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA3AAAAI/CAYAAAA2kzvaAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nOzdfbBdZ30f+u8vcgyFvNhEoqWWxDGtmuLQNKYal5ROS3kJcsxYmSm5kZtMDCXV5E5c0iTcVi4d0zrNHRN66yYTh8YFF5LJxVA3TdRY1PHwMsnc1o5lCC+260Q1Dj7YjQED7ZTGjsLv/rGXw+boHOlYWz57L+3PZ0ajvZ717L0e+znP1vqeZ61nVXcHAACAxfd1824AAAAAmyPAAQAAjIQABwAAMBICHAAAwEgIcAAAACMhwAEAAIzEOfNuwFrbt2/vlZWVeTcDAABgLu6+++7PdfeO9fYtXIBbWVnJ0aNH590MAACAuaiq399o3ykvoayqm6rq0ar65Ab7q6p+tqqOVdXHq+rFU/uurKrfG/5ceXrNBwAAINncPXDvSrLvJPsvTbJn+HMwyduTpKqek+QtSf5qkkuSvKWqzp+lsQAAAMvslAGuu38zyWMnqbI/yS/2xB1Jzquq5yV5dZLbu/ux7v5Ckttz8iAIAADASZyJVSgvSPLQ1PbqULZROQAAAKfhTCxiUuuU9UnKT/yAqoOZXH6Z3bt3n4EmwTitHLp13k142jx43WXzbgIjttHY8HM1Xk/X952fCeBJi3xeNct31ZmYgVtNsmtqe2eSh09SfoLuvrG793b33h071l0tEwAAYOmdiQB3OMkPDqtRviTJl7r7kSS3Jfmuqjp/WLzku4YyAAAATsMpL6GsqvckeVmS7VW1msnKkl+fJN39r5McSfLdSY4l+XKS1w/7Hquqn0xy1/BR13b3yRZDAQAA4CROGeC6+4pT7O8kP7LBvpuS3HR6TQMAAGDambiEEgAAgC0gwAEAAIyEAAcAADASAhwAAMBICHAAAAAjIcABAACMxCkfIwAAsBkrh26ddxMAznpm4AAAAEZCgAMAABgJAQ4AAGAkBDgAAICREOAAAABGQoADAAAYCQEOAABgJAQ4AACAkRDgAAAARkKAAwAAGIlz5t0AYDmsHLp103UfvO6yp7ElnE38XM3HU/n/vtX8TMzHRv/f/T/mTFvk75+tYgYOAABgJDYV4KpqX1XdX1XHqurQOvuvr6rfGf78blV9cWrfH0/tO3wmGw8AALBMTnkJZVVtS3JDklclWU1yV1Ud7u57n6zT3T82Vf/vJ7l46iP+d3d/x5lrMgAAwHLazAzcJUmOdfcD3f1EkpuT7D9J/SuSvOdMNA4AAICv2kyAuyDJQ1Pbq0PZCarq+UkuTPLBqeJnVtXRqrqjqr7ntFsKAACw5DazCmWtU9Yb1D2Q5Jbu/uOpst3d/XBVvSDJB6vqE939377mAFUHkxxMkt27d2+iSQAAAMtnMzNwq0l2TW3vTPLwBnUPZM3lk9398PD3A0k+nK+9P+7JOjd2997u3rtjx45NNAkAAGD5bCbA3ZVkT1VdWFXnZhLSTlhNsqq+Ncn5Sf7LVNn5VfWM4fX2JC9Ncu/a9wIAAHBqp7yEsruPV9VVSW5Lsi3JTd19T1Vdm+Rodz8Z5q5IcnN3T19e+cIkv1BVX8kkLF43vXolAAAAm7eZe+DS3UeSHFlTds2a7X+6zvv+c5K/NEP7AAAAGGzqQd4AAADM36Zm4IDTt3Lo1nk3ARaSsQFnv6cyzh+87rKnsSWwWGb5N9AMHAAAwEgIcAAAACMhwAEAAIyEAAcAADASAhwAAMBICHAAAAAjIcABAACMhAAHAAAwEgIcAADASAhwAAAAI3HOvBsAsNbKoVvXLX/wusu2uCWcTdb7ufIztdz8TCwW/cFaG50PLDszcAAAACMhwAEAAIyEAAcAADASAhwAAMBICHAAAAAjIcABAACMxKYCXFXtq6r7q+pYVR1aZ//rquqzVfU7w58fmtp3ZVX93vDnyjPZeAAAgGVyyufAVdW2JDckeVWS1SR3VdXh7r53TdX3dvdVa977nCRvSbI3SSe5e3jvF85I6wEAAJbIZmbgLklyrLsf6O4nktycZP8mP//VSW7v7seG0HZ7kn2n11QAAIDltpkAd0GSh6a2V4eytf52VX28qm6pql1P8b0AAACcwikvoUxS65T1mu3/mOQ93f14Vf1wkncnefkm35uqOpjkYJLs3r17E00CAM6klUO3zrsJLBg/E7CYNjMDt5pk19T2ziQPT1fo7s939+PD5r9J8lc2+97h/Td2997u3rtjx47Nth0AAGCpbCbA3ZVkT1VdWFXnJjmQ5PB0hap63tTm5UnuG17fluS7qur8qjo/yXcNZQAAADxFp7yEsruPV9VVmQSvbUlu6u57quraJEe7+3CSN1bV5UmOJ3ksyeuG9z5WVT+ZSQhMkmu7+7Gn4b8DAADgrLeZe+DS3UeSHFlTds3U66uTXL3Be29KctMMbQQAACCbfJA3AAAA8yfAAQAAjIQABwAAMBICHAAAwEgIcAAAACMhwAEAAIzEph4jAJzayqFb590EWFjGBwCcGWbgAAAARkKAAwAAGAkBDgAAYCQEOAAAgJEQ4AAAAEZCgAMAABgJAQ4AAGAkBDgAAICREOAAAABGQoADAAAYCQEOAABgJM6ZdwMANmvl0K2brvvgdZc9jS3hbOFnirWeys9E4ufi6WaMnn2e6hjjRJuagauqfVV1f1Udq6pD6+z/8aq6t6o+XlUfqKrnT+3746r6neHP4TPZeAAAgGVyyhm4qtqW5IYkr0qymuSuqjrc3fdOVftokr3d/eWq+j+T/HSS7xv2/e/u/o4z3G4AAICls5kZuEuSHOvuB7r7iSQ3J9k/XaG7P9TdXx4270iy88w2EwAAgM0EuAuSPDS1vTqUbeQNSd4/tf3MqjpaVXdU1fecRhsBAADI5hYxqXXKet2KVT+QZG+SvzlVvLu7H66qFyT5YFV9orv/25r3HUxyMEl27969qYYDAAAsm83MwK0m2TW1vTPJw2srVdUrk7w5yeXd/fiT5d398PD3A0k+nOTite/t7hu7e293792xY8dT+g8AAABYFpsJcHcl2VNVF1bVuUkOJPma1SSr6uIkv5BJeHt0qvz8qnrG8Hp7kpcmmV78BAAAgE065SWU3X28qq5KcluSbUlu6u57quraJEe7+3CStyX5hiT/rqqS5NPdfXmSFyb5har6SiZh8bo1q1cCAACwSZt6kHd3H0lyZE3ZNVOvX7nB+/5zkr80SwMBAACY2NSDvAEAAJi/Tc3AwbJaOXTrvJsAC8v4GCf9xlp+JmBczMABAACMhAAHAAAwEgIcAADASAhwAAAAIyHAAQAAjIQABwAAMBICHAAAwEgIcAAAACMhwAEAAIyEAAcAADAS58y7AQDAbFYO3TrvJrBg/EzA2csMHAAAwEgIcAAAACMhwAEAAIyEAAcAADASAhwAAMBICHAAAAAjsakAV1X7qur+qjpWVYfW2f+MqnrvsP/OqlqZ2nf1UH5/Vb36zDUdAABguZwywFXVtiQ3JLk0yUVJrqiqi9ZUe0OSL3T3n09yfZK3Du+9KMmBJN+WZF+Snx8+DwAAgKdoMzNwlyQ51t0PdPcTSW5Osn9Nnf1J3j28viXJK6qqhvKbu/vx7v5UkmPD5wEAAPAUbSbAXZDkoant1aFs3TrdfTzJl5J8yybfCwAAwCacs4k6tU5Zb7LOZt6bqjqY5OCw+XhVfXIT7WLxbE/yuXk3gtNy1vVdvXXeLdgyZ13fLaqn4WdK343Xn/TdEn3XLLxN9oVxN07L2G/P32jHZgLcapJdU9s7kzy8QZ3VqjonyTcneWyT701335jkxiSpqqPdvXcT7WLB6Lvx0nfjpe/GS9+Nl74bL303Tvrta23mEsq7kuypqgur6txMFiU5vKbO4SRXDq9fm+SD3d1D+YFhlcoLk+xJ8ttnpukAAADL5ZQzcN19vKquSnJbkm1Jburue6rq2iRHu/twkncm+aWqOpbJzNuB4b33VNX7ktyb5HiSH+nuP36a/lsAAADOapu5hDLdfSTJkTVl10y9/sMk37vBe38qyU89hTbd+BTqslj03Xjpu/HSd+Ol78ZL342Xvhsn/TalJlc6AgAAsOg2cw8cAAAAC2ChAlxV7auq+6vqWFUdmnd72FhV7aqqD1XVfVV1T1X96FD+nKq6vap+b/j7/Hm3lRNV1baq+mhV/fqwfWFV3Tn023uHBYtYMFV1XlXdUlX/dRh732nMjUNV/djwXfnJqnpPVT3TuFtMVXVTVT06/UijjcZZTfzscN7y8ap68fxazgZ997bhO/PjVfUfquq8qX1XD313f1W9ej6tJlm/76b2vamquqq2D9tLP+4WJsBV1bYkNyS5NMlFSa6oqovm2ypO4niSn+juFyZ5SZIfGfrrUJIPdPeeJB8Ytlk8P5rkvqnttya5fui3LyR5w1xaxan8TJL/1N1/MclfzqQPjbkFV1UXJHljkr3d/aJMFgQ7EONuUb0ryb41ZRuNs0szWWF7TybPs337FrWR9b0rJ/bd7Ule1N3fnuR3k1ydJMM5y4Ek3za85+eHc1Hm4105se9SVbuSvCrJp6eKl37cLUyAS3JJkmPd/UB3P5Hk5iT759wmNtDdj3T3R4bX/zOTE8kLMumzdw/V3p3ke+bTQjZSVTuTXJbkHcN2JXl5kluGKvptAVXVNyX5G5ms+pvufqK7vxhjbizOSfKnhmelPivJIzHuFlJ3/2YmK2pP22ic7U/yiz1xR5Lzqup5W9NS1lqv77r7N7r7+LB5RybPJE4mfXdzdz/e3Z9KciyTc1HmYINxlyTXJ/mHSaYX7Vj6cbdIAe6CJA9Nba8OZSy4qlpJcnGSO5P86e5+JJmEvCTPnV/L2MC/yuTL8CvD9rck+eLUP3DG3mJ6QZLPJvm3w+Wv76iqZ8eYW3jd/Zkk/yKT3yA/kuRLSe6OcTcmG40z5y7j8neTvH94re8WXFVdnuQz3f2xNbuWvu8WKcDVOmWWyFxwVfUNSf59kn/Q3f9j3u3h5KrqNUke7e67p4vXqWrsLZ5zkrw4ydu7++Ik/ysulxyF4X6p/UkuTPJnkzw7k0uA1jLuxsf350hU1Zszuf3jl58sWqeavlsQVfWsJG9Ocs16u9cpW6q+W6QAt5pk19T2ziQPz6ktbEJVfX0m4e2Xu/tXhuI/eHIae/j70Xm1j3W9NMnlVfVgJpcpvzyTGbnzhku7EmNvUa0mWe3uO4ftWzIJdMbc4ntlkk9192e7+4+S/EqSvxbjbkw2GmfOXUagqq5M8pok399ffX6Wvltsfy6TX3p9bDhn2ZnkI1X1Z6LvFirA3ZVkz7Aq17mZ3Fh6eM5tYgPDfVPvTHJfd//LqV2Hk1w5vL4yya9tddvYWHdf3d07u3slkzH2we7+/iQfSvLaoZp+W0Dd/d+TPFRV3zoUvSLJvTHmxuDTSV5SVc8avjuf7Dvjbjw2GmeHk/zgsCreS5J86clLLVkMVbUvyT9Kcnl3f3lq1+EkB6rqGVV1YSYLYvz2PNrIibr7E9393O5eGc5ZVpO8ePi3cOnH3UI9yLuqvjuT2YBtSW7q7p+ac5PYQFX99SS/leQT+eq9VP84k/vg3pdkdyYnLd/b3evdlMqcVdXLkrypu19TVS/IZEbuOUk+muQHuvvxebaPE1XVd2Sy+My5SR5I8vpMfhFnzC24qvpnSb4vk0u4PprkhzK5Z8O4WzBV9Z4kL0uyPckfJHlLkl/NOuNsCOQ/l8nqeV9O8vruPjqPdrNh312d5BlJPj9Uu6O7f3io/+ZM7os7nsmtIO9f+5lsjfX6rrvfObX/wUxW8v2ccbdgAQ4AAICNLdIllAAAAJyEAAcAADASAhwAAMBICHAAAAAjIcABAACMhAAHAAAwEgIcAADASAhwAAAAI3HOvBuw1vbt23tlZWXezQAAAJiLu++++3PdvWO9fQsX4FZWVnL06NF5NwMAAGAuqur3N9rnEkoAAICREOAAAABGQoADAAAYCQEOAABgJAQ4AACAkRDgAAAARkKAAwAAGAkBDgAAYCQEOAAAgJEQ4AAAAEbinHk3AJiPlUO3bunxHrzusi09HgDA2cgMHAAAwEgIcAAAACMhwAEAAIyEAAcAADASFjGBBbLVC4sAADCuxd3MwAEAAIzETAGuqvZV1f1VdayqDq2zf3dVfaiqPlpVH6+q757leAAAAMvstANcVW1LckOSS5NclOSKqrpoTbV/kuR93X1xkgNJfv50jwcAALDsZpmBuyTJse5+oLufSHJzkv1r6nSSbxpef3OSh2c4HgAAwFKbZRGTC5I8NLW9muSvrqnzT5P8RlX9/STPTvLKGY4HAACw1GYJcLVOWa/ZviLJu7r7/6mq70zyS1X1ou7+ytd8UNXBJAeTZPfu3TM0CQCYl61cxW2WFdwAxmyWSyhXk+ya2t6ZEy+RfEOS9yVJd/+XJM9Msn3tB3X3jd29t7v37tixY4YmAQAAnL1mCXB3JdlTVRdW1bmZLFJyeE2dTyd5RZJU1QszCXCfneGYAAAAS+u0A1x3H09yVZLbktyXyWqT91TVtVV1+VDtJ5L8var6WJL3JHldd6+9zBIAAIBNmOUeuHT3kSRH1pRdM/X63iQvneUYAAAATMz0IG8AAAC2zkwzcAAAAE+HrVzZdkzMwAEAAIyEAAcAADASAhwAAMBICHAAAAAjIcABAACMhAAHAAAwEgIcAADASAhwAAAAIyHAAQAAjIQABwAAMBICHAAAwEicM+8GAAA8VSuHbt3S4z143WVbejyAjZiBAwAAGImZAlxV7auq+6vqWFUd2qDO/1FV91bVPVX1/85yPAAAgGV22pdQVtW2JDckeVWS1SR3VdXh7r53qs6eJFcneWl3f6GqnjtrgwEAAJbVLDNwlyQ51t0PdPcTSW5Osn9Nnb+X5Ibu/kKSdPejMxwPAABgqc0S4C5I8tDU9upQNu0vJPkLVfX/VdUdVbVvhuMBAAAstVlWoax1ynqdz9+T5GVJdib5rap6UXd/8Ws+qOpgkoNJsnv37hmaBAAAcPaaZQZuNcmuqe2dSR5ep86vdfcfdfenktyfSaD7Gt19Y3fv7e69O3bsmKFJAAAAZ69ZAtxdSfZU1YVVdW6SA0kOr6nzq0n+VpJU1fZMLql8YIZjAgAALK3TDnDdfTzJVUluS3Jfkvd19z1VdW1VXT5Uuy3J56vq3iQfSvJ/dffnZ200AADAMprlHrh095EkR9aUXTP1upP8+PAHAACAGcz0IG8AAAC2zkwzcACbtXLo1i093oPXXbalx+PM2cqfFT8nAIyNGTgAAICREOAAAABGQoADAAAYCQEOAABgJAQ4AACAkRDgAAAARkKAAwAAGAkBDgAAYCQEOAAAgJEQ4AAAAEZCgAMAABiJc+bdAFhkK4dunXcTAE6b7zCAxTTL97MZOAAAgJGYKcBV1b6qur+qjlXVoZPUe21VdVXtneV4AAAAy+y0A1xVbUtyQ5JLk1yU5Iqqumidet+Y5I1J7jzdYwEAADDbDNwlSY519wPd/USSm5PsX6feTyb56SR/OMOxAAAAlt4sAe6CJA9Nba8OZX+iqi5Osqu7f32G4wAAAJDZVqGsdcr6T3ZWfV2S65O87pQfVHUwycEk2b179wxNAgDgqdjq1UofvO6yLT0enG1mmYFbTbJrantnkoentr8xyYuSfLiqHkzykiSH11vIpLtv7O693b13x44dMzQJAADg7DVLgLsryZ6qurCqzk1yIMnhJ3d295e6e3t3r3T3SpI7klze3UdnajEAAMCSOu0A193Hk1yV5LYk9yV5X3ffU1XXVtXlZ6qBAAAATMxyD1y6+0iSI2vKrtmg7stmORYAAMCym+lB3gAAAGydmWbgADj7bfUKdcDZzaqXMBszcAAAACMhwAEAAIyEAAcAADASAhwAAMBICHAAAAAjIcABAACMhAAHAAAwEgIcAADASAhwAAAAIyHAAQAAjIQABwAAMBLnzLsBADAvK4du3dLjPXjdZVt6PIAzaau/M1nfTDNwVbWvqu6vqmNVdWid/T9eVfdW1cer6gNV9fxZjgcAALDMTjvAVdW2JDckuTTJRUmuqKqL1lT7aJK93f3tSW5J8tOnezwAAIBlN8sM3CVJjnX3A939RJKbk+yfrtDdH+ruLw+bdyTZOcPxAAAAltosAe6CJA9Nba8OZRt5Q5L3z3A8AACApTbLIia1TlmvW7HqB5LsTfI3N9h/MMnBJNm9e/cMTQIAOPMseAMsillm4FaT7Jra3pnk4bWVquqVSd6c5PLufny9D+ruG7t7b3fv3bFjxwxNAgAAOHvNEuDuSrKnqi6sqnOTHEhyeLpCVV2c5BcyCW+PznAsAACApXfaAa67jye5KsltSe5L8r7uvqeqrq2qy4dqb0vyDUn+XVX9TlUd3uDjAAAAOIWZHuTd3UeSHFlTds3U61fO8vkAAAB81UwP8gYAAGDrzDQDB7CorBgHjNlWf4cB42EGDgAAYCQEOAAAgJEQ4AAAAEZCgAMAABgJAQ4AAGAkrELJqFiVCxgz32Gw9axKzNnGDBwAAMBICHAAAAAjIcABAACMhAAHAAAwEgIcAADASFiFEmBkrGQIAMvLDBwAAMBIzBTgqmpfVd1fVceq6tA6+59RVe8d9t9ZVSuzHA8AAGCZnXaAq6ptSW5IcmmSi5JcUVUXran2hiRf6O4/n+T6JG893eMBAAAsu1lm4C5Jcqy7H+juJ5LcnGT/mjr7k7x7eH1LkldUVc1wTAAAgKU1S4C7IMlDU9urQ9m6dbr7eJIvJfmWGY4JAACwtGZZhXK9mbQ+jTqpqoNJDg6bj1fVJ2doF/OzPcnn5t0ITou+m1HN7wJxfTde+m689N14Pe19N8d/D85myzjmnr/RjlkC3GqSXVPbO5M8vEGd1ao6J8k3J3ls7Qd1941JbkySqjra3XtnaBdzou/GS9+Nl74bL303XvpuvPTdOOm3rzXLJZR3JdlTVRdW1blJDiQ5vKbO4SRXDq9fm+SD3X3CDBwAAACndtozcN19vKquSnJbkm1Jburue6rq2iRHu/twkncm+aWqOpbJzNuBM9FoAACAZTTLJZTp7iNJjqwpu2bq9R8m+d6n+LE3ztIm5krfjZe+Gy99N176brz03Xjpu3HSb1PKFY0AAADjMMs9cAAAAGyhhQpwVbWvqu6vqmNVdWje7WFjVbWrqj5UVfdV1T1V9aND+XOq6vaq+r3h7/Pn3VZOVFXbquqjVfXrw/aFVXXn0G/vHRYmYsFU1XlVdUtV/ddh7H2nMTcOVfVjw3flJ6vqPVX1TONuMVXVTVX16PQjjTYaZzXxs8N5y8er6sXzazkb9N3bhu/Mj1fVf6iq86b2XT303f1V9er5tJpk/b6b2vemquqq2j5sL/24W5gAV1XbktyQ5NIkFyW5oqoumm+rOInjSX6iu1+Y5CVJfmTor0NJPtDde5J8YNhm8fxokvumtt+a5Pqh376Q5A1zaRWn8jNJ/lN3/8UkfzmTPjTmFlxVXZDkjUn2dveLMln460CMu0X1riT71pRtNM4uTbJn+HMwydu3qI2s7105se9uT/Ki7v72JL+b5OokGc5ZDiT5tuE9Pz+cizIf78qJfZeq2pXkVUk+PVW89ONuYQJckkuSHOvuB7r7iSQ3J9k/5zaxge5+pLs/Mrz+n5mcSF6QSZ+9e6j27iTfM58WspGq2pnksiTvGLYrycuT3DJU0W8LqKq+KcnfyGR133T3E939xRhzY3FOkj81PBP1WUkeiXG3kLr7N3PiM2s3Gmf7k/xiT9yR5Lyqet7WtJS11uu77v6N7j4+bN6RyXOLk0nf3dzdj3f3p5Icy+RclDnYYNwlyfVJ/mGS6UU7ln7cLVKAuyDJQ1Pbq0MZC66qVpJcnOTOJH+6ux9JJiEvyXPn1zI28K8y+TL8yrD9LUm+OPUPnLG3mF6Q5LNJ/u1w+es7qurZMeYWXnd/Jsm/yOQ3yI8k+VKSu2PcjclG48y5y7j83STvH17ruwVXVZcn+Ux3f2zNrqXvu0UKcLVOmSUyF1xVfUOSf5/kH3T3/5h3ezi5qnpNkke7++7p4nWqGnuL55wkL07y9u6+OMn/isslR2G4X2p/kguT/Nkkz87kEqC1jLvx8f05ElX15kxu//jlJ4vWqabvFkRVPSvJm5Ncs97udcqWqu8WKcCtJtk1tb0zycNzagubUFVfn0l4++Xu/pWh+A+enMYe/n50Xu1jXS9NcnlVPZjJZcovz2RG7rzh0q7E2FtUq0lWu/vOYfuWTAKdMbf4XpnkU9392e7+oyS/kuSvxbgbk43GmXOXEaiqK5O8Jsn391efn6XvFtufy+SXXh8bzll2JvlIVf2Z6LuFCnB3JdkzrMp1biY3lh6ec5vYwHDf1DuT3Nfd/3Jq1+EkVw6vr0zya1vdNjbW3Vd3987uXslkjH2wu78/yYeSvHaopt8WUHf/9yQPVdW3DkWvSHJvjLkx+HSSl1TVs4bvzif7zrgbj43G2eEkPzisiveSJF968lJLFkNV7Uvyj5Jc3t1fntp1OMmBqnpGVV2YyYIYvz2PNnKi7v5Edz+3u1eGc5bVJC8e/i1c+nG3UA/yrqrvzmQ2YFuSm7r7p+bcJDZQVX89yW8l+US+ei/VP87kPrj3JdmdyUnL93b3ejelMmdV9bIkb+ru11TVCzKZkXtOko8m+YHufnye7eNEVfUdmSw+c26SB5K8PpNfxBlzC66q/lmS78vkEq6PJvmhTO7ZMO4WTFW9J8nLkmxP8gdJ3pLkV7POOBsC+c9lsnrel5O8vruPzqPdbNh3Vyd5RpLPD9Xu6O4fHuq/OZP74o5ncivI+9d+Jltjvb7r7ndO7X8wk5V8P2fcLViAAwAAYGOLdAklAAAAJyHAAQAAjIQABwAAMBICHAAAwEgIcAAAACMhwAEAAIyEAAcAADASAhwAAMBInDPvBqy1ffv2XllZmXczAAAA5uLuu+/+XHfvWG/fwgW4lZWVHD16dN7NAAAAmIuq+v2N9rmEEloA/qAAAB4FSURBVAAAYCQEOAAAgJEQ4AAAAEZi4e6BAzZn5dCt827Clnvwusvm3QQAgLkyAwcAADASAhwAAMBICHAAAAAjIcABAACMhAAHAAAwEgIcAADASAhwAAAAIyHAAQAAjIQABwAAMBICHAAAwEgIcAAAACMhwAEAAIyEAAcAADASAhwAAMBICHAAAAAjIcABAACMxDnzbgAA8PRYOXTrvJuw5R687rJ5NwHgaWUGDgAAYCQEOAAAgJEQ4AAAAEZCgAMAABgJAQ4AAGAkBDgAAICREOAAAABGQoADAAAYiZkCXFXtq6r7q+pYVR06Sb3XVlVX1d5ZjgcAALDMzjndN1bVtiQ3JHlVktUkd1XV4e6+d029b0zyxiR3ztJQWFYrh26ddxMWxnr/Lx687rI5tARYVL4ngLPdLDNwlyQ51t0PdPcTSW5Osn+dej+Z5KeT/OEMxwIAAFh6swS4C5I8NLW9OpT9iaq6OMmu7v71GY4DAABAZriEMkmtU9Z/srPq65Jcn+R1p/ygqoNJDibJ7t27Z2gSAADAUzOmW1ZmmYFbTbJrantnkoentr8xyYuSfLiqHkzykiSH11vIpLtv7O693b13x44dMzQJAADg7DXLDNxdSfZU1YVJPpPkQJK/8+TO7v5Sku1PblfVh5O8qbuPznBMAGCNMf3mGIDZnPYMXHcfT3JVktuS3Jfkfd19T1VdW1WXn6kGAgAAMDHLDFy6+0iSI2vKrtmg7stmORYkfssMwFO3jP92eHQCnL1mepA3AAAAW0eAAwAAGAkBDgAAYCRmugcOABbNMt7vBMDyMAMHAAAwEgIcAADASAhwAAAAI+EeOAAAWFLLeN/w2J+TaAYOAABgJMzAAQAAS2Pss45m4AAAAEbCDBwwamP/LdrpGPu1+8DTb6PvRt8fMH5m4AAAAEZCgAMAABgJAQ4AAGAkBDgAAICREOAAAABGwiqUACNjdbmvWsZVSAFOh+/Ls4cZOAAAgJEQ4AAAAEZCgAMAABgJAQ4AAGAkBDgAAICREOAAAABGwmMEAACWxDIuJb+Mj1jh7CbAsbCW8R8ZAAA4GZdQAgAAjIQABwAAMBIuoQQA4Ky13i0Z7otjzMzAAQAAjIQABwAAMBICHAAAwEjMFOCqal9V3V9Vx6rq0Dr7f7yq7q2qj1fVB6rq+bMcDwAAYJmddoCrqm1JbkhyaZKLklxRVRetqfbRJHu7+9uT3JLkp0/3eAAAAMtulhm4S5Ic6+4HuvuJJDcn2T9dobs/1N1fHjbvSLJzhuMBAAAstVkC3AVJHpraXh3KNvKGJO+f4XgAAABLbZbnwNU6Zb1uxaofSLI3yd/cYP/BJAeTZPfu3TM0CWB5rfesIwBO5PuSMZtlBm41ya6p7Z1JHl5bqapemeTNSS7v7sfX+6DuvrG793b33h07dszQJAAAgLPXLAHuriR7qurCqjo3yYEkh6crVNXFSX4hk/D26AzHAgAAWHqnHeC6+3iSq5LcluS+JO/r7nuq6tqqunyo9rYk35Dk31XV71TV4Q0+DgAAgFOY5R64dPeRJEfWlF0z9fqVs3w+AAAAXzXTg7wBAADYOgIcAADASAhwAAAAIyHAAQAAjIQABwAAMBICHAAAwEgIcAAAACMhwAEAAIyEAAcAADASAhwAAMBICHAAAAAjIcABAACMhAAHAAAwEgIcAADASAhwAAAAIyHAAQAAjIQABwAAMBICHAAAwEgIcAAAACMhwAEAAIzEOfNuAF9r5dCt824CAACwoMzAAQAAjIQABwAAMBICHAAAwEgIcAAAACMhwAEAAIyEAAcAADASAhwAAMBICHAAAAAjIcABAACMhAAHAAAwEgIcAADASAhwAAAAIzFTgKuqfVV1f1Udq6pD6+x/RlW9d9h/Z1WtzHI8AACAZXbaAa6qtiW5IcmlSS5KckVVXbSm2huSfKG7/3yS65O89XSPBwAAsOxmmYG7JMmx7n6gu59IcnOS/Wvq7E/y7uH1LUleUVU1wzEBAACW1iwB7oIkD01trw5l69bp7uNJvpTkW2Y4JgAAwNI6Z4b3rjeT1qdRJ1V1MMnBYfPxqvrkDO1ifrYn+dy8G8Fp0Xfjpe/GS9+Nl74bL303TsvYb8/faMcsAW41ya6p7Z1JHt6gzmpVnZPkm5M8tvaDuvvGJDcmSVUd7e69M7SLOdF346XvxkvfjZe+Gy99N176bpz029ea5RLKu5LsqaoLq+rcJAeSHF5T53CSK4fXr03ywe4+YQYOAACAUzvtGbjuPl5VVyW5Lcm2JDd19z1VdW2So919OMk7k/xSVR3LZObtwJloNAAAwDKa5RLKdPeRJEfWlF0z9foPk3zvU/zYG2dpE3Ol78ZL342XvhsvfTde+m689N046bcp5YpGAACAcZjlHjgAAAC20EIFuKraV1X3V9Wxqjo07/awsaraVVUfqqr7quqeqvrRofw5VXV7Vf3e8Pf5824rJ6qqbVX10ar69WH7wqq6c+i39w4LE7Fgquq8qrqlqv7rMPa+05gbh6r6seG78pNV9Z6qeqZxt5iq6qaqenT6kUYbjbOa+NnhvOXjVfXi+bWcDfrubcN35ser6j9U1XlT+64e+u7+qnr1fFpNsn7fTe17U1V1VW0ftpd+3C1MgKuqbUluSHJpkouSXFFVF823VZzE8SQ/0d0vTPKSJD8y9NehJB/o7j1JPjBss3h+NMl9U9tvTXL90G9fSPKGubSKU/mZJP+pu/9ikr+cSR8acwuuqi5I8sYke7v7RZks/HUgxt2ieleSfWvKNhpnlybZM/w5mOTtW9RG1veunNh3tyd5UXd/e5LfTXJ1kgznLAeSfNvwnp8fzkWZj3flxL5LVe1K8qokn54qXvpxtzABLsklSY519wPd/USSm5Psn3Ob2EB3P9LdHxle/89MTiQvyKTP3j1Ue3eS75lPC9lIVe1MclmSdwzbleTlSW4Zqui3BVRV35Tkb2Syum+6+4nu/mKMubE4J8mfGp6J+qwkj8S4W0jd/Zs58Zm1G42z/Ul+sSfuSHJeVT1va1rKWuv1XXf/RncfHzbvyOS5xcmk727u7se7+1NJjmVyLsocbDDukuT6JP8wyfSiHUs/7hYpwF2Q5KGp7dWhjAVXVStJLk5yZ5I/3d2PJJOQl+S582sZG/hXmXwZfmXY/pYkX5z6B87YW0wvSPLZJP92uPz1HVX17BhzC6+7P5PkX2TyG+RHknwpyd0x7sZko3Hm3GVc/m6S9w+v9d2Cq6rLk3ymuz+2ZtfS990iBbhap8wSmQuuqr4hyb9P8g+6+3/Muz2cXFW9Jsmj3X33dPE6VY29xXNOkhcneXt3X5zkf8XlkqMw3C+1P8mFSf5skmdncgnQWsbd+Pj+HImqenMmt3/88pNF61TTdwuiqp6V5M1Jrllv9zplS9V3ixTgVpPsmtremeThObWFTaiqr88kvP1yd//KUPwHT05jD38/Oq/2sa6XJrm8qh7M5DLll2cyI3fecGlXYuwtqtUkq91957B9SyaBzphbfK9M8qnu/mx3/1GSX0ny12LcjclG48y5ywhU1ZVJXpPk+/urz8/Sd4vtz2XyS6+PDecsO5N8pKr+TPTdQgW4u5LsGVblOjeTG0sPz7lNbGC4b+qdSe7r7n85tetwkiuH11cm+bWtbhsb6+6ru3tnd69kMsY+2N3fn+RDSV47VNNvC6i7/3uSh6rqW4eiVyS5N8bcGHw6yUuq6lnDd+eTfWfcjcdG4+xwkh8cVsV7SZIvPXmpJYuhqvYl+UdJLu/uL0/tOpzkQFU9o6ouzGRBjN+eRxs5UXd/oruf290rwznLapIXD/8WLv24W6gHeVfVd2cyG7AtyU3d/VNzbhIbqKq/nuS3knwiX72X6h9nch/c+5LszuSk5Xu7e72bUpmzqnpZkjd192uq6gWZzMg9J8lHk/xAdz8+z/Zxoqr6jkwWnzk3yQNJXp/JL+KMuQVXVf8syfdlcgnXR5P8UCb3bBh3C6aq3pPkZUm2J/mDJG9J8qtZZ5wNgfznMlk978tJXt/dR+fRbjbsu6uTPCPJ54dqd3T3Dw/135zJfXHHM7kV5P1rP5OtsV7fdfc7p/Y/mMlKvp8z7hYswAEAALCxRbqEEgAAgJMQ4AAAAEZCgAMAABgJAQ4AAGAkBDgAAICREOAAAABGQoADAAAYCQEOAABgJM6ZdwPW2r59e6+srMy7GQAAAHNx9913f667d6y375QBrqpuSvKaJI9294vW2V9JfibJdyf5cpLXdfdHhn1XJvknQ9V/3t3vPtXxVlZWcvTo0VNVAwAAOCtV1e9vtG8zl1C+K8m+k+y/NMme4c/BJG8fDvqcJG9J8leTXJLkLVV1/uaaDAAAwFqnDHDd/ZtJHjtJlf1JfrEn7khyXlU9L8mrk9ze3Y919xeS3J6TB0EAAABO4kwsYnJBkoemtleHso3KAQAAOA1nYhGTWqesT1J+4gdUHczk8svs3r37DDQJAIBZrBy6ddN1H7zusqexJcC0MxHgVpPsmtremeThofxla8o/vN4HdPeNSW5Mkr17964b8rbCU/mierr4Ajw9/pGBrTfrd6axCFvv6TrX2ehzjfOTW4Rzz7HxM3VmAtzhJFdV1c2ZLFjype5+pKpuS/J/Ty1c8l1Jrj4Dx2OJnYkvOmEPnhonfMDpWm+cG+Mwm808RuA9mcykba+q1UxWlvz6JOnuf53kSCaPEDiWyWMEXj/se6yqfjLJXcNHXdvdJ1sMhTihATCzN15O1oGnm3PlTQS47r7iFPs7yY9ssO+mJDedXtOY5h/F+fAlAWcP36MTZ+IqBJd9Lb5F7qNl/Ld1kfvjbLFM3/Fn4hJKABbAMv3jdTJOlM4c/y/hqTNueLqdNQHuqfw2x8ACgK2xjLMtAE+nsybAwVYy08FYOHk+Of9/gNNlQmDxna3f8Wd9gDubB9fZ+kMJwNnPL8KAeRn7989ZH+AAAM5mZ/Mvq4ETfd28GwAAAMDmmIEDIInf4p+N9CnA2UeAA1gAVtIFADZDgGNhOUkFWC4W5zo5/y4CiQAHAAAsuTH9AkmAOwuNfWnUsRrTwAe/yQcWjfMX2BwBDmCBCVrgxB5gmgAHACMn6AMsDwFuSbi8DwAAnppFPIcW4AAAtoCZUuBMEODgaebeDQAAzpSvm3cDAAAA2BwBDgAAYCRcQgkAC2gRb5wHYP42FeCqal+Sn0myLck7uvu6NfuvT/K3hs1nJXlud5837PvjJJ8Y9n26uy8/Ew0HgKfLIt+7aiEMlolfZMCJThngqmpbkhuSvCrJapK7qupwd9/7ZJ3u/rGp+n8/ycVTH/G/u/s7zlyTOds4GQEAgM3ZzAzcJUmOdfcDSVJVNyfZn+TeDepfkeQtZ6Z5AAAAi2WeV2psZhGTC5I8NLW9OpSdoKqen+TCJB+cKn5mVR2tqjuq6ntOu6UAAABLbjMzcLVOWW9Q90CSW7r7j6fKdnf3w1X1giQfrKpPdPd/+5oDVB1McjBJdu/evYkmAZw5i3y/E7A+90YBy2ozM3CrSXZNbe9M8vAGdQ8kec90QXc/PPz9QJIP52vvj3uyzo3dvbe79+7YsWMTTQIAAFg+mwlwdyXZU1UXVtW5mYS0w2srVdW3Jjk/yX+ZKju/qp4xvN6e5KXZ+N45AAAATuKUl1B29/GquirJbZk8RuCm7r6nqq5NcrS7nwxzVyS5ubunL698YZJfqKqvZBIWr5tevRIAAIDN29Rz4Lr7SJIja8quWbP9T9d5339O8pdmaB8AAACDTQU4zl4WbwCAM8vzTYGn02bugQMAAGABCHAAAAAj4RJKgHV4xhQAsIjMwAEAAIyEAAcAADASLqGEOXB5HgAAp0OAA5aGpb0BgLFzCSUAAMBICHAAAAAj4RJKZrbeZWnu5QIAgDPPDBwAAMBImIEDgDmzwA4Am2UGDgAAYCTMwAEAnCazp8BWE+B4WnhQNQAAnHkCHCwBK4UCAJwd3AMHAAAwEgIcAADASGzqEsqq2pfkZ5JsS/KO7r5uzf7XJXlbks8MRT/X3e8Y9l2Z5J8M5f+8u999BtoNMBfu7wQA1rNV5winDHBVtS3JDUlelWQ1yV1Vdbi7711T9b3dfdWa9z4nyVuS7E3SSe4e3vuFM9J6RsdqXQAAcPo2cwnlJUmOdfcD3f1EkpuT7N/k5786ye3d/dgQ2m5Psu/0mgoAALDcNhPgLkjy0NT26lC21t+uqo9X1S1VtespvhcAAIBT2Mw9cLVOWa/Z/o9J3tPdj1fVDyd5d5KXb/K9qaqDSQ4mye7duzfRJFge7rkCAOBJm5mBW02ya2p7Z5KHpyt09+e7+/Fh898k+Subfe/w/hu7e293792xY8dm2w4AALBUNjMDd1eSPVV1YSarTB5I8nemK1TV87r7kWHz8iT3Da9vS/J/V9X5w/Z3Jbl65lYDwBYzGw7AIjhlgOvu41V1VSZhbFuSm7r7nqq6NsnR7j6c5I1VdXmS40keS/K64b2PVdVPZhICk+Ta7n7safjvAAAAOOtt6jlw3X0kyZE1ZddMvf7/27v/mDvPuo7j74+dKw6iAzZU+oN1WpWBymYzphglsMXiltY/IKuBsODMooGAP4iuLplKsqQTI2KYhGWrA0OoOBGekM2J24wm0tHCZD+d1q5uzzoZP6dx2tnx9Y/7rpw9Pac9W5/13FfP+5U0z31f93WeXk++z3We+3uuH/dWJoysVdV2YPsxtFHSlHxMgyRJ0oltqgROkiSpBeM+yHKaq6QTyTSbmEiSJEmSBsAROKlRTpeUJEmaP47ASZIkSVIjHIGT5pRbokuSJLXHEThJkiRJaoQjcDqMIzOSJEnSMDkCJ0mSJEmNMIGTJEmSpEY4hVLS0/gQXEmSpOFyBE6SJEmSGmECJ0mSJEmNMIGTJEmSpEaYwEmSJElSI0zgJEmSJKkR7kIp6YQ06YH0kiRJLTOBkyRJOgo/FJI0FE6hlCRJkqRGmMBJkiRJUiOcQilJy2Dc9Kp92y6cQUskSdKJbKoRuCQbkzyQZE+Sy8dc/7Uk9yW5K8mtSV42cu2pJP/Y/1tYzsZLkiRJ0jw56ghckhXANcAFwCKwK8lCVd03Uu1OYENVPZHkl4HfAy7ur/13Vb1qmdstSZIkSXNnmhG4c4E9VbW3qp4EdgCbRytU1e1V9UR/uhNYvbzNlCRJkiRNswZuFfDwyPki8Ooj1L8UuHnk/HlJdgMHgW1V9cln3EpJkgZq0vbyroGUJD0XpkngMqasxlZM3gJsAH56pHhtVe1PciZwW5K7q+pfl7zuMuAygLVr107VcEmz542rJEnS8TXNFMpFYM3I+Wpg/9JKSc4HrgA2VdWBQ+VVtb//uhf4W+Dspa+tqmurakNVbTj99NOf0Q8gSZIkSfNimgRuF7A+ybokJwNbgKftJpnkbOBDdMnbYyPlL0yysj8+DXgNMLr5iSRJkiRpSkedQllVB5O8A7gFWAFsr6p7k7wH2F1VC8B7gRcAf54E4KGq2gS8HPhQkm/SJYvbluxeKUmSJEma0lQP8q6qm4CblpRdOXJ8/oTX/QPww8fSQEmSJElSZ6oHeUuSJEmSZm+qEThJ823SbpOSJEk6vhyBkyRJkqRGmMBJkiRJUiNM4CRJkiSpEa6BkyRJkqTnyLi9BPZtu/BZfz9H4CRJkiSpESZwkiRJktQIp1BKkqQT2qRHoRzLFCZJmhUTOEnLbrnnekuSJKnjFEpJkiRJaoQJnCRJkiQ1wgROkiRJkhrhGjhNbdIicEmSJEnHR5MJnImEJGno3MxHkvRccAqlJEmSJDWiyRE4Se2Zx+cwzePPLEnHw7GOcPv+rJYNOoFzqqTmjb/zHf+w6kRlH5eGzanPasGgEzhJOhpviCVJ0jyZKoFLshF4P7ACuK6qti25vhL4CPBjwFeBi6tqX39tK3Ap8BTwzqq6ZdlaL2mumKxJeq75PiNp6I6awCVZAVwDXAAsAruSLFTVfSPVLgW+XlXfn2QLcDVwcZKzgC3AK4CXAn+T5Aeq6qnl/kEktWkeb5bm8WeWhsi+qGn4e6KhmWYE7lxgT1XtBUiyA9gMjCZwm4Hf6Y9vBD6QJH35jqo6ADyYZE///T67PM2XJEmSTLTUlmNZ7z9NArcKeHjkfBF49aQ6VXUwyePAi/vynUteu+pI/9ndjzxuB5QkSZKkMaZJ4DKmrKasM81rSXIZcFl/euDfrr7oninapeE5DfjKrBuhZ8XYtcvYtcvYtcvYtcvYtWlu4par///wZZPqTJPALQJrRs5XA/sn1FlMchLwXcDXpnwtVXUtcC1Akt1VtWGKdmlgjF27jF27jF27jF27jF27jF2bjNvTfdsUdXYB65OsS3Iy3aYkC0vqLACX9MdvBG6rqurLtyRZmWQdsB743PI0XZIkSZLmy1FH4Po1be8AbqF7jMD2qro3yXuA3VW1AFwP/Gm/ScnX6JI8+nofp9vw5CDwdneglCRJkqRnZ6rnwFXVTcBNS8quHDn+H+BNE157FXDVM2jTtc+grobF2LXL2LXL2LXL2LXL2LXL2LXJuI1IN9NRkiRJkjR006yBkyRJkiQNwKASuCQbkzyQZE+Sy2fdHk2WZE2S25Pcn+TeJO/qy1+U5DNJ/qX/+sJZt1WHS7IiyZ1JPt2fr0tyRx+3P+s3LNLAJDk1yY1J/qnvez9un2tDkl/t3yvvSfKxJM+z3w1Tku1JHktyz0jZ2H6Wzh/19y13JTlndi3XhNi9t3/PvCvJXyY5deTa1j52DyT5mdm0WjA+diPX3p2kkpzWn899vxtMApdkBXAN8AbgLODnk5w121bpCA4Cv15VLwfOA97ex+ty4NaqWg/c2p9reN4F3D9yfjXwvj5uXwcunUmrdDTvB/6qqn4I+FG6GNrnBi7JKuCdwIaqeiXdhmBbsN8N1Q3AxiVlk/rZG+h22F5P9zzbDx6nNmq8Gzg8dp8BXllVPwL8M7AVoL9n2QK8on/NH/f3opqNGzg8diRZA1wAPDRSPPf9bjAJHHAusKeq9lbVk8AOYPOM26QJqurRqvpCf/yfdDeSq+hi9uG+2oeBn5tNCzVJktXAhcB1/XmA1wE39lWM2wAl+U7gp+h2/aWqnqyqb2Cfa8VJwHf0z0o9BXgU+90gVdXf0e2oPWpSP9sMfKQ6O4FTk3zv8WmplhoXu6r666o62J/upHsmMXSx21FVB6rqQWAP3b2oZmBCvwN4H/AbwOimHXPf74aUwK0CHh45X+zLNHBJzgDOBu4AvruqHoUuyQNeMruWaYI/pHsz/GZ//mLgGyN/4Ox7w3Qm8GXgT/rpr9cleT72ucGrqkeA36f7BPlR4HHg89jvWjKpn3nv0pZfAG7uj43dwCXZBDxSVV9ccmnuYzekBC5jytwic+CSvAD4C+BXquo/Zt0eHVmSi4DHqurzo8Vjqtr3huck4Bzgg1V1NvBfOF2yCf16qc3AOuClwPPppgAtZb9rj++fjUhyBd3yj48eKhpTzdgNRJJTgCuAK8ddHlM2V7EbUgK3CKwZOV8N7J9RWzSFJN9Ol7x9tKo+0Rd/6dAwdv/1sVm1T2O9BtiUZB/dNOXX0Y3IndpP7QL73lAtAotVdUd/fiNdQmefG77zgQer6stV9b/AJ4CfwH7Xkkn9zHuXBiS5BLgIeHN96/lZxm7Yvo/uQ68v9vcsq4EvJPkejN2gErhdwPp+V66T6RaWLsy4TZqgXzd1PXB/Vf3ByKUF4JL++BLgU8e7bZqsqrZW1eqqOoOuj91WVW8Gbgfe2FczbgNUVf8OPJzkB/ui1wP3YZ9rwUPAeUlO6d87D8XOfteOSf1sAXhrvyveecDjh6ZaahiSbAR+E9hUVU+MXFoAtiRZmWQd3YYYn5tFG3W4qrq7ql5SVWf09yyLwDn938K573eDepB3kp+lGw1YAWyvqqtm3CRNkOQngb8H7uZba6l+i24d3MeBtXQ3LW+qqnGLUjVjSV4LvLuqLkpyJt2I3IuAO4G3VNWBWbZPh0vyKrrNZ04G9gJvo/sgzj43cEl+F7iYbgrXncAv0q3ZsN8NTJKPAa8FTgO+BPw28EnG9LM+If8A3e55TwBvq6rds2i3JsZuK7AS+GpfbWdV/VJf/wq6dXEH6ZaC3Lz0e+r4GBe7qrp+5Po+up18v2K/G1gCJ0mSJEmabEhTKCVJkiRJR2ACJ0mSJEmNMIGTJEmSpEaYwEmSJElSI0zgJEmSJKkRJnCSJEmS1AgTOEmSJElqhAmcJEmSJDXi/wA4UGgf9O5IfwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1080x720 with 4 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# \"HBEGITS\"\n",
    "draw_ss_predicts(1,\"H\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>chip_id</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>dist_0</th>\n",
       "      <th>dist_1</th>\n",
       "      <th>dist_2</th>\n",
       "      <th>dist_3</th>\n",
       "      <th>...</th>\n",
       "      <th>solv_access_16</th>\n",
       "      <th>solv_access_17</th>\n",
       "      <th>solv_access_18</th>\n",
       "      <th>solv_access_19</th>\n",
       "      <th>solv_access_20</th>\n",
       "      <th>solv_access_21</th>\n",
       "      <th>solv_access_22</th>\n",
       "      <th>solv_access_23</th>\n",
       "      <th>solv_access_24</th>\n",
       "      <th>ss_majority_categorical</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>2.931498</td>\n",
       "      <td>4.466173</td>\n",
       "      <td>4.830971</td>\n",
       "      <td>5.929355</td>\n",
       "      <td>...</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.524194</td>\n",
       "      <td>0.080808</td>\n",
       "      <td>0.018293</td>\n",
       "      <td>0.396907</td>\n",
       "      <td>0.644330</td>\n",
       "      <td>0.463710</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_1</td>\n",
       "      <td>12</td>\n",
       "      <td>37</td>\n",
       "      <td>3.148242</td>\n",
       "      <td>4.942146</td>\n",
       "      <td>5.011508</td>\n",
       "      <td>6.109199</td>\n",
       "      <td>...</td>\n",
       "      <td>0.524194</td>\n",
       "      <td>0.080808</td>\n",
       "      <td>0.018293</td>\n",
       "      <td>0.396907</td>\n",
       "      <td>0.644330</td>\n",
       "      <td>0.463710</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_2</td>\n",
       "      <td>24</td>\n",
       "      <td>49</td>\n",
       "      <td>3.400494</td>\n",
       "      <td>6.638870</td>\n",
       "      <td>9.738158</td>\n",
       "      <td>12.659022</td>\n",
       "      <td>...</td>\n",
       "      <td>0.080808</td>\n",
       "      <td>0.018293</td>\n",
       "      <td>0.396907</td>\n",
       "      <td>0.644330</td>\n",
       "      <td>0.463710</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_3</td>\n",
       "      <td>36</td>\n",
       "      <td>61</td>\n",
       "      <td>3.473039</td>\n",
       "      <td>5.952474</td>\n",
       "      <td>7.897916</td>\n",
       "      <td>8.550533</td>\n",
       "      <td>...</td>\n",
       "      <td>0.018293</td>\n",
       "      <td>0.396907</td>\n",
       "      <td>0.644330</td>\n",
       "      <td>0.463710</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_4</td>\n",
       "      <td>48</td>\n",
       "      <td>73</td>\n",
       "      <td>3.076461</td>\n",
       "      <td>6.190106</td>\n",
       "      <td>9.162886</td>\n",
       "      <td>11.840832</td>\n",
       "      <td>...</td>\n",
       "      <td>0.396907</td>\n",
       "      <td>0.644330</td>\n",
       "      <td>0.463710</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_5</td>\n",
       "      <td>60</td>\n",
       "      <td>85</td>\n",
       "      <td>3.465978</td>\n",
       "      <td>6.825651</td>\n",
       "      <td>10.210162</td>\n",
       "      <td>13.602541</td>\n",
       "      <td>...</td>\n",
       "      <td>0.644330</td>\n",
       "      <td>0.463710</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_6</td>\n",
       "      <td>72</td>\n",
       "      <td>97</td>\n",
       "      <td>3.187567</td>\n",
       "      <td>4.704804</td>\n",
       "      <td>4.751395</td>\n",
       "      <td>5.968911</td>\n",
       "      <td>...</td>\n",
       "      <td>0.463710</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.103774</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_7</td>\n",
       "      <td>84</td>\n",
       "      <td>109</td>\n",
       "      <td>3.311531</td>\n",
       "      <td>5.471488</td>\n",
       "      <td>9.184711</td>\n",
       "      <td>12.485128</td>\n",
       "      <td>...</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.103774</td>\n",
       "      <td>0.110294</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_8</td>\n",
       "      <td>96</td>\n",
       "      <td>121</td>\n",
       "      <td>3.268903</td>\n",
       "      <td>4.931504</td>\n",
       "      <td>6.315815</td>\n",
       "      <td>9.363784</td>\n",
       "      <td>...</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.103774</td>\n",
       "      <td>0.110294</td>\n",
       "      <td>0.053254</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_9</td>\n",
       "      <td>108</td>\n",
       "      <td>133</td>\n",
       "      <td>3.271093</td>\n",
       "      <td>6.159716</td>\n",
       "      <td>9.532755</td>\n",
       "      <td>13.027534</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.103774</td>\n",
       "      <td>0.110294</td>\n",
       "      <td>0.053254</td>\n",
       "      <td>0.103659</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_10</td>\n",
       "      <td>120</td>\n",
       "      <td>145</td>\n",
       "      <td>3.066599</td>\n",
       "      <td>4.990942</td>\n",
       "      <td>5.992573</td>\n",
       "      <td>8.496792</td>\n",
       "      <td>...</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.103774</td>\n",
       "      <td>0.110294</td>\n",
       "      <td>0.053254</td>\n",
       "      <td>0.103659</td>\n",
       "      <td>0.115385</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11 rows × 422 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0  Unnamed: 0.1 protein_id  chip_id  start  stop    dist_0  \\\n",
       "0            0             0     12asA0   chip_0      0    25  2.931498   \n",
       "1            1             1     12asA0   chip_1     12    37  3.148242   \n",
       "2            2             2     12asA0   chip_2     24    49  3.400494   \n",
       "3            3             3     12asA0   chip_3     36    61  3.473039   \n",
       "4            4             4     12asA0   chip_4     48    73  3.076461   \n",
       "5            5             5     12asA0   chip_5     60    85  3.465978   \n",
       "6            6             6     12asA0   chip_6     72    97  3.187567   \n",
       "7            7             7     12asA0   chip_7     84   109  3.311531   \n",
       "8            8             8     12asA0   chip_8     96   121  3.268903   \n",
       "9            9             9     12asA0   chip_9    108   133  3.271093   \n",
       "10          10            10     12asA0  chip_10    120   145  3.066599   \n",
       "\n",
       "      dist_1     dist_2     dist_3  ...  solv_access_16  solv_access_17  \\\n",
       "0   4.466173   4.830971   5.929355  ...        0.076923        0.524194   \n",
       "1   4.942146   5.011508   6.109199  ...        0.524194        0.080808   \n",
       "2   6.638870   9.738158  12.659022  ...        0.080808        0.018293   \n",
       "3   5.952474   7.897916   8.550533  ...        0.018293        0.396907   \n",
       "4   6.190106   9.162886  11.840832  ...        0.396907        0.644330   \n",
       "5   6.825651  10.210162  13.602541  ...        0.644330        0.463710   \n",
       "6   4.704804   4.751395   5.968911  ...        0.463710        0.121951   \n",
       "7   5.471488   9.184711  12.485128  ...        0.121951        0.476190   \n",
       "8   4.931504   6.315815   9.363784  ...        0.476190        0.024390   \n",
       "9   6.159716   9.532755  13.027534  ...        0.024390        0.201183   \n",
       "10  4.990942   5.992573   8.496792  ...        0.201183        0.262887   \n",
       "\n",
       "    solv_access_18  solv_access_19  solv_access_20  solv_access_21  \\\n",
       "0         0.080808        0.018293        0.396907        0.644330   \n",
       "1         0.018293        0.396907        0.644330        0.463710   \n",
       "2         0.396907        0.644330        0.463710        0.121951   \n",
       "3         0.644330        0.463710        0.121951        0.476190   \n",
       "4         0.463710        0.121951        0.476190        0.024390   \n",
       "5         0.121951        0.476190        0.024390        0.201183   \n",
       "6         0.476190        0.024390        0.201183        0.262887   \n",
       "7         0.024390        0.201183        0.262887        0.014085   \n",
       "8         0.201183        0.262887        0.014085        0.136364   \n",
       "9         0.262887        0.014085        0.136364        0.103774   \n",
       "10        0.014085        0.136364        0.103774        0.110294   \n",
       "\n",
       "    solv_access_22  solv_access_23  solv_access_24  ss_majority_categorical  \n",
       "0         0.463710        0.121951        0.476190                        H  \n",
       "1         0.121951        0.476190        0.024390                     None  \n",
       "2         0.476190        0.024390        0.201183                     None  \n",
       "3         0.024390        0.201183        0.262887                     None  \n",
       "4         0.201183        0.262887        0.014085                     None  \n",
       "5         0.262887        0.014085        0.136364                     None  \n",
       "6         0.014085        0.136364        0.103774                     None  \n",
       "7         0.136364        0.103774        0.110294                     None  \n",
       "8         0.103774        0.110294        0.053254                     None  \n",
       "9         0.110294        0.053254        0.103659                     None  \n",
       "10        0.053254        0.103659        0.115385                        H  \n",
       "\n",
       "[11 rows x 422 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_25.loc[0:10]\n",
    "\n",
    "\n",
    "# def ramachandran_plot(pdb_id, degrees=True):\n",
    "#     df = get_torsion_angles(pdb_id, degrees)\n",
    "#     x = df['Phi']\n",
    "#     y = df['Psi']\n",
    "#     # Generate plot\n",
    "#     plt.plot(x, y, \".\")\n",
    "#     plt.title('Ramachandran Plot')\n",
    "#     if degrees:\n",
    "#         plt.xlabel(f'$\\Phi$ Angle (Degrees)')\n",
    "#         plt.xlim(-180, 180)\n",
    "#         plt.ylabel(f'$\\Psi$ Angle (Radians)')\n",
    "#         plt.ylim(-180, 180)\n",
    "#         plt.show()\n",
    "#     else:\n",
    "#         plt.xlabel(f'$\\Phi$ Angle (Radians)')\n",
    "#         plt.xlim(-math.pi, math.pi)\n",
    "#         plt.ylabel(f'$\\Psi$ Angle (Radians)')\n",
    "#         plt.ylim(-math.pi, math.pi)\n",
    "#         plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ground_truth_window_avg(df):\n",
    "    aa_length=max(df[\"stop\"])\n",
    "    mean_ss=[]\n",
    "    gt=[\"ground_truth_\"+ss for ss in \"HBEGITS\"]\n",
    "\n",
    "    for aa in range(0,aa_length):\n",
    "        chip=df[(df[\"start\"]<=aa) & (df[\"stop\"]>aa)]\n",
    "        mean_ss.append(chip[gt].mean(0))\n",
    "    gt_avg=pandas.DataFrame(mean_ss)\n",
    "    \n",
    "    return gt_avg\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each protein:\n",
    "#     for chip in protein\n",
    "#         get avg ss's for chip\n",
    "#     get avg ss for each aa\n",
    "#     get psipred for each aa\n",
    "#     present as bar graph\n",
    "\n",
    "#     get gt for protein !!!!\n",
    "\n",
    "#     get thd_rmse for protein\n",
    "#     get psipred_rmse for protein\n",
    "#     put rmse's in vectors\n",
    "\n",
    "# graph rmse's\n",
    "\n",
    "# the following is pseudocode\n",
    "\n",
    "def rmse(a,b):\n",
    "    return np.sqrt(np.mean((a-b)**2))\n",
    "\n",
    "def eval_test_data_new(leaf_df,dataset):\n",
    "    for protein in pandas.unique(leaf_df['protein_id']):\n",
    "        \n",
    "        mean_ss=[]\n",
    "        gt_names=[\"pred_gt_\"+ss for ss in \"HBEGITS\"]\n",
    "\n",
    "        # this part is avg'ing across the wrong stuff\n",
    "        # all the values end up the same\n",
    "        \n",
    "        for aa in range(length_dict[protein]):\n",
    "            chip=leaf_df[(leaf_df[\"start\"]<=aa) & (leaf_df[\"stop\"]>aa)]\n",
    "            mean_ss.append([protein]+list(chip[gt_names].mean(0)))\n",
    "        gt_avg=pandas.DataFrame(mean_ss)\n",
    "        gt_avg.columns=[\"protein_id\"]+[\"pred_ss_\"+ss for ss in \"HBEGITS\"]\n",
    "        \n",
    "        \n",
    "        ssgt=[]\n",
    "        \n",
    "        for sec_s in \"HBEGITS\":\n",
    "        # get gt for protein\n",
    "            ss=get_ground_truth_api(protein)['Secondary_Structure']\n",
    "            gt=[int(x) for x in ss==sec_s]\n",
    "            gt_avg[\"ground_truth_\"+sec_s]=gt\n",
    "           \n",
    "        gt_avg['psipred_helix']=input_features[protein][0].astype(np.float)\n",
    "        gt_avg['psipred_sheet']=input_features[protein][1].astype(np.float)\n",
    "        gt_avg['psipred_coil']=input_features[protein][2].astype(np.float)\n",
    "        \n",
    "        return gt_avg\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "    \n",
    "def seg_rows_formatted(thd_name,file_id,dataset):\n",
    "    seg=\"GROUPDATA_\"+thd_name+\" \"+file_id+\".json\"\n",
    "    return get_rows_from_segmentation(thd_name+\"/\"+seg,dataset)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_pred_gt(leafdf,dataset):\n",
    "\n",
    "    gt=[\"ground_truth_\"+ss for ss in \"HBEGITS\"]\n",
    "\n",
    "    ss_pred=pandas.DataFrame()\n",
    "\n",
    "    for index, row in leafdf.iterrows():\n",
    "        ssp=seg_rows_formatted(row['thd_id'],row['file_id'],dataset)[gt].mean(0)\n",
    "        ss_pred=ss_pred.append(ssp, ignore_index=True)\n",
    "        \n",
    "    ss_pred.columns=[\"pred_gt_\"+ss for ss in \"HBEGITS\"]\n",
    "    return pandas.concat([leafdf,ss_pred],axis=1)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [],
   "source": [
    "thd_leafs=add_pred_gt(pandas.read_csv(\"datasets/test6_sequences_pred.csv\"),dataset_5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "prediction_df=eval_test_data_new(thd_leafs,dataset_5)\n",
    "\n",
    "\n",
    "# 1rybA\n",
    "\n",
    "# input_features['1rybA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.5665646488490348\n",
      "0.23321427765305483\n",
      "0.4473615254274545\n",
      "0.36468050545726594\n",
      "8.880383122962865e-05\n",
      "0.3246723014279299\n",
      "0.2343651456926101\n"
     ]
    }
   ],
   "source": [
    "rmse_df=pandas.DataFrame()\n",
    "\n",
    "for ss in \"HBEGITS\":\n",
    "    \n",
    "    print(rmse(prediction_df[\"pred_ss_\"+ss],prediction_df[\"ground_truth_\"+ss]))\n",
    "    \n",
    "#     rmse(1-prediction_df[\"psipred_helix\"],prediction_df[\"ground_truth_\"+ss])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      0.002\n",
       "1      0.070\n",
       "2      0.197\n",
       "3      0.847\n",
       "4      0.981\n",
       "5      0.989\n",
       "6      0.978\n",
       "7      0.808\n",
       "8      0.532\n",
       "9      0.102\n",
       "10     0.021\n",
       "11     0.018\n",
       "12     0.053\n",
       "13     0.142\n",
       "14     0.165\n",
       "15     0.174\n",
       "16     0.217\n",
       "17     0.138\n",
       "18     0.112\n",
       "19     0.200\n",
       "20     0.334\n",
       "21     0.466\n",
       "22     0.640\n",
       "23     0.863\n",
       "24     0.978\n",
       "25     0.985\n",
       "26     0.991\n",
       "27     0.994\n",
       "28     0.990\n",
       "29     0.990\n",
       "       ...  \n",
       "156    0.936\n",
       "157    0.982\n",
       "158    0.984\n",
       "159    0.987\n",
       "160    0.991\n",
       "161    0.993\n",
       "162    0.993\n",
       "163    0.993\n",
       "164    0.992\n",
       "165    0.989\n",
       "166    0.986\n",
       "167    0.986\n",
       "168    0.990\n",
       "169    0.993\n",
       "170    0.993\n",
       "171    0.994\n",
       "172    0.995\n",
       "173    0.987\n",
       "174    0.983\n",
       "175    0.964\n",
       "176    0.892\n",
       "177    0.329\n",
       "178    0.025\n",
       "179    0.154\n",
       "180    0.253\n",
       "181    0.246\n",
       "182    0.230\n",
       "183    0.153\n",
       "184    0.079\n",
       "185    0.033\n",
       "Name: psipred_helix, Length: 186, dtype: float64"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
