{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n",
    "import math\n",
    "import imageio\n",
    "import json\n",
    "import time\n",
    "\n",
    "from collections import Counter\n",
    "from matplotlib.pyplot import figure\n",
    "from matplotlib import animation\n",
    "\n",
    "from Bio.PDB import *\n",
    "import nglview as nv\n",
    "import math\n",
    "import warnings\n",
    "from Bio.PDB.StructureBuilder import PDBConstructionWarning\n",
    "warnings.filterwarnings(\"ignore\", category=PDBConstructionWarning)\n",
    "\n",
    "# this is dangerous of course but uncomment it when you want to run the code\n",
    "# warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "\n",
    "\n",
    "\n",
    "# seg_dir = \"C:\\Users\\Kevin\\Documents\\TDA\\afrl\\Segmentations\"\n",
    "seg_dir = \"C:/Users/Kevin/Documents/TDA/afrl/Segmentations\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# aaron's code to make sure the folders are all in their proper place\n",
    "\n",
    "dirlocal = os.path.curdir\n",
    "data_path = os.path.join(dirlocal, 'Data/')\n",
    "protein_path = os.path.join(data_path, 'Proteins/')\n",
    "pdb_path = os.path.join(data_path, 'PDB/')\n",
    "fasta_path = os.path.join(data_path, 'fasta/')\n",
    "dssp_path = os.path.join(data_path, 'DSSP/')\n",
    "\n",
    "if not os.path.exists(data_path):\n",
    "    os.makedirs(data_path)\n",
    "if not os.path.exists(protein_path):\n",
    "    os.makedirs(protein_path)\n",
    "if not os.path.exists(pdb_path):\n",
    "    os.makedirs(pdb_path)\n",
    "ss_path = os.path.join(data_path, 'ss.txt')\n",
    "feature_path = os.path.join(data_path, 'sample-input-features.npy')\n",
    "distance_path = os.path.join(data_path, 'sample-distance-maps-cb.npy')\n",
    "full_feature_path = os.path.join(data_path, 'full-input-features.npy')\n",
    "full_distance_path = os.path.join(data_path, 'full-distance-maps-cb.npy')\n",
    "test_feature_path = os.path.join(data_path, 'testset-input-features.npy')\n",
    "test_distance_path = os.path.join(data_path, 'testset-distance-maps-cb.npy')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## aaron's code to get raw features from pdb\n",
    "\n",
    "import numpy as np\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from matplotlib.pyplot import figure\n",
    "%matplotlib inline\n",
    "import time\n",
    "import requests\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "# ! pip install biopython\n",
    "# ! pip install nglview\n",
    "# ! jupyter-nbextension enable nglview --py --sys-prefix\n",
    "\n",
    "from Bio.PDB import *\n",
    "import nglview as nv\n",
    "import math\n",
    "import warnings\n",
    "from Bio.PDB.StructureBuilder import PDBConstructionWarning\n",
    "warnings.filterwarnings(\"ignore\", category=PDBConstructionWarning)\n",
    "\n",
    "# this is dangerous of course but uncomment it when you want to run the code\n",
    "# warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_chip_start_stop(df, step_size, chip_size):\n",
    "\n",
    "    ch=[int(x.split(\"_\")[-1]) for x in df['chip_id']]\n",
    "\n",
    "    start=[c*step_size for c in ch]\n",
    "    stop=[chip_size+c*step_size for c in ch]\n",
    "\n",
    "    df['start']=start\n",
    "    df['stop']=stop\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "###  aaron's functions: \n",
    "\n",
    "# gets dataframe containing torsion angles, peptides, chain, etc\n",
    "# from pdb file\n",
    "def get_torsion_angles(pdb_id, degrees=False):\n",
    "    pdb_struct = get_pdb_structure(pdb_id)\n",
    "    torsion_angles = []\n",
    "    for model in pdb_struct:\n",
    "        for chain in model:\n",
    "            polypeptides = PPBuilder().build_peptides(chain)\n",
    "            for poly_index, poly in enumerate(polypeptides):\n",
    "                #print(\"Model %s Chain %s\" % (str(model.id), str(chain.id)))\n",
    "                #print(\"(part %i of %i)\" % (poly_index+1, len(polypeptides)))\n",
    "                #print(\"length %i\" % (len(poly)))\n",
    "                #print(\"from %s%i\" % (poly[0].resname, poly[0].id[1]))\n",
    "                #print(\"to %s%i\" % (poly[-1].resname, poly[-1].id[1]))\n",
    "                phi_psi = poly.get_phi_psi_list()\n",
    "                for res_index, residue in enumerate(poly):\n",
    "                    res_name = \"%s%i\" % (residue.resname, residue.id[1])\n",
    "                    #print(res_name, tuple(math.degrees(b) for b in phi_psi[res_index] if b))\n",
    "                    deg = phi_psi[res_index]\n",
    "                    if degrees:\n",
    "                        deg = tuple(math.degrees(b) if b else None for b in deg)\n",
    "                    phi, psi = deg\n",
    "                    model_name, model_id = model.full_id\n",
    "                    torsion_angles.append([model_name, model_id, chain.id, residue.resname, residue.id[1], phi, psi])\n",
    "    return pd.DataFrame(torsion_angles, columns=['Model_Name', 'Model_ID', 'Chain', 'Residue_Name', 'Residue_ID', 'Phi', 'Psi'])\n",
    "\n",
    "\n",
    "# generates ramachandran plot from a given pdb file\n",
    "def ramachandran_plot(pdb_id, degrees=True):\n",
    "    df = get_torsion_angles(pdb_id, degrees)\n",
    "    x = df['Phi']\n",
    "    y = df['Psi']\n",
    "    # Generate plot\n",
    "    plt.plot(x, y, \".\")\n",
    "    plt.title('Ramachandran Plot')\n",
    "    if degrees:\n",
    "        plt.xlabel(f'$\\Phi$ Angle (Degrees)')\n",
    "        plt.xlim(-180, 180)\n",
    "        plt.ylabel(f'$\\Psi$ Angle (Radians)')\n",
    "        plt.ylim(-180, 180)\n",
    "        plt.show()\n",
    "    else:\n",
    "        plt.xlabel(f'$\\Phi$ Angle (Radians)')\n",
    "        plt.xlim(-math.pi, math.pi)\n",
    "        plt.ylabel(f'$\\Psi$ Angle (Radians)')\n",
    "        plt.ylim(-math.pi, math.pi)\n",
    "        plt.show()\n",
    "\n",
    "# helper function which generates a biopython structure\n",
    "# for display in display_protein()\n",
    "def get_pdb_structure(pdb_id):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    parser = PDBParser()\n",
    "    file_path = os.path.join(pdb_path, f\"{pdb_id}.pdb\")\n",
    "    try:\n",
    "        struct = parser.get_structure(pdb_id, file_path)\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "        return None\n",
    "    return struct\n",
    "\n",
    "\n",
    "#  downloads the protein in question from pdb\n",
    "def get_pdb_file(pdb_id, replace=False):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    file_path = os.path.join(pdb_path, f'{pdb_id}.pdb')\n",
    "    if os.path.isfile(file_path) and not replace:\n",
    "        return True\n",
    "    parser = PDBParser()\n",
    "    url = f'https://files.rcsb.org/download/{pdb_id}.pdb'\n",
    "    resp = requests.get(url)\n",
    "    try:\n",
    "        #print(file_path)\n",
    "        file = open(file_path, \"wb\")\n",
    "        file.write(resp.content)\n",
    "        file.close()\n",
    "    except Exception as e: \n",
    "        print(e)\n",
    "        return False\n",
    "    return True\n",
    "\n",
    "# neato 3d viewer of a protein's structure\n",
    "def display_protein(pdb_id):\n",
    "    pdb_struct = get_pdb_structure(pdb_id)\n",
    "    view = nv.show_biopython(pdb_struct)\n",
    "    return view\n",
    "\n",
    "\n",
    "#  returns a dataframe with secondary structure, amino acid code, chain id, etc\n",
    "# deprecated\n",
    "def get_secondary_structure(pdb_id):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    indexes = get_pdb_ss_seq(pdb_id)\n",
    "    chain_df = []\n",
    "    for i, chain_ss in enumerate(indexes):\n",
    "        chain_id, seq, ss = chain_ss\n",
    "        #unknown_list = [i for i, c in enumerate(seq) if c == 'X']\n",
    "        #print(unknown_list)\n",
    "        seq_list = [c for c in seq]\n",
    "        # [f(x) if condition else g(x) for x in sequence]\n",
    "        ss_list = ['L' if c is ' ' else c for c in ss]\n",
    "        #print(seq_list + ss_list)\n",
    "        chain_df.append(pd.DataFrame({'Chain': chain_id, 'Amino_Acid':seq_list, 'Secondary_Structure':ss_list}))\n",
    "    return pd.concat(chain_df)\n",
    "\n",
    "# get matrix of pairwise carbon beta distances\n",
    "def get_cb_distances(full_pdb_id, partial=False):\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    get_pdb_file(pdb_id)\n",
    "    pdb_struct = get_pdb_structure(pdb_id)  # Returns the BioPython structure of the PDB file\n",
    "    ppb = PPBuilder()\n",
    "    chain_dist_matrix = {}\n",
    "    for model in pdb_struct:\n",
    "        for chain in model:\n",
    "            if chain.id == full_pdb_id[4] and partial:\n",
    "                residues = []\n",
    "                for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                    for residue in pp:\n",
    "                        residues.append(residue)\n",
    "                lng = len(residues)\n",
    "                dist_matrix = np.zeros((lng, lng), np.float)\n",
    "                for i in range(0, lng):\n",
    "                    for j in range(i, lng):\n",
    "                        try:\n",
    "                            diff_vector  = residues[i]['C'].coord - residues[j]['C'].coord\n",
    "                        except Exception:\n",
    "                            # This just gets the average value around the unknown residue distance\n",
    "                            count = 0\n",
    "                            sums = 0\n",
    "                            if j - 1 > 0:\n",
    "                                count += 1\n",
    "                                sums += dist_matrix[i][j - 1]\n",
    "                            if i - 1 > 0:\n",
    "                                count += 1\n",
    "                                sums += dist_matrix[i - 1][j]\n",
    "                            if count > 0:\n",
    "                                diff_vector = sums / count\n",
    "                            else:\n",
    "                                diff_vector = 0\n",
    "                        dist = np.sqrt(np.sum(diff_vector * diff_vector))\n",
    "                        dist_matrix[i][j] = dist\n",
    "                        dist_matrix[j][i] = dist\n",
    "                    chain_dist_matrix = dist_matrix\n",
    "        \n",
    "    if partial:\n",
    "        start_pos, end_pos, bool_arr = get_seq_pos(full_pdb_id)\n",
    "        if start_pos == -1:\n",
    "            return pd.DataFrame(chain_dist_matrix[bool_arr]).iloc[:, bool_arr]\n",
    "        return pd.DataFrame(chain_dist_matrix[start_pos:end_pos]).iloc[:, start_pos:end_pos]\n",
    "    else:\n",
    "        return chain_dist_matrix\n",
    "    \n",
    "    \n",
    "# gets alignment of given sequence on top of the original sequence\n",
    "# from a fasta file downloaded from the cloud repository\n",
    "def get_seq_alignment(full_pdb_id):\n",
    "    seq = ''\n",
    "    for record in SeqIO.parse(os.path.join(fasta_path, f'{full_pdb_id}.fasta'), \"fasta\"):\n",
    "        seq = record.seq\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    #print(pdb_id)\n",
    "    get_pdb_file(pdb_id)\n",
    "    struct = get_pdb_structure(pdb_id)\n",
    "    ppb = PPBuilder()\n",
    "    for model in struct:\n",
    "        for chain in model:\n",
    "            if chain.id == full_pdb_id[4]:\n",
    "                ground_seq = ''\n",
    "                for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                    ground_seq += pp.get_sequence()\n",
    "                alignments = pairwise2.align.globalxx(ground_seq, str(seq))\n",
    "                al_seq = alignments[0][1]\n",
    "                al_seq_bool = [False if a is '-' else True for a in al_seq]\n",
    "                return al_seq_bool\n",
    "\n",
    "            \n",
    "def get_seq_pos(full_pdb_id):\n",
    "    seq = ''\n",
    "    for record in SeqIO.parse(os.path.join(fasta_path, f'{full_pdb_id}.fasta'), \"fasta\"):\n",
    "        seq = str(record.seq)\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    get_pdb_file(pdb_id)\n",
    "    struct = get_pdb_structure(pdb_id)\n",
    "    ppb = PPBuilder()\n",
    "    for model in struct:\n",
    "        for chain in model:\n",
    "            if chain.id == full_pdb_id[4]:\n",
    "                ground_seq = ''\n",
    "                for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                    ground_seq += pp.get_sequence()\n",
    "                start_pos = ground_seq.find(seq)\n",
    "                if start_pos == -1:\n",
    "                    alignments = pairwise2.align.globalxx(ground_seq, seq)\n",
    "                    al_seq = alignments[0][1]\n",
    "                    bool_seq = [False if a == '-' else True for a in al_seq]\n",
    "                    #start_pos = largest_substring_pos(ground_seq, seq)\n",
    "                    #start_pos = int(alignments[0][2] - 1)\n",
    "                    #end_pos = start_pos + len(al_seq)\n",
    "                    return (-1, -1, bool_seq)\n",
    "                else:\n",
    "                    end_pos = start_pos + len(seq)\n",
    "                    return (start_pos, end_pos, '')\n",
    "\n",
    "\n",
    "            \n",
    "            \n",
    "# calls dssp api to get dssp file\n",
    "def pdb_id_to_dssp_file(pdb_id, replace=False):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    rest_url = 'http://www.cmbi.umcn.nl/xssp/'\n",
    "    # Read the pdb id data into a variable\n",
    "    data = {'data': pdb_id}\n",
    "\n",
    "    # Send a request to the server to create hssp data from the pdb file data.\n",
    "    # If an error occurs, an exception is raised and the program exits. If the\n",
    "    # request is successful, the id of the job running on the server is\n",
    "    # returned.\n",
    "    url_create = f'{rest_url}api/create/pdb_id/dssp/'\n",
    "    r = requests.post(url_create, data=data)\n",
    "    r.raise_for_status()\n",
    "\n",
    "    job_id = json.loads(r.text)['id']\n",
    "    #print(f'Job submitted successfully. Id is: {job_id}')\n",
    "\n",
    "    # Loop until the job running on the server has finished, either successfully\n",
    "    # or due to an error.\n",
    "    ready = False\n",
    "    while not ready:\n",
    "        # Check the status of the running job. If an error occurs an exception\n",
    "        # is raised and the program exits. If the request is successful, the\n",
    "        # status is returned.\n",
    "        url_status = f'{rest_url}api/status/pdb_id/dssp/{job_id}/'\n",
    "        r = requests.get(url_status)\n",
    "        r.raise_for_status()\n",
    "\n",
    "        status = json.loads(r.text)['status']\n",
    "        #print(f'Job status is: {status}')\n",
    "\n",
    "        # If the status equals SUCCESS, exit out of the loop by changing the\n",
    "        # condition ready. This causes the code to drop into the `else` block\n",
    "        # below.\n",
    "        #\n",
    "        # If the status equals either FAILURE or REVOKED, an exception is raised\n",
    "        # containing the error message. The program exits.\n",
    "        #\n",
    "        # Otherwise, wait for five seconds and start at the beginning of the\n",
    "        # loop again.\n",
    "        if status == 'SUCCESS':\n",
    "            ready = True\n",
    "        elif status in ['FAILURE', 'REVOKED']:\n",
    "            raise Exception(json.loads(r.text)['message'])\n",
    "        else:\n",
    "            time.sleep(5)\n",
    "    else:\n",
    "        # Requests the result of the job. If an error occurs an exception is\n",
    "        # raised and the program exits. If the request is successful, the result\n",
    "        # is returned.\n",
    "        url_result = f'{rest_url}api/result/pdb_id/dssp/{job_id}/'\n",
    "        r = requests.get(url_result)\n",
    "        r.raise_for_status()\n",
    "        result = json.loads(r.text)['result']\n",
    "        try:\n",
    "            file_path = os.path.join(dssp_path, f'{pdb_id}.dssp')\n",
    "            if os.path.isfile(file_path) and not replace:\n",
    "                return True\n",
    "            #print(file_path)\n",
    "            file = open(file_path, \"w\")\n",
    "            file.write(result)\n",
    "            file.close()\n",
    "        except Exception as e: \n",
    "            print(e)\n",
    "            return False\n",
    "        # Return the result to the caller, which prints it to the screen.\n",
    "        return True\n",
    "\n",
    "# uses api call file to generate secondary structure, phi/psi, solvent-accessibility stuff\n",
    "# because of indexing stuff, it crashes if we use partial = True\n",
    "# ask aaron about that\n",
    "def get_ground_truth_api(full_pdb_id, partial=False):\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    get_pdb_file(pdb_id)\n",
    "    struct = get_pdb_structure(pdb_id)\n",
    "    ppb = PPBuilder()\n",
    "    struct_info = []\n",
    "    for model in struct:\n",
    "        file = os.path.join(pdb_path, f'{pdb_id}.pdb')\n",
    "        pdb_id_to_dssp_file(pdb_id)\n",
    "        dssp = DSSP(model=model, in_file=os.path.join(dssp_path, f'{pdb_id}.dssp'), file_type='DSSP')\n",
    "        seq_count = 0\n",
    "        #dssp = DSSP(model=model, in_file=file)\n",
    "        for chain in model:\n",
    "            if chain.id == full_pdb_id[4] or not partial:\n",
    "                residues = []\n",
    "                seq = ''\n",
    "                for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                    seq += pp.get_sequence()\n",
    "                    for residue in pp:\n",
    "                        residues.append(residue)\n",
    "                for i, residue in enumerate(residues):\n",
    "                    try:\n",
    "                        key = list(dssp.keys())[seq_count]   \n",
    "                        dssp_info = dssp[key]\n",
    "                        amino_acid = dssp_info[1]\n",
    "                        sec_struct = dssp_info[2]\n",
    "                        solv_acc = dssp_info[3]\n",
    "                        phi = dssp_info[4]\n",
    "                        psi = dssp_info[5]\n",
    "                    except Exception:\n",
    "                        # DSSP didn't capture this amino acid\n",
    "                        #print(full_pdb_id, seq_count, len(dssp), len(seq), residue)\n",
    "                        amino_acid = seq[i]\n",
    "                        sec_struct = '-'\n",
    "                        solv_acc = 0\n",
    "                        phi = 360\n",
    "                        psi = 360\n",
    "                    seq_count += 1\n",
    "                    # Keys 6 through 13 is bonding energy / relidx (no clue what this is)\n",
    "                    struct_info.append([model.full_id[0], model.full_id[1], chain.id,\n",
    "                        residue.resname, residue.id[1], amino_acid, sec_struct, solv_acc, phi, psi])\n",
    "    info_df = pd.DataFrame(struct_info, \n",
    "        columns=['Model_Name', 'Model_ID', 'Chain', 'Residue_Name',\n",
    "        'Residue_ID', 'Amino_Acid', 'Secondary_Structure', 'Solvent_Accessability', \n",
    "        'Phi', 'Psi'])\n",
    "    if partial:    \n",
    "        start_pos, end_pos, bool_arr = get_seq_pos(full_pdb_id)\n",
    "        if start_pos == -1:\n",
    "            return info_df[bool_arr]\n",
    "        return info_df[start_pos:end_pos]\n",
    "    else:\n",
    "        return info_df\n",
    "\n",
    "\n",
    "    \n",
    "# the following two functions download the files from the api\n",
    "# to create a cache on the disk\n",
    "# speeds things up immensely, only need to run once though\n",
    "def get_ground_truth_files(full_pdb_id, replace=False, partial=False):\n",
    "    bad_list = ['4osnA0']\n",
    "    if full_pdb_id in bad_list:\n",
    "        return 'Bad File'\n",
    "    if partial:\n",
    "        file_path = os.path.join(protein_path, full_pdb_id + '.csv')\n",
    "    else:\n",
    "        file_path = os.path.join(protein_path, full_pdb_id[0:4] + '.csv')\n",
    "        \n",
    "    if os.path.isfile(file_path) and not replace:\n",
    "        return True\n",
    "    \n",
    "    df = get_ground_truth(full_pdb_id, partial)\n",
    "    file_path = ''\n",
    "    try:\n",
    "        df.to_csv(file_path)\n",
    "    except Exception:\n",
    "        return False\n",
    "    return True\n",
    "    \n",
    "def get_cb_distance_files(full_pdb_id, replace=False, partial=False):\n",
    "    bad_list = ['4osnA0']\n",
    "    if full_pdb_id in bad_list:\n",
    "        return 'Bad File'\n",
    "    file_path = ''\n",
    "    if partial:\n",
    "        file_path = os.path.join(distances_path, full_pdb_id + '.csv')\n",
    "        if os.path.isfile(file_path) and not replace:\n",
    "            return True\n",
    "        try:\n",
    "            df = get_cb_distances(full_pdb_id, partial)\n",
    "            df.to_csv(file_path)\n",
    "        except Exception:\n",
    "            return False\n",
    "        return True\n",
    "    else:\n",
    "        df = get_cb_distances(full_pdb_id, partial)\n",
    "        keys = df.keys\n",
    "        for key in keys:\n",
    "            file_path = os.path.join(distances_path, full_pdb_id[0:4] + key + '0.csv')\n",
    "            if os.path.isfile(file_path) and not replace:\n",
    "                return True\n",
    "            try:\n",
    "                df[key].to_csv(file_path)\n",
    "            except Exception:\n",
    "                return False\n",
    "        return True\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_25=pandas.read_csv(\"datasets/protein_all_25_12_no_na.csv\")\n",
    "dataset_10=pandas.read_csv(\"datasets/protein_all_10_5_no_na.csv\")\n",
    "dataset_5=pandas.read_csv(\"datasets/protein_all_5_3_no_na.csv\")\n",
    "\n",
    "fix_chip_start_stop(dataset_25,12,25)\n",
    "fix_chip_start_stop(dataset_10,5,10)\n",
    "fix_chip_start_stop(dataset_5,3,5)\n",
    "\n",
    "\n",
    "pdb_list, length_dict, aa_seq, input_features = np.load(\"datasets/testset-input-features.npy\",allow_pickle=True)\n",
    "pdb_list_y, distance_maps_cb = np.load(\"datasets/testset-distance-maps-cb.npy\",encoding=\"latin1\",allow_pickle=True)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions\n",
    "def get_rows_from_segmentation(seg):\n",
    "    with open(seg_dir+seg) as jsonf:\n",
    "        a=json.load(jsonf)\n",
    "        df=dataset.iloc[[int(m) for m in a['rowList']],:]\n",
    "        return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ground_truth_window_avg(df):\n",
    "    aa_length=max(df[\"stop\"])\n",
    "    mean_ss=[]\n",
    "    gt=[\"ground_truth_\"+ss for ss in \"HBEGITS\"]\n",
    "\n",
    "    for aa in range(0,aa_length):\n",
    "        chip=df[(df[\"start\"]<=aa) & (df[\"stop\"]>aa)]\n",
    "        mean_ss.append(chip[gt].mean(0))\n",
    "    gt_avg=pandas.DataFrame(mean_ss)\n",
    "    \n",
    "    return gt_avg\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def draw_ss_predicts(index,ss):\n",
    "    \n",
    "#     H, G and I to H; E to E; the rest to C\n",
    "    ss_dict={\n",
    "        \"H\":0,\n",
    "        \"G\":0,\n",
    "        \"I\":0,\n",
    "        \"E\":1,\n",
    "        \"B\":2,\n",
    "        \"T\":2,\n",
    "        \"S\":2             \n",
    "            }\n",
    "    \n",
    "    aa_length=length_dict[pdb_list[index]]\n",
    "    \n",
    "    plt.figure(figsize=(15,10))\n",
    "\n",
    "    df25=dataset_25[dataset_25[\"protein_id\"]==dataset_25[\"protein_id\"].unique()[index]] \n",
    "    df10=dataset_10[dataset_10[\"protein_id\"]==dataset_10[\"protein_id\"].unique()[index]] \n",
    "    df5=dataset_5[dataset_5[\"protein_id\"]==dataset_5[\"protein_id\"].unique()[index]] \n",
    "\n",
    "    plt.subplot(4,1,1)\n",
    "    y=ground_truth_window_avg(df5)[\"ground_truth_\"+ss]\n",
    "    x=range(0,len(y))\n",
    "    plt.bar(x,y,1)\n",
    "    plt.xlim((0,aa_length))    \n",
    "\n",
    "    plt.subplot(4,1,2)\n",
    "    y=ground_truth_window_avg(df10)[\"ground_truth_\"+ss]\n",
    "    x=range(0,len(y))\n",
    "    plt.bar(x,y,1)\n",
    "    plt.xlim((0,aa_length))\n",
    "\n",
    "    plt.subplot(4,1,3)\n",
    "    y=ground_truth_window_avg(df25)[\"ground_truth_\"+ss]\n",
    "    x=range(0,len(y))\n",
    "    plt.bar(x,y,1)\n",
    "    plt.xlim((0,aa_length))\n",
    "\n",
    "    plt.subplot(4,1,4)\n",
    "    y=[1-n for n in input_features[pdb_list[index]][ss_dict[ss]].astype(np.float)]\n",
    "    x=range(0,len(y))\n",
    "    plt.bar(x,y,1)\n",
    "    plt.xlim((0,aa_length))\n",
    "\n",
    "    \n",
    "    \n",
    "def get_inv_psipred_from_ss(protein):\n",
    "#     H, G and I to H; E to E; the rest to C\n",
    "    ss_dict={\n",
    "        \"H\":0,\n",
    "        \"G\":0,\n",
    "        \"I\":0,\n",
    "        \"E\":1,\n",
    "        \"B\":2,\n",
    "        \"T\":2,\n",
    "        \"S\":2             \n",
    "            }\n",
    "    \n",
    "    # I subtract it from 1 to make it on the same scale as the other predictions\n",
    "    # (i.e. high means more likely to be X)\n",
    "    \n",
    "    return [1-n for n in input_features[protein][ss_dict[ss]].astype(np.float)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each protein:\n",
    "#     for chip in protein\n",
    "#         get avg ss's for chip\n",
    "#     get avg ss for each aa\n",
    "#     get psipred for each aa\n",
    "#     present as bar graph\n",
    "\n",
    "#     get gt for protein !!!!\n",
    "\n",
    "#     get thd_rmse for protein\n",
    "#     get psipred_rmse for protein\n",
    "#     put rmse's in vectors\n",
    "\n",
    "# graph rmse's\n",
    "\n",
    "# the following is pseudocode\n",
    "\n",
    "def rmse(a,b):\n",
    "    return np.sqrt(np.mean((a-b)**2))\n",
    "\n",
    "\n",
    "def eval_test_data(leaf_df):\n",
    "    # for each protein:\n",
    "    rmse_df=pandas.DataFrame()\n",
    "    \n",
    "    for protein in unique(leaf_df['protein_id']):\n",
    "        # get avg ss for each aa\n",
    "        avg_ss=ground_truth_window_avg(leaf_df[leaf_df['protein_id']==protein])\n",
    "        \n",
    "            \n",
    "        # get psipred for each aa\n",
    "        # we'll get this later, just focus on the other stuff now\n",
    "#         get_inv_psipred_from_ss()\n",
    "        \n",
    "        # present as bar graph\n",
    "\n",
    "        gt_df=Pandas.DataFrame()\n",
    "        \n",
    "        for sec_s in \"HBEGITS\":\n",
    "        # get gt for protein\n",
    "            ss=get_ground_truth_api(protein)['Secondary_Structure']\n",
    "            gt=[int(x) for x in ss==sec_s]\n",
    "            gt_df=pd.concat(gt_df,gt)\n",
    "\n",
    "        # get thd_rmse's for protein\n",
    "        thd_rmse=[rmse(avg_ss[ss_x],gt_df[ss_x]) for ss_x in len(\"HBEGITS\")]\n",
    "        \n",
    "        rmse_df.append([protein]+thd_rmse)\n",
    "        \n",
    "    rmse_df.columns=[\"protein_id\"]+[\"RMSE_\"+ss for ss in \"HBEGITS\"]\n",
    "        \n",
    "    return rmse_df\n",
    "        # get psipred_rmse for protein\n",
    "        # put rmse's in vectors\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA3AAAAI/CAYAAAA2kzvaAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nOzdf6yk910f+vcnaxx0SS443QVFtjdr2m0VA5WduzK5igThkh/r+MpLVbhdI4rDTbu9lV3awr3SplQOMreVCaJRUU2apawSUIkxaSnbeqnJzQ9xdanDrsE3ydpyWYyJTzbCCU7TSgH7rvO5f5xxMj4+5+zszuzMPGdeL+nozPN9npn5nDnf+c685/s8z1R3BwAAgOX3skUXAAAAwGQEOAAAgIEQ4AAAAAZCgAMAABgIAQ4AAGAgBDgAAICBuGLRBWy0e/fu3rdv36LLAAAAWIiHH374C929Z7N1Sxfg9u3bl9OnTy+6DAAAgIWoqj/eat0Fd6GsquNV9XRVfXqL9VVVP1dVZ6vqk1X1urF1t1fVH4x+br+08gEAAEgmOwbu/UkObrP+5iT7Rz9Hkrw3SarqVUneleQ7k9yU5F1VddU0xQIAAKyyCwa47v7tJM9ss8mhJL/U6x5K8k1V9eokb03y4e5+pru/mOTD2T4IAgAAsI1ZnIXy6iRPjS2vjdq2agcAAOASzOIkJrVJW2/T/tIbqDqS9d0vs3fv3hmUxCLtO/rAokvIk/fcsugSdpzx/6vHd92l9HWPHZfb5RyD9d/Z2/j/WqbH2Ov5sCxzX1qUnfreZRYzcGtJrh1bvibJuW3aX6K7j3X3ge4+sGfPpmfLBAAAWHmzCHAnkvzw6GyUr0/ype7+XJIHk7ylqq4anbzkLaM2AAAALsEFd6Gsqg8meWOS3VW1lvUzS35dknT3v0xyMsnbkpxN8uUkPzJa90xV/VSSU6Oburu7tzsZCgAAANu4YIDr7tsusL6T3LHFuuNJjl9aaQAAAIybxS6UAAAAzIEABwAAMBACHAAAwEAIcAAAAAMhwAEAAAyEAAcAADAQF/waAYBVtu/oA4suAQAu2lavX0/ec8ucK7n8Jnmt3m6boT0mZuAAAAAGQoADAAAYCAEOAABgIAQ4AACAgRDgAAAABkKAAwAAGAgBDgAAYCAEOAAAgIEQ4AAAAAZCgAMAABiIKxZdAFwO+44+sGn7k/fcMudKdqatHt9kZzzG2/19l+u2d+LjthP+pmV2Ofspi7OTxwiWw3gfG2q/mvX4N7THxAwcAADAQEwU4KrqYFU9XlVnq+roJuvfU1WPjH7+c1X9l7F1z4+tOzHL4gEAAFbJBXehrKpdSe5N8uYka0lOVdWJ7n70hW26+x+Obf/3ktw4dhN/1t03zK5kAACA1TTJDNxNSc529xPd/VyS+5Ic2mb725J8cBbFAQAA8DWTBLirkzw1trw2anuJqnpNkuuSfHSs+eur6nRVPVRV33fJlQIAAKy4Sc5CWZu09RbbHk7yoe5+fqxtb3efq6pvTfLRqvpUd//hi+6g6kiSI0myd+/eCUoCAABYPZPMwK0luXZs+Zok57bY9nA27D7Z3edGv59I8vG8+Pi4F7Y51t0HuvvAnj17JigJAABg9UwS4E4l2V9V11XVlVkPaS85m2RV/ZUkVyX5T2NtV1XVy0eXdyd5Q5JHN14XAACAC7vgLpTdfb6q7kzyYJJdSY5395mqujvJ6e5+IczdluS+7h7fvfK1Sd5XVV/Jeli8Z/zslQAAAExukmPg0t0nk5zc0HbXhuWf3OR6v5PkO6aoDwAAgJGJvsgbAACAxZtoBg42s+/oA4sugRnzP2WWtupPT95zy5wrYVa2GyP8X7e3zOPrUGtb1T63zP8v5sMMHAAAwEAIcAAAAAMhwAEAAAyEAAcAADAQAhwAAMBACHAAAAADIcABAAAMhAAHAAAwEAIcAADAQAhwAAAAA3HFoguAedp39IEXLT95zy0LqmTnGn+MPb6TW6XHbZX+1lnbOIaxOjxvgBeYgQMAABgIAQ4AAGAgBDgAAICBEOAAAAAGQoADAAAYCAEOAABgICYKcFV1sKoer6qzVXV0k/Vvr6rPV9Ujo5+/Nbbu9qr6g9HP7bMsHgAAYJVc8HvgqmpXknuTvDnJWpJTVXWiux/dsOmvdvedG677qiTvSnIgSSd5eHTdL86kegAAgBUyyQzcTUnOdvcT3f1ckvuSHJrw9t+a5MPd/cwotH04ycFLKxUAAGC1TRLgrk7y1Njy2qhto79eVZ+sqg9V1bUXeV0AAAAu4IK7UCapTdp6w/K/T/LB7n62qv63JB9I8j9NeN1U1ZEkR5Jk7969E5QEMJ19Rx9YdAmD5bED2Bm2G8+fvOeWOVayuUW83iz7Y5JMNgO3luTaseVrkpwb36C7/7S7nx0t/kKS/2HS646uf6y7D3T3gT179kxaOwAAwEqZJMCdSrK/qq6rqiuTHE5yYnyDqnr12OKtSR4bXX4wyVuq6qqquirJW0ZtAAAAXKQL7kLZ3eer6s6sB69dSY5395mqujvJ6e4+keRHq+rWJOeTPJPk7aPrPlNVP5X1EJgkd3f3M5fh7wAAANjxJjkGLt19MsnJDW13jV1+Z5J3bnHd40mOT1EjAAAAmfCLvAEAAFg8AQ4AAGAgBDgAAICBEOAAAAAGQoADAAAYCAEOAABgICb6GgFIkn1HH1h0CVwG/q/ALIyPJU/ec8sCK1kOyz62Lnt9k1ilPrcT/l/Mjhk4AACAgRDgAAAABkKAAwAAGAgBDgAAYCAEOAAAgIEQ4AAAAAZCgAMAABgIAQ4AAGAgBDgAAICBEOAAAAAGQoADAAAYiCsWXQAs0r6jD2za/uQ9t8y5kp1pq8c38RhvZ5Uet1X6Wye13WMCiecNrLqJZuCq6mBVPV5VZ6vq6Cbrf6yqHq2qT1bVR6rqNWPrnq+qR0Y/J2ZZPAAAwCq54AxcVe1Kcm+SNydZS3Kqqk5096Njm/1+kgPd/eWq+rtJ3p3kb4zW/Vl33zDjugEAAFbOJDNwNyU5291PdPdzSe5Lcmh8g+7+WHd/ebT4UJJrZlsmAAAAkwS4q5M8Nba8NmrbyjuS/ObY8tdX1emqeqiqvu8SagQAACCTncSkNmnrTTes+qEkB5J891jz3u4+V1XfmuSjVfWp7v7DDdc7kuRIkuzdu3eiwgEAAFbNJDNwa0muHVu+Jsm5jRtV1ZuS/ESSW7v72Rfau/vc6PcTST6e5MaN1+3uY919oLsP7Nmz56L+AAAAgFUxSYA7lWR/VV1XVVcmOZzkRWeTrKobk7wv6+Ht6bH2q6rq5aPLu5O8Icn4yU8AAACY0AV3oezu81V1Z5IHk+xKcry7z1TV3UlOd/eJJD+T5BVJfq2qkuQz3X1rktcmeV9VfSXrYfGeDWevBAAAYEITfZF3d59McnJD211jl9+0xfV+J8l3TFMgAAAA6yb6Im8AAAAWb6IZOFbLvqMPLLoELgP/V4/BpfK4cbG26zNP3nPLHCu5/Jb5+bHMtc3aTuhzy/b/Gq9nXo/hsj0GGy3iMdmMGTgAAICBEOAAAAAGQoADAAAYCAEOAABgIAQ4AACAgRDgAAAABkKAAwAAGAgBDgAAYCAEOAAAgIEQ4AAAAAbiikUXADBL+44+sOgSBstjB8Bmtnt9ePKeW2Z6e0Ox1d9wKY/HxTIDBwAAMBACHAAAwEAIcAAAAAMhwAEAAAyEAAcAADAQAhwAAMBATBTgqupgVT1eVWer6ugm619eVb86Wv+Jqto3tu6do/bHq+qtsysdAABgtVwwwFXVriT3Jrk5yfVJbquq6zds9o4kX+zuv5TkPUl+enTd65McTvJtSQ4m+fnR7QEAAHCRJpmBuynJ2e5+orufS3JfkkMbtjmU5AOjyx9K8r1VVaP2+7r72e7+oyRnR7cHAADARZokwF2d5Kmx5bVR26bbdPf5JF9K8hcmvC4AAAATuGKCbWqTtp5wm0mum6o6kuTIaPHZqvr0BHXBZVM//dWLu5N8YXGV7FxjjzGT2Z3kC6v0uK3S3zpAlzw2+r9eXiv4+F6wL67gYzJXHt8X2V0/PbP3ja/ZasUkAW4tybVjy9ckObfFNmtVdUWSb0zyzITXTXcfS3IsSarqdHcfmKAuuOz0R5aFvsgy0R9ZFvoiy2Re/XGSXShPJdlfVddV1ZVZPynJiQ3bnEhy++jy9yf5aHf3qP3w6CyV1yXZn+R3Z1M6AADAarngDFx3n6+qO5M8mGRXkuPdfaaq7k5yurtPJPnFJL9cVWezPvN2eHTdM1V1f5JHk5xPckd3P3+Z/hYAAIAdbZJdKNPdJ5Oc3NB219jlP0/yA1tc958k+ScXUdOxi9gWLjf9kWWhL7JM9EeWhb7IMplLf6z1PR0BAABYdpMcAwcAAMASWKoAV1UHq+rxqjpbVUcXXQ+rpaqerKpPVdUjVXV61PaqqvpwVf3B6PdVi66TnamqjlfV0+Nfo7JV/6t1PzcaKz9ZVa9bXOXsRFv0x5+sqs+OxshHquptY+veOeqPj1fVWxdTNTtRVV1bVR+rqseq6kxV/f1Ru/GRudqmL859bFyaAFdVu5Lcm+TmJNcnua2qrl9sVayg7+nuG8ZOAXs0yUe6e3+Sj4yW4XJ4f5KDG9q26n83Z/2svvuz/h2a751TjayO9+el/TFJ3jMaI28YHR+f0Wv14STfNrrOz49e02EWzif58e5+bZLXJ7lj1OeMj8zbVn0xmfPYuDQBLslNSc529xPd/VyS+5IcWnBNcCjJB0aXP5Dk+xZYCztYd/921s/iO26r/ncoyS/1uoeSfFNVvXo+lbIKtuiPWzmU5L7ufra7/yjJ2ay/psPUuvtz3f17o8v/LcljSa6O8ZE526YvbuWyjY3LFOCuTvLU2PJatn9QYNY6yW9V1cNVdWTU9i3d/blk/Ymb5JsXVh2raKv+Z7xkUe4c7ZZ2fGyXcv2RuaiqfUluTPKJGB9ZoA19MZnz2LhMAa42aXOKTObpDd39uqzvfnFHVX3XoguCLRgvWYT3JvmLSW5I8rkkPztq1x+57KrqFUn+TZJ/0N3/dbtNN2nTH5mZTfri3MfGZQpwa0muHVu+Jsm5BdXCCuruc6PfTyf59axPc//JC7tejH4/vbgKWUFb9T/jJXPX3X/S3c9391eS/EK+tiuQ/shlVVVfl/U3zP+6u//tqNn4yNxt1hcXMTYuU4A7lWR/VV1XVVdm/aC/EwuuiRVRVd9QVa984XKStyT5dNb74O2jzW5P8huLqZAVtVX/O5Hkh0dnW3t9ki+9sCsRXC4bjiP6a1kfI5P1/ni4ql5eVddl/eQRvzvv+tiZqqqS/GKSx7r7n42tMj4yV1v1xUWMjVfM4kZmobvPV9WdSR5MsivJ8e4+s+CyWB3fkuTX15+buSLJr3T3f6yqU0nur6p3JPlMkh9YYI3sYFX1wSRvTLK7qtaSvCvJPdm8/51M8rasHxD95SQ/MveC2dG26I9vrKobsr4L0JNJ/k6SdPeZqro/yaNZP0vbHd39/CLqZkd6Q5K/meRTVfXIqO0fxfjI/G3VF2+b99hY3XYLBgAAGIJl2oUSAACAbQhwAAAAAyHAAQAADIQABwAAMBACHAAAwEAIcAAAAAMhwAEAAAyEAAcAADAQVyy6gI12797d+/btW3QZAAAAC/Hwww9/obv3bLZu6QLcvn37cvr06UWXAQAAsBBV9cdbrbMLJQAAwEAIcAAAAAMhwAEAAAyEAAcAADAQS3cSEwAAltu+ow9Mdf0n77llRpXA6jEDBwAAMBACHAAAwEAIcAAAAAMhwAEAAAyEAAcAADAQzkLJUnJ2KwAAeCkzcAAAAAMhwAEAAAyEAAcAADAQAhwAAMBAOIkJMzftCUgAuDycIApg+MzAAQAADMRUAa6qDlbV41V1tqqObrJ+b1V9rKp+v6o+WVVvm+b+AAAAVtklB7iq2pXk3iQ3J7k+yW1Vdf2Gzf5xkvu7+8Ykh5P8/KXeHwAAwKqbZgbupiRnu/uJ7n4uyX1JDm3YppP896PL35jk3BT3BwAAsNKmOYnJ1UmeGlteS/KdG7b5ySS/VVV/L8k3JHnTFPcHAACw0qYJcLVJW29Yvi3J+7v7Z6vqf0zyy1X17d39lRfdUNWRJEeSZO/evVOUBDBbztoHACyTaXahXEty7djyNXnpLpLvSHJ/knT3f0ry9Ul2b7yh7j7W3Qe6+8CePXumKAkAAGDnmibAnUqyv6quq6ors36SkhMbtvlMku9Nkqp6bdYD3OenuE8AAICVdckBrrvPJ7kzyYNJHsv62SbPVNXdVXXraLMfT/K3q+r/TfLBJG/v7o27WQIAADCBaY6BS3efTHJyQ9tdY5cfTfKGae4DAACAdVN9kTcAAADzM9UMHAAwP9OeFRWWxSz6srP8sqrMwAEAAAyEAAcAADAQAhwAAMBACHAAAAADIcABAAAMhLNQAgAAgzXtWU2HdkZTM3AAAAADIcABAAAMhAAHAAAwEAIcAADAQAhwAAAAA+EslACX0bRnxkqGd3YsYPnNYmwCFsMMHAAAwEBMFeCq6mBVPV5VZ6vq6Bbb/C9V9WhVnamqX5nm/gAAAFbZJe9CWVW7ktyb5M1J1pKcqqoT3f3o2Db7k7wzyRu6+4tV9c3TFgwAALCqppmBuynJ2e5+orufS3JfkkMbtvnbSe7t7i8mSXc/PcX9AQAArLRpAtzVSZ4aW14btY37y0n+clX9P1X1UFUdnOL+AAAAVto0Z6GsTdp6k9vfn+SNSa5J8n9X1bd393950Q1VHUlyJEn27t07RUkAALAapj2bqLMcD9M0M3BrSa4dW74myblNtvmN7v7/uvuPkjye9UD3It19rLsPdPeBPXv2TFESAADAzjVNgDuVZH9VXVdVVyY5nOTEhm3+XZLvSZKq2p31XSqfmOI+AQAAVtYlB7juPp/kziQPJnksyf3dfaaq7q6qW0ebPZjkT6vq0SQfS/J/dPefTls0AADAKprmGLh098kkJze03TV2uZP82OgHAACAKUz1Rd4AAADMz1QzcLCsnJVpNjyOALNnbAWmYQYOAABgIAQ4AACAgRDgAAAABkKAAwAAGAgBDgAAYCAEOAAAgIEQ4AAAAAZCgAMAABgIAQ4AAGAgBDgAAICBEOAAAAAG4opFF8Dy2Xf0gUWXADOjP8PszOL59OQ9t8ygktVmXGMnWYb+PG0N8x7XzMABAAAMxFQBrqoOVtXjVXW2qo5us933V1VX1YFp7g8AAGCVXXKAq6pdSe5NcnOS65PcVlXXb7LdK5P8aJJPXOp9AQAAMN0M3E1Jznb3E939XJL7khzaZLufSvLuJH8+xX0BAACsvGkC3NVJnhpbXhu1fVVV3Zjk2u7+D1PcDwAAAJnuLJS1SVt/dWXVy5K8J8nbL3hDVUeSHEmSvXv3TlESALPmzIMAsDymmYFbS3Lt2PI1Sc6NLb8yybcn+XhVPZnk9UlObHYik+4+1t0HuvvAnj17pigJAABg55omwJ1Ksr+qrquqK5McTnLihZXd/aXu3t3d+7p7X5KHktza3aenqhgAAGBFXXKA6+7zSe5M8mCSx5Lc391nquruqrp1VgUCAACwbppj4NLdJ5Oc3NB21xbbvnGa+wIAAFh1U32RNwAAAPMz1QwcsLxmceZAmJVp+6OzWLJMjK/MyqL7krMMD5MZOAAAgIEQ4AAAAAZCgAMAABgIAQ4AAGAgBDgAAICBEOAAAAAGQoADAAAYCAEOAABgIAQ4AACAgRDgAAAABkKAAwAAGIgrFl0AsHPtO/rAVNd/8p5bZlQJLN60zwfgxbzGsKqmmoGrqoNV9XhVna2qo5us/7GqerSqPllVH6mq10xzfwAAAKvskgNcVe1Kcm+Sm5Ncn+S2qrp+w2a/n+RAd//VJB9K8u5LvT8AAIBVN80M3E1Jznb3E939XJL7khwa36C7P9bdXx4tPpTkminuDwAAYKVNE+CuTvLU2PLaqG0r70jym1PcHwAAwEqb5iQmtUlbb7ph1Q8lOZDku7dYfyTJkSTZu3fvFCUB7DwO1AeA5TWLk1RdzGv1NDNwa0muHVu+Jsm5jRtV1ZuS/ESSW7v72c1uqLuPdfeB7j6wZ8+eKUoCAADYuaYJcKeS7K+q66rqyiSHk5wY36CqbkzyvqyHt6enuC8AAICVd8kBrrvPJ7kzyYNJHktyf3efqaq7q+rW0WY/k+QVSX6tqh6pqhNb3BwAAAAXMNUXeXf3ySQnN7TdNXb5TdPcPgAAAF8z1Rd5AwAAMD9TzcDBTjXvswnB5TSL/gwAm/EaM39m4AAAAAZCgAMAABgIAQ4AAGAgBDgAAICBEOAAAAAGwlkodyBnA2Kn0Jd5gTPD7hzT/i/9H4FVZwYOAABgIAQ4AACAgRDgAAAABkKAAwAAGAgBDgAAYCCchRKWlDMwAsyesZVZ0ZdYFDNwAAAAAzFVgKuqg1X1eFWdraqjm6x/eVX96mj9J6pq3zT3BwAAsMouOcBV1a4k9ya5Ocn1SW6rqus3bPaOJF/s7r+U5D1JfvpS7w8AAGDVTTMDd1OSs939RHc/l+S+JIc2bHMoyQdGlz+U5Hurqqa4TwAAgJU1TYC7OslTY8tro7ZNt+nu80m+lOQvTHGfAAAAK2uas1BuNpPWl7BNqupIkiOjxWer6tNT1AWztDvJFy7limWHYWbrkvsi6zwnZ2ph/dH/kQ28TrNMZtkfX7PVttMEuLUk144tX5Pk3BbbrFXVFUm+MckzG2+ou48lOZYkVXW6uw9MURfMjP7IstAXWSb6I8tCX2SZzKs/TrML5akk+6vquqq6MsnhJCc2bHMiye2jy9+f5KPd/ZIZOAAAAC7skmfguvt8Vd2Z5MEku5Ic7+4zVXV3ktPdfSLJLyb55ao6m/WZt8OzKBoAAGAVTbMLZbr7ZJKTG9ruGrv850l+4CJv9tg0NcGM6Y8sC32RZaI/siz0RZbJXPpj2aMRAABgGKY5Bg4AAIA5WqoAV1UHq+rxqjpbVUcXXQ+rpaqerKpPVdUjVXV61PaqqvpwVf3B6PdVi66TnamqjlfV0+Nfo7JV/6t1PzcaKz9ZVa9bXOXsRFv0x5+sqs+OxshHquptY+veOeqPj1fVWxdTNTtRVV1bVR+rqseq6kxV/f1Ru/GRudqmL859bFyaAFdVu5Lcm+TmJNcnua2qrl9sVayg7+nuG8ZOAXs0yUe6e3+Sj4yW4XJ4f5KDG9q26n83J9k/+jmS5L1zqpHV8f68tD8myXtGY+QNo+PgM3qtPpzk20bX+fnRazrMwvkkP97dr03y+iR3jPqc8ZF526ovJnMeG5cmwCW5KcnZ7n6iu59Lcl+SQwuuCQ4l+cDo8geSfN8Ca2EH6+7fzku/J3Or/ncoyS/1uoeSfFNVvXo+lbIKtuiPWzmU5L7ufra7/yjJ2ay/psPUuvtz3f17o8v/LcljSa6O8ZE526YvbuWyjY3LFOCuTvLU2PJatn9QYNY6yW9V1cNVdWTU9i3d/blk/Ymb5JsXVh2raKv+Z7xkUe4c7ZZ2fGyXcv2RuaiqfUluTPKJGB9ZoA19MZnz2LhMAa42aXOKTObpDd39uqzvfnFHVX3XoguCLRgvWYT3JvmLSW5I8rkkPztq1x+57KrqFUn+TZJ/0N3/dbtNN2nTH5mZTfri3MfGZQpwa0muHVu+Jsm5BdXCCuruc6PfTyf59axPc//JC7tejH4/vbgKWUFb9T/jJXPX3X/S3c9391eS/EK+tiuQ/shlVVVfl/U3zP+6u//tqNn4yNxt1hcXMTYuU4A7lWR/VV1XVVdm/aC/EwuuiRVRVd9QVa984XKStyT5dNb74O2jzW5P8huLqZAVtVX/O5Hkh0dnW3t9ki+9sCsRXC4bjiP6a1kfI5P1/ni4ql5eVddl/eQRvzvv+tiZqqqS/GKSx7r7n42tMj4yV1v1xUWMjVfM4kZmobvPV9WdSR5MsivJ8e4+s+CyWB3fkuTX15+buSLJr3T3f6yqU0nur6p3JPlMkh9YYI3sYFX1wSRvTLK7qtaSvCvJPdm8/51M8rasHxD95SQ/MveC2dG26I9vrKobsr4L0JNJ/k6SdPeZqro/yaNZP0vbHd39/CLqZkd6Q5K/meRTVfXIqO0fxfjI/G3VF2+b99hY3XYLBgAAGIJl2oUSAACAbQhwAAAAAyHAAQAADIQABwAAMBACHAAAwEAIcAAAAAMhwAEAAAyEAAcAADAQVyy6gI12797d+/btW3QZAAAAC/Hwww9/obv3bLZu6QLcvn37cvr06UWXAQAAsBBV9cdbrbMLJQAAwEAIcAAAAAMhwAEAAAzE0h0Dx8627+gDiy7hRZ6855ZFlwAAU/HaCqvFDBwAAMBACHAAAAADIcABAAAMhAAHAAAwEAIcAADAQAhwAAAAAyHAAQAADIQABwAAMBACHAAAwEAIcAAAAAMhwAEAAAyEAAcAADAQAhwAAMBACHAAAAADIcABAAAMxBWLLgAAgJ1j39EHvnr5yXtuWWAlsDMJcMCONv5GYhl4MwMATMMulAAAAAMhwAEAAAyEAAcAADAQjoEDABiQZTu2dzvLVqvjkNkJzMABAAAMhAAHAAAwEAIcAADAQAhwAAAAAzFVgKuqg1X1eFWdraqj22z3/VXVVXVgmvsDAABYZZd8Fsqq2pXk3iRvTrKW5FRVnejuRzds98okP5rkE9MUynAt2xmoxo3X5sxUzIM+BwBMY5oZuJuSnO3uJ7r7uST3JTm0yXY/leTdSf58ivsCAABYedMEuKuTPDW2vDZq+6qqujHJtd39H6a4HwAAADLdF3nXJm391ZVVL0vyniRvv+ANVR1JciRJ9u7dO0VJAMBWlm2XdrsRA1y8aWbg1pJcO7Z8TZJzY8uvTPLtST5eVU8meX2SE5udyKS7j3X3ge4+sGfPnilKAgAA2LmmmYE7lWR/VV2X5LNJDif5wRdWdveXkux+YbmqPp7kf+/u01PcJ8C2lm2GAYDl4URS7ASXPAPX3eeT3JnkwSSPJbm/u89U1d1VdeusCgQAAGDdNDNw6e6TSU5uaLtri23fOM19MTkzEDAMy/Zc9Wk0bG7ZnquwaMv8nOv2rwIAABoSSURBVFiF17KpvsgbAACA+RHgAAAABkKAAwAAGIipjoEDWOb94AEulbENhmmZn7uzOj7PDBwAAMBACHAAAAADYRdKAGAhfKkyDMMy75a4igQ4AJihZXujIxjB8CzbOMJysQslAADAQJiBA4AdzCf5MAyeq0zKDBwAAMBAmIGDkWX75MtxK8ybE0oAwPIzAwcAADAQAhwAAMBACHAAAAAD4Rg4AABWzrId+w6TMgMHAAAwEGbgYEkt8xkBfWoJX+P5AMA8mYEDAAAYCAEOAABgIOxCCQAsnF1RASZjBg4AAGAgBDgAAICBEOAAAAAGwjFwALzEsh2PtGxfpQEAiyLA7QDL9kYLAAC4POxCCQAAMBACHAAAwEDYhRKApTe+q7jj4QBYZWbgAAAABkKAAwAAGAgBDgAAYCCmCnBVdbCqHq+qs1V1dJP1P1ZVj1bVJ6vqI1X1mmnuDwAAYJVdcoCrql1J7k1yc5Lrk9xWVddv2Oz3kxzo7r+a5ENJ3n2p9wcAALDqppmBuynJ2e5+orufS3JfkkPjG3T3x7r7y6PFh5JcM8X9AQAArLRpAtzVSZ4aW14btW3lHUl+c4r7AwAAWGnTfA9cbdLWm25Y9UNJDiT57i3WH0lyJEn27t07RUmwM41/BxasOs8HAFbZNDNwa0muHVu+Jsm5jRtV1ZuS/ESSW7v72c1uqLuPdfeB7j6wZ8+eKUoCAADYuaYJcKeS7K+q66rqyiSHk5wY36CqbkzyvqyHt6enuC8AAICVd8kBrrvPJ7kzyYNJHktyf3efqaq7q+rW0WY/k+QVSX6tqh6pqhNb3BwAAAAXMM0xcOnuk0lObmi7a+zym6a5fQAAAL5mqi/yBgAAYH4EOAAAgIEQ4AAAAAZCgAMAABgIAQ4AAGAgBDgAAICBEOAAAAAGQoADAAAYCAEOAABgIAQ4AACAgRDgAAAABkKAAwAAGAgBDgAAYCAEOAAAgIEQ4AAAAAZCgAMAABgIAQ4AAGAgBDgAAICBEOAAAAAGQoADAAAYiCsWXcBQ7Dv6wKJLAAAAVpwZOAAAgIEQ4AAAAAZCgAMAABgIAQ4AAGAgBDgAAICBEOAAAAAGQoADAAAYCAEOAABgIAQ4AACAgRDgAAAABkKAAwAAGAgBDgAAYCCmCnBVdbCqHq+qs1V1dJP1L6+qXx2t/0RV7Zvm/gAAAFbZJQe4qtqV5N4kNye5PsltVXX9hs3ekeSL3f2XkrwnyU9f6v0BAACsumlm4G5Kcra7n+ju55Lcl+TQhm0OJfnA6PKHknxvVdUU9wkAALCypglwVyd5amx5bdS26TbdfT7Jl5L8hSnuEwAAYGVdMcV1N5tJ60vYJlV1JMmR0eKzVfXpKeqCWdqd5AuLLgKiL7Jc9EeWhb7IMtm2P9bFHUz2mq1WTBPg1pJcO7Z8TZJzW2yzVlVXJPnGJM9svKHuPpbkWJJU1enuPjBFXTAz+iPLQl9kmeiPLAt9kWUyr/44zS6Up5Lsr6rrqurKJIeTnNiwzYkkt48uf3+Sj3b3S2bgAAAAuLBLnoHr7vNVdWeSB5PsSnK8u89U1d1JTnf3iSS/mOSXq+ps1mfeDs+iaAAAgFU0zS6U6e6TSU5uaLtr7PKfJ/mBi7zZY9PUBDOmP7Is9EWWif7IstAXWSZz6Y9lj0YAAIBhmOYYOAAAAOZoqQJcVR2sqser6mxVHV10PayWqnqyqj5VVY9U1elR26uq6sNV9Qej31ctuk52pqo6XlVPj3+Nylb9r9b93Gis/GRVvW5xlbMTbdEff7KqPjsaIx+pqreNrXvnqD8+XlVvXUzV7ERVdW1VfayqHquqM1X190ftxkfmapu+OPexcWkCXFXtSnJvkpuTXJ/ktqq6frFVsYK+p7tvGDsF7NEkH+nu/Uk+MlqGy+H9SQ5uaNuq/92cZP/o50iS986pRlbH+/PS/pgk7xmNkTeMjoPP6LX6cJJvG13n50ev6TAL55P8eHe/Nsnrk9wx6nPGR+Ztq76YzHlsXJoAl+SmJGe7+4nufi7JfUkOLbgmOJTkA6PLH0jyfQushR2su387L/2ezK3636Ekv9TrHkryTVX16vlUyirYoj9u5VCS+7r72e7+oyRns/6aDlPr7s919++NLv+3JI8luTrGR+Zsm764lcs2Ni5TgLs6yVNjy2vZ/kGBWeskv1VVD1fVkVHbt3T355L1J26Sb15Ydayirfqf8ZJFuXO0W9rxsV3K9Ufmoqr2JbkxySdifGSBNvTFZM5j4zIFuNqkzSkymac3dPfrsr77xR1V9V2LLgi2YLxkEd6b5C8muSHJ55L87Khdf+Syq6pXJPk3Sf5Bd//X7TbdpE1/ZGY26YtzHxuXKcCtJbl2bPmaJOcWVAsrqLvPjX4/neTXsz7N/Scv7Hox+v304ipkBW3V/4yXzF13/0l3P9/dX0nyC/narkD6I5dVVX1d1t8w/+vu/rejZuMjc7dZX1zE2LhMAe5Ukv1VdV1VXZn1g/5OLLgmVkRVfUNVvfKFy0nekuTTWe+Dt482uz3JbyymQlbUVv3vRJIfHp1t7fVJvvTCrkRwuWw4juivZX2MTNb74+GqenlVXZf1k0f87rzrY2eqqkryi0ke6+5/NrbK+MhcbdUXFzE2XjGLG5mF7j5fVXcmeTDJriTHu/vMgstidXxLkl9ff27miiS/0t3/sapOJbm/qt6R5DNJfmCBNbKDVdUHk7wxye6qWkvyriT3ZPP+dzLJ27J+QPSXk/zI3AtmR9uiP76xqm7I+i5ATyb5O0nS3Weq6v4kj2b9LG13dPfzi6ibHekNSf5mkk9V1SOjtn8U4yPzt1VfvG3eY2N12y0YAABgCJZpF0oAAAC2IcABAAAMhAAHAAAwEAIcAADAQAhwAAAAAyHAAQAADIQABwAAMBACHAAAwEBccaENqup4kv85ydPd/e2brK8k/zzr33r/5SRv7+7fG627Pck/Hm36f3b3By50f7t37+59+/ZN/AcAAADsJA8//PAXunvPZusuGOCSvD/Jv0jyS1usvznJ/tHPdyZ5b5LvrKpXJXlXkgNJOsnDVXWiu7+43Z3t27cvp0+fnqAsAACAnaeq/nirdRfchbK7fzvJM9tscijJL/W6h5J8U1W9Oslbk3y4u58ZhbYPJzl4caUDAADwglkcA3d1kqfGltdGbVu1AwAAcAlmEeBqk7bepv2lN1B1pKpOV9Xpz3/+8zMoCQAAYOeZ5Bi4C1lLcu3Y8jVJzo3a37ih/eOb3UB3H0tyLEkOHDiwacgbon1HH/jq5SfvuWWBlexM449v4jFmsTzfGTLjKcBwzCLAnUhyZ1Xdl/WTmHypuz9XVQ8m+adVddVou7ckeecM7m+pbXwRZOfZ6n/sDc/O5/nNqvCBBNPwOrkcPI93rkm+RuCDWZ9J211Va1k/s+TXJUl3/8skJ7P+FQJns/41Aj8yWvdMVf1UklOjm7q7u7c7GQpc0HZvoBf9guETbF6wSi+a2z0nF/23L3Nti3IpH0KsUn9mcvrS8pn0PZLHfvguGOC6+7YLrO8kd2yx7niS45dW2s7izT0M07Szbqv83PeGYedZ5f7M6u6FsGxj2bLVs8x26gd5s9iFcuWt6oAGMClvOBbHaxTLaNk/DBjK82YodSZeB2ZJgGPHM2DAbA3pDQPALHlPwTIQ4AAAYEUIoeuG/DgIcABzNOQXjJ3E/2E2PI7AslmFvUQEOJjAKgwGsBN4rgJMblHHIvrwZzoCHMAGQgDAfHlDvxxm+X/wWnr5CHAA8UIDAAyDAMfSG+oba58mciH6CABwsQQ4AABYYkP9MJvLQ4ADAKZmRhkYqqGNXwIcAABLY2hvpmHeBLhLYBr74hiIAQC4XFbtvbkABwCwwnzQCsMiwAEwV94swjCs2qwGDIUAx0rZ+GLkzSMAq0g4g+ES4Jgrn7wDADCNVf8AQoADAGAp+eAXXkqAY+ZW/VMRAAC4XAQ4AGDhzLRcXj5c5WJdynNSP5sPAQ4AAFhqwuHXTBTgqupgkn+eZFeSf9Xd92xY/54k3zNa/O+SfHN3f9No3fNJPjVa95nuvnUWhQMAwKI4szWLcsEAV1W7ktyb5M1J1pKcqqoT3f3oC9t09z8c2/7vJblx7Cb+rLtvmF3JO4NdRYBZM64A0zLLActvkhm4m5Kc7e4nkqSq7ktyKMmjW2x/W5J3zaY8ABbNGzpgGZjxgnUvm2Cbq5M8Nba8Nmp7iap6TZLrknx0rPnrq+p0VT1UVd93yZUCAACsuElm4GqTtt5i28NJPtTdz4+17e3uc1X1rUk+WlWf6u4/fNEdVB1JciRJ9u7dO0FJAMBOZaYFYGuTBLi1JNeOLV+T5NwW2x5Ocsd4Q3efG/1+oqo+nvXj4/5wwzbHkhxLkgMHDmwVDgGAHcTuuQAXb5IAdyrJ/qq6Lslnsx7SfnDjRlX1V5JcleQ/jbVdleTL3f1sVe1O8oYk755F4SwXL8IAAHD5XTDAdff5qrozyYNZ/xqB4919pqruTnK6u0+MNr0tyX3dPT6D9tok76uqr2T9eLt7xs9eCVsRCGE6zkgJADvTRN8D190nk5zc0HbXhuWf3OR6v5PkO6aoDwAAgJGJAhxsxiwZAADMlwDHwtjFCwAALs4k3wMHAADAEhDgAAAABkKAAwAAGAgBDgAAYCAEOAAAgIFwFkoAAFaar0ZiSAQ4AF5ild7M+EoTAIbELpQAAAADYQaOpbBKn/YDAMClMgMHAAAwEAIcAADAQNiFEgBYak40A4u18VAXz8PFEuAABsqxo8B2jBGwMwlwwEryxgYAGCLHwAEAAAyEGTgAAGAhHON68QQ4ABbGgfE7kzdkAJfPRLtQVtXBqnq8qs5W1dFN1r+9qj5fVY+Mfv7W2Lrbq+oPRj+3z7J4AACAVXLBGbiq2pXk3iRvTrKW5FRVnejuRzds+qvdfeeG674qybuSHEjSSR4eXfeLM6keAABghUyyC+VNSc529xNJUlX3JTmUZGOA28xbk3y4u58ZXffDSQ4m+eCllQuwM9nlDICh8Jq1WJMEuKuTPDW2vJbkOzfZ7q9X1Xcl+c9J/mF3P7XFda++xFoBAIAV4St/NjfJMXC1SVtvWP73SfZ1919N8n8l+cBFXDdVdaSqTlfV6c9//vMTlAQAALB6JpmBW0ty7djyNUnOjW/Q3X86tvgLSX567Lpv3HDdj2+8g+4+luRYkhw4cOAlAQ8Alondh4Zru0/0x/+X/sfAspokwJ1Ksr+qrkvy2SSHk/zg+AZV9eru/txo8dYkj40uP5jkn1bVVaPltyR559RVAwDMmN21gCG4YIDr7vNVdWfWw9iuJMe7+0xV3Z3kdHefSPKjVXVrkvNJnkny9tF1n6mqn8p6CEySu184oQkAAAAXZ6Iv8u7uk0lObmi7a+zyO7PFzFp3H09yfIoaYfDsigOwMxnfgXmbKMABADAbQh8wDQEOAKaw8bgpb8gBuJwEOCbm4O6dySfBAAyR1y9W1coEOE9yAC7EbBqT8IEmsEgrE+AAAGCZmGDgUrxs0QUAAAAwGTNwJPEJEAAADIEAN1ACFwBszXFqwE61YwOcgRteTOgHuDTeUwDLZMcGOGBr3owALAdnPgUu1o4KcEN9U2pmBAAAmMTgA9xQQxvAsjGevpTHZPn40BNYdYMPcKvEGwm4MLsjAQA7mQC3ooRBAJgts4PAPAhwS8bgDwCwM/kAnVkQ4ICvsvshTM8HcYuxE98Y60vAZgQ4AABg4XwQMxkBDrhoPhWG+djuzYznHsBqWskA5wURZmfZd7sUNgGYh2lfb7xeMamJAlxVHUzyz5PsSvKvuvueDet/LMnfSnI+yeeT/K/d/cejdc8n+dRo0890960zqh2Yo524WwMwbDtxXNrqb/LmfrXsxL7N7FwwwFXVriT3JnlzkrUkp6rqRHc/OrbZ7yc50N1frqq/m+TdSf7GaN2fdfcNsyxapwYAAJbNPHLKJDNwNyU5291PJElV3ZfkUJKvBrju/tjY9g8l+aFZFrmqJu0Ay74LGwAArLpZzaRPEuCuTvLU2PJaku/cZvt3JPnNseWvr6rTWd+98p7u/ncXXSXANszKb8+uV7CzGQNhtUwS4GqTtt50w6ofSnIgyXePNe/t7nNV9a1JPlpVn+ruP9xwvSNJjiTJ3r17JyocAABg0eb9IcokAW4tybVjy9ckObdxo6p6U5KfSPLd3f3sC+3dfW70+4mq+niSG5O8KMB197Ekx5LkwIEDm4ZDFsOneqvN/3/n8z9ePv4nAGxnkgB3Ksn+qrouyWeTHE7yg+MbVNWNSd6X5GB3Pz3WflWSL3f3s1W1O8kbsn6CEwDY8S7n7qt2jQVYTRcMcN19vqruTPJg1r9G4Hh3n6mqu5Oc7u4TSX4mySuS/FpVJV/7uoDXJnlfVX0lycuyfgzco5veETBTi/oU35tK5s2MFQCzMun7mEW+9kz0PXDdfTLJyQ1td41dftMW1/udJN8xTYHMnzdDALPnww0AZmGiAAc7lTdUAFwOvuIHuFwEOADgshFkWHX2bGLWBDhYIG9smDdvJACYJa8r8yfAwZwZ6AAAuFQCHAAsmA92AJjUIAKcFzYAAGCnmOYwmkEEOC6OMysCrBbjPsDqEOAAAGCG7D3G5STAAQBz440twHQEOOCymfVuXXYTA2BWfJjAUAlwADBn3jgCcKkEOACAy0xoB2ZFgAMAGBBhEBZjWZ57Ahwsqe0GiVU5/mtZBkoAgGUhwG3gJAks0rSBReDZ+Xb6/9gYDADbe9n/3979hch1lnEc//5IbC5U/FcsmgYNGsH0JlaphYLUCzUtwipYur3QIIF4kaCCN60XWoRCc6GiWAvRhraihqAW9yIYtQreWJsowSQNxaWNdpvQoJUoiJHEx4tzgsO6s5nE3TlnZ76fmznnnTObZ5dnnjlPznve6ToASZIkSdJovAI3okn/X2+pS9dy1eVa3pO+jzUNzHNJmmw2cJLGYvFJ5bBGzZNPSZKk4ZxCKUmSJElrhFfgpDVoEhZ68EqbJElS42rO7UZq4JJsB74GrAO+XVUPLnp+A/A48G7gL8DdVXW6fe4+YCdwCfh0VR2+2l+iK32IQboS83TyjDrdVJIkTZ8rNnBJ1gEPAR8AFoAjSeaq6pmBw3YCf62qtyeZBfYCdyfZCswCNwFvBn6e5B1VdWmlfxEtbRKu1HTBpkjqL9+fkhbzfEfTZJQrcLcA81X1HECSA8AMMNjAzQD3t9s/AL6RJO34gaq6ADyfZL79eb9emfB1NTzpWZ5/H/XVtOamJ2SSVtq01lNdm77myygN3EbghYH9BeC9w46pqotJzgNvaMefWvTajcv9Y8dfPN/bP5YkqRt+LkgalfVCk26UBi5LjNWIx4zyWpLsAna1uxf+uPfDJ0aISxqH64E/dx2EhLmofjEf1RfmovpkRfIxewF4y7DnR2ngFoBNA/s3AmeGHLOQZD3wGuDlEV9LVe0D9gEkOVpV7xkhLmnVmY/qC3NRfWI+qi/MRfXJuPJxlO+BOwJsSbI5yXU0i5LMLTpmDtjRbn8M+EVVVTs+m2RDks3AFuDplQldkiRJkqbLFa/Atfe07QEO03yNwP6qOpnkS8DRqpoDHgG+0y5S8jJNk0d73EGaBU8uArtdgVKSJEmSrs1I3wNXVYeAQ4vGvjCw/U/griGvfQB44Cpi2ncVx0qrzXxUX5iL6hPzUX1hLqpPxpKPaWY6SpIkSZL6bpR74CRJkiRJPdCrBi7J9iTPJplPcm/X8Wi6JDmd5HiSY0mOtmOvT/KzJH9oH1/XdZyaTEn2JzmX5MTA2JL5l8bX21r5+yQ3dxe5JtGQfLw/yYttjTyW5M6B5+5r8/HZJB/qJmpNoiSbkvwyyakkJ5N8ph23PmqslsnFsdfG3jRwSdYBDwF3AFuBe5Js7TYqTaH3V9W2gSVg7wWerKotwJPtvrQaHgW2Lxobln930Kzqu4XmOzQfHlOMmh6P8r/5CPDVtkZua++Pp/2sngVual/zzfYzXVoJF4HPVdU7gVuB3W3OWR81bsNyEcZcG3vTwAG3APNV9VxV/Qs4AMx0HJM0AzzWbj8GfKTDWDTBqupXNKv4DhqWfzPA49V4CnhtkjeNJ1JNgyH5OMwMcKCqLlTV88A8zWe69H+rqrNV9bt2++/AKWAj1keN2TK5OMyq1cY+NXAbgRcG9hdY/o8irbQCfprkt0l2tWM3VNVZaN64wBs7i07TaFj+WS/VlT3ttLT9A1PKzUeNRZK3Au8CfoP1UR1alIsw5trYpwYuS4y5RKbG6baquplm+sXuJO/rOiBpCOuluvAw8DZgG3AW+HI7bj5q1SV5FfBD4LNV9bflDl1izHzUilkiF8deG/vUwC0Amwb2bwTOdBSLplBVnWkfzwFP0Fzmfuny1Iv28Vx3EWoKDcs/66XGrqpeqqpLVfVv4Fv8dyqQ+ahVleQVNCfM362qH7XD1keN3VK52EVt7FMDdwTYkmRzkutobvqb6zgmTYkkr0zy6svbwAeBEzQ5uKM9bAfw424i1JQaln9zwCfa1dZuBc5fnkokrZZF9xF9lKZGQpOPs0k2JNlMs3jE0+OOT5MpSYBHgFNV9ZWBp6yPGqthudhFbVy/Ej9kJVTVxSR7gMPAOmB/VZ3sOCxNjxuAJ5r3JuuB71XVT5IcAQ4m2Qn8Cbirwxg1wZJ8H7gduD7JAvBF4EGWzr9DwJ00N0T/A/jk2APWRBuSj7cn2UYzBeg08CmAqjqZ5CDwDM0qbbur6lIXcWsi3QZ8HDie5Fg79nmsjxq/Ybl4z7hrY6qcFixJkiRJa0GfplBKkiRJkpZhAydJkiRJa4QNnCRJkiStETZwkiRJkrRG2MBJkiRJ0hphAydJkiRJa4QNnCRJkiStETZwkiRJkrRG/AcmGZbaxPjoTwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1080x720 with 4 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# \"HBEGITS\"\n",
    "draw_ss_predicts(1,\"H\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>Unnamed: 0.1</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>chip_id</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>dist_0</th>\n",
       "      <th>dist_1</th>\n",
       "      <th>dist_2</th>\n",
       "      <th>dist_3</th>\n",
       "      <th>...</th>\n",
       "      <th>solv_access_16</th>\n",
       "      <th>solv_access_17</th>\n",
       "      <th>solv_access_18</th>\n",
       "      <th>solv_access_19</th>\n",
       "      <th>solv_access_20</th>\n",
       "      <th>solv_access_21</th>\n",
       "      <th>solv_access_22</th>\n",
       "      <th>solv_access_23</th>\n",
       "      <th>solv_access_24</th>\n",
       "      <th>ss_majority_categorical</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_0</td>\n",
       "      <td>0</td>\n",
       "      <td>25</td>\n",
       "      <td>2.931498</td>\n",
       "      <td>4.466173</td>\n",
       "      <td>4.830971</td>\n",
       "      <td>5.929355</td>\n",
       "      <td>...</td>\n",
       "      <td>0.076923</td>\n",
       "      <td>0.524194</td>\n",
       "      <td>0.080808</td>\n",
       "      <td>0.018293</td>\n",
       "      <td>0.396907</td>\n",
       "      <td>0.644330</td>\n",
       "      <td>0.463710</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_1</td>\n",
       "      <td>12</td>\n",
       "      <td>37</td>\n",
       "      <td>3.148242</td>\n",
       "      <td>4.942146</td>\n",
       "      <td>5.011508</td>\n",
       "      <td>6.109199</td>\n",
       "      <td>...</td>\n",
       "      <td>0.524194</td>\n",
       "      <td>0.080808</td>\n",
       "      <td>0.018293</td>\n",
       "      <td>0.396907</td>\n",
       "      <td>0.644330</td>\n",
       "      <td>0.463710</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_2</td>\n",
       "      <td>24</td>\n",
       "      <td>49</td>\n",
       "      <td>3.400494</td>\n",
       "      <td>6.638870</td>\n",
       "      <td>9.738158</td>\n",
       "      <td>12.659022</td>\n",
       "      <td>...</td>\n",
       "      <td>0.080808</td>\n",
       "      <td>0.018293</td>\n",
       "      <td>0.396907</td>\n",
       "      <td>0.644330</td>\n",
       "      <td>0.463710</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_3</td>\n",
       "      <td>36</td>\n",
       "      <td>61</td>\n",
       "      <td>3.473039</td>\n",
       "      <td>5.952474</td>\n",
       "      <td>7.897916</td>\n",
       "      <td>8.550533</td>\n",
       "      <td>...</td>\n",
       "      <td>0.018293</td>\n",
       "      <td>0.396907</td>\n",
       "      <td>0.644330</td>\n",
       "      <td>0.463710</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_4</td>\n",
       "      <td>48</td>\n",
       "      <td>73</td>\n",
       "      <td>3.076461</td>\n",
       "      <td>6.190106</td>\n",
       "      <td>9.162886</td>\n",
       "      <td>11.840832</td>\n",
       "      <td>...</td>\n",
       "      <td>0.396907</td>\n",
       "      <td>0.644330</td>\n",
       "      <td>0.463710</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_5</td>\n",
       "      <td>60</td>\n",
       "      <td>85</td>\n",
       "      <td>3.465978</td>\n",
       "      <td>6.825651</td>\n",
       "      <td>10.210162</td>\n",
       "      <td>13.602541</td>\n",
       "      <td>...</td>\n",
       "      <td>0.644330</td>\n",
       "      <td>0.463710</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>6</td>\n",
       "      <td>6</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_6</td>\n",
       "      <td>72</td>\n",
       "      <td>97</td>\n",
       "      <td>3.187567</td>\n",
       "      <td>4.704804</td>\n",
       "      <td>4.751395</td>\n",
       "      <td>5.968911</td>\n",
       "      <td>...</td>\n",
       "      <td>0.463710</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.103774</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_7</td>\n",
       "      <td>84</td>\n",
       "      <td>109</td>\n",
       "      <td>3.311531</td>\n",
       "      <td>5.471488</td>\n",
       "      <td>9.184711</td>\n",
       "      <td>12.485128</td>\n",
       "      <td>...</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.103774</td>\n",
       "      <td>0.110294</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>8</td>\n",
       "      <td>8</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_8</td>\n",
       "      <td>96</td>\n",
       "      <td>121</td>\n",
       "      <td>3.268903</td>\n",
       "      <td>4.931504</td>\n",
       "      <td>6.315815</td>\n",
       "      <td>9.363784</td>\n",
       "      <td>...</td>\n",
       "      <td>0.476190</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.103774</td>\n",
       "      <td>0.110294</td>\n",
       "      <td>0.053254</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_9</td>\n",
       "      <td>108</td>\n",
       "      <td>133</td>\n",
       "      <td>3.271093</td>\n",
       "      <td>6.159716</td>\n",
       "      <td>9.532755</td>\n",
       "      <td>13.027534</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024390</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.103774</td>\n",
       "      <td>0.110294</td>\n",
       "      <td>0.053254</td>\n",
       "      <td>0.103659</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>10</td>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_10</td>\n",
       "      <td>120</td>\n",
       "      <td>145</td>\n",
       "      <td>3.066599</td>\n",
       "      <td>4.990942</td>\n",
       "      <td>5.992573</td>\n",
       "      <td>8.496792</td>\n",
       "      <td>...</td>\n",
       "      <td>0.201183</td>\n",
       "      <td>0.262887</td>\n",
       "      <td>0.014085</td>\n",
       "      <td>0.136364</td>\n",
       "      <td>0.103774</td>\n",
       "      <td>0.110294</td>\n",
       "      <td>0.053254</td>\n",
       "      <td>0.103659</td>\n",
       "      <td>0.115385</td>\n",
       "      <td>H</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11 rows × 422 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Unnamed: 0  Unnamed: 0.1 protein_id  chip_id  start  stop    dist_0  \\\n",
       "0            0             0     12asA0   chip_0      0    25  2.931498   \n",
       "1            1             1     12asA0   chip_1     12    37  3.148242   \n",
       "2            2             2     12asA0   chip_2     24    49  3.400494   \n",
       "3            3             3     12asA0   chip_3     36    61  3.473039   \n",
       "4            4             4     12asA0   chip_4     48    73  3.076461   \n",
       "5            5             5     12asA0   chip_5     60    85  3.465978   \n",
       "6            6             6     12asA0   chip_6     72    97  3.187567   \n",
       "7            7             7     12asA0   chip_7     84   109  3.311531   \n",
       "8            8             8     12asA0   chip_8     96   121  3.268903   \n",
       "9            9             9     12asA0   chip_9    108   133  3.271093   \n",
       "10          10            10     12asA0  chip_10    120   145  3.066599   \n",
       "\n",
       "      dist_1     dist_2     dist_3  ...  solv_access_16  solv_access_17  \\\n",
       "0   4.466173   4.830971   5.929355  ...        0.076923        0.524194   \n",
       "1   4.942146   5.011508   6.109199  ...        0.524194        0.080808   \n",
       "2   6.638870   9.738158  12.659022  ...        0.080808        0.018293   \n",
       "3   5.952474   7.897916   8.550533  ...        0.018293        0.396907   \n",
       "4   6.190106   9.162886  11.840832  ...        0.396907        0.644330   \n",
       "5   6.825651  10.210162  13.602541  ...        0.644330        0.463710   \n",
       "6   4.704804   4.751395   5.968911  ...        0.463710        0.121951   \n",
       "7   5.471488   9.184711  12.485128  ...        0.121951        0.476190   \n",
       "8   4.931504   6.315815   9.363784  ...        0.476190        0.024390   \n",
       "9   6.159716   9.532755  13.027534  ...        0.024390        0.201183   \n",
       "10  4.990942   5.992573   8.496792  ...        0.201183        0.262887   \n",
       "\n",
       "    solv_access_18  solv_access_19  solv_access_20  solv_access_21  \\\n",
       "0         0.080808        0.018293        0.396907        0.644330   \n",
       "1         0.018293        0.396907        0.644330        0.463710   \n",
       "2         0.396907        0.644330        0.463710        0.121951   \n",
       "3         0.644330        0.463710        0.121951        0.476190   \n",
       "4         0.463710        0.121951        0.476190        0.024390   \n",
       "5         0.121951        0.476190        0.024390        0.201183   \n",
       "6         0.476190        0.024390        0.201183        0.262887   \n",
       "7         0.024390        0.201183        0.262887        0.014085   \n",
       "8         0.201183        0.262887        0.014085        0.136364   \n",
       "9         0.262887        0.014085        0.136364        0.103774   \n",
       "10        0.014085        0.136364        0.103774        0.110294   \n",
       "\n",
       "    solv_access_22  solv_access_23  solv_access_24  ss_majority_categorical  \n",
       "0         0.463710        0.121951        0.476190                        H  \n",
       "1         0.121951        0.476190        0.024390                     None  \n",
       "2         0.476190        0.024390        0.201183                     None  \n",
       "3         0.024390        0.201183        0.262887                     None  \n",
       "4         0.201183        0.262887        0.014085                     None  \n",
       "5         0.262887        0.014085        0.136364                     None  \n",
       "6         0.014085        0.136364        0.103774                     None  \n",
       "7         0.136364        0.103774        0.110294                     None  \n",
       "8         0.103774        0.110294        0.053254                     None  \n",
       "9         0.110294        0.053254        0.103659                     None  \n",
       "10        0.053254        0.103659        0.115385                        H  \n",
       "\n",
       "[11 rows x 422 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_25.loc[0:10]\n",
    "\n",
    "\n",
    "# def ramachandran_plot(pdb_id, degrees=True):\n",
    "#     df = get_torsion_angles(pdb_id, degrees)\n",
    "#     x = df['Phi']\n",
    "#     y = df['Psi']\n",
    "#     # Generate plot\n",
    "#     plt.plot(x, y, \".\")\n",
    "#     plt.title('Ramachandran Plot')\n",
    "#     if degrees:\n",
    "#         plt.xlabel(f'$\\Phi$ Angle (Degrees)')\n",
    "#         plt.xlim(-180, 180)\n",
    "#         plt.ylabel(f'$\\Psi$ Angle (Radians)')\n",
    "#         plt.ylim(-180, 180)\n",
    "#         plt.show()\n",
    "#     else:\n",
    "#         plt.xlabel(f'$\\Phi$ Angle (Radians)')\n",
    "#         plt.xlim(-math.pi, math.pi)\n",
    "#         plt.ylabel(f'$\\Psi$ Angle (Radians)')\n",
    "#         plt.ylim(-math.pi, math.pi)\n",
    "#         plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 1,\n",
       " 1,\n",
       " 1,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0,\n",
       " 0]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
