{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n",
    "import math\n",
    "import imageio\n",
    "import json\n",
    "import time\n",
    "\n",
    "from collections import Counter\n",
    "from matplotlib.pyplot import figure\n",
    "from matplotlib import animation\n",
    "\n",
    "from Bio.PDB import *\n",
    "import nglview as nv\n",
    "import math\n",
    "import warnings\n",
    "from Bio.PDB.StructureBuilder import PDBConstructionWarning\n",
    "warnings.filterwarnings(\"ignore\", category=PDBConstructionWarning)\n",
    "\n",
    "# this is dangerous of course but uncomment it when you want to run the code\n",
    "# warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n",
    "\n",
    "\n",
    "\n",
    "# seg_dir = \"C:\\Users\\Kevin\\Documents\\TDA\\afrl\\Segmentations\"\n",
    "# seg_dir = \"C:/Users/Kevin/Documents/TDA/afrl/Segmentations\"\n",
    "seg_dir = \"../Batch_protein_all_5_3_no_na.csv_2019.07.08 16.10/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# aaron's code to make sure the folders are all in their proper place\n",
    "\n",
    "dirlocal = os.path.curdir\n",
    "data_path = os.path.join(dirlocal, 'Data/')\n",
    "protein_path = os.path.join(data_path, 'Proteins/')\n",
    "pdb_path = os.path.join(data_path, 'PDB/')\n",
    "fasta_path = os.path.join(data_path, 'fasta/')\n",
    "dssp_path = os.path.join(data_path, 'DSSP/')\n",
    "\n",
    "if not os.path.exists(data_path):\n",
    "    os.makedirs(data_path)\n",
    "if not os.path.exists(protein_path):\n",
    "    os.makedirs(protein_path)\n",
    "if not os.path.exists(pdb_path):\n",
    "    os.makedirs(pdb_path)\n",
    "ss_path = os.path.join(data_path, 'ss.txt')\n",
    "feature_path = os.path.join(data_path, 'sample-input-features.npy')\n",
    "distance_path = os.path.join(data_path, 'sample-distance-maps-cb.npy')\n",
    "full_feature_path = os.path.join(data_path, 'full-input-features.npy')\n",
    "full_distance_path = os.path.join(data_path, 'full-distance-maps-cb.npy')\n",
    "test_feature_path = os.path.join(data_path, 'testset-input-features.npy')\n",
    "test_distance_path = os.path.join(data_path, 'testset-distance-maps-cb.npy')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## aaron's code to get raw features from pdb\n",
    "\n",
    "import numpy as np\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from matplotlib.pyplot import figure\n",
    "%matplotlib inline\n",
    "import time\n",
    "import requests\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "# ! pip install biopython\n",
    "# ! pip install nglview\n",
    "# ! jupyter-nbextension enable nglview --py --sys-prefix\n",
    "\n",
    "from Bio.PDB import *\n",
    "import nglview as nv\n",
    "import math\n",
    "import warnings\n",
    "from Bio.PDB.StructureBuilder import PDBConstructionWarning\n",
    "warnings.filterwarnings(\"ignore\", category=PDBConstructionWarning)\n",
    "\n",
    "# this is dangerous of course but uncomment it when you want to run the code\n",
    "# warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fix_chip_start_stop(df, step_size, chip_size):\n",
    "\n",
    "    ch=[int(x.split(\"_\")[-1]) for x in df['chip_id']]\n",
    "\n",
    "    start=[c*step_size for c in ch]\n",
    "    stop=[chip_size+c*step_size for c in ch]\n",
    "\n",
    "    df['start']=start\n",
    "    df['stop']=stop\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "###  aaron's functions: \n",
    "\n",
    "# gets dataframe containing torsion angles, peptides, chain, etc\n",
    "# from pdb file\n",
    "def get_torsion_angles(pdb_id, degrees=False):\n",
    "    pdb_struct = get_pdb_structure(pdb_id)\n",
    "    torsion_angles = []\n",
    "    for model in pdb_struct:\n",
    "        for chain in model:\n",
    "            polypeptides = PPBuilder().build_peptides(chain)\n",
    "            for poly_index, poly in enumerate(polypeptides):\n",
    "                #print(\"Model %s Chain %s\" % (str(model.id), str(chain.id)))\n",
    "                #print(\"(part %i of %i)\" % (poly_index+1, len(polypeptides)))\n",
    "                #print(\"length %i\" % (len(poly)))\n",
    "                #print(\"from %s%i\" % (poly[0].resname, poly[0].id[1]))\n",
    "                #print(\"to %s%i\" % (poly[-1].resname, poly[-1].id[1]))\n",
    "                phi_psi = poly.get_phi_psi_list()\n",
    "                for res_index, residue in enumerate(poly):\n",
    "                    res_name = \"%s%i\" % (residue.resname, residue.id[1])\n",
    "                    #print(res_name, tuple(math.degrees(b) for b in phi_psi[res_index] if b))\n",
    "                    deg = phi_psi[res_index]\n",
    "                    if degrees:\n",
    "                        deg = tuple(math.degrees(b) if b else None for b in deg)\n",
    "                    phi, psi = deg\n",
    "                    model_name, model_id = model.full_id\n",
    "                    torsion_angles.append([model_name, model_id, chain.id, residue.resname, residue.id[1], phi, psi])\n",
    "    return pd.DataFrame(torsion_angles, columns=['Model_Name', 'Model_ID', 'Chain', 'Residue_Name', 'Residue_ID', 'Phi', 'Psi'])\n",
    "\n",
    "\n",
    "# generates ramachandran plot from a given pdb file\n",
    "def ramachandran_plot(pdb_id, degrees=True):\n",
    "    df = get_torsion_angles(pdb_id, degrees)\n",
    "    x = df['Phi']\n",
    "    y = df['Psi']\n",
    "    # Generate plot\n",
    "    plt.plot(x, y, \".\")\n",
    "    plt.title('Ramachandran Plot')\n",
    "    if degrees:\n",
    "        plt.xlabel(f'$\\Phi$ Angle (Degrees)')\n",
    "        plt.xlim(-180, 180)\n",
    "        plt.ylabel(f'$\\Psi$ Angle (Radians)')\n",
    "        plt.ylim(-180, 180)\n",
    "        plt.show()\n",
    "    else:\n",
    "        plt.xlabel(f'$\\Phi$ Angle (Radians)')\n",
    "        plt.xlim(-math.pi, math.pi)\n",
    "        plt.ylabel(f'$\\Psi$ Angle (Radians)')\n",
    "        plt.ylim(-math.pi, math.pi)\n",
    "        plt.show()\n",
    "\n",
    "# helper function which generates a biopython structure\n",
    "# for display in display_protein()\n",
    "def get_pdb_structure(pdb_id):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    parser = PDBParser()\n",
    "    file_path = os.path.join(pdb_path, f\"{pdb_id}.pdb\")\n",
    "    try:\n",
    "        struct = parser.get_structure(pdb_id, file_path)\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "        return None\n",
    "    return struct\n",
    "\n",
    "\n",
    "#  downloads the protein in question from pdb\n",
    "def get_pdb_file(pdb_id, replace=False):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    file_path = os.path.join(pdb_path, f'{pdb_id}.pdb')\n",
    "    if os.path.isfile(file_path) and not replace:\n",
    "        return True\n",
    "    parser = PDBParser()\n",
    "    url = f'https://files.rcsb.org/download/{pdb_id}.pdb'\n",
    "    resp = requests.get(url)\n",
    "    try:\n",
    "        #print(file_path)\n",
    "        file = open(file_path, \"wb\")\n",
    "        file.write(resp.content)\n",
    "        file.close()\n",
    "    except Exception as e: \n",
    "        print(e)\n",
    "        return False\n",
    "    return True\n",
    "\n",
    "# neato 3d viewer of a protein's structure\n",
    "def display_protein(pdb_id):\n",
    "    pdb_struct = get_pdb_structure(pdb_id)\n",
    "    view = nv.show_biopython(pdb_struct)\n",
    "    return view\n",
    "\n",
    "\n",
    "#  returns a dataframe with secondary structure, amino acid code, chain id, etc\n",
    "# deprecated\n",
    "def get_secondary_structure(pdb_id):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    indexes = get_pdb_ss_seq(pdb_id)\n",
    "    chain_df = []\n",
    "    for i, chain_ss in enumerate(indexes):\n",
    "        chain_id, seq, ss = chain_ss\n",
    "        #unknown_list = [i for i, c in enumerate(seq) if c == 'X']\n",
    "        #print(unknown_list)\n",
    "        seq_list = [c for c in seq]\n",
    "        # [f(x) if condition else g(x) for x in sequence]\n",
    "        ss_list = ['L' if c is ' ' else c for c in ss]\n",
    "        #print(seq_list + ss_list)\n",
    "        chain_df.append(pd.DataFrame({'Chain': chain_id, 'Amino_Acid':seq_list, 'Secondary_Structure':ss_list}))\n",
    "    return pd.concat(chain_df)\n",
    "\n",
    "# get matrix of pairwise carbon beta distances\n",
    "def get_cb_distances(full_pdb_id, partial=False):\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    get_pdb_file(pdb_id)\n",
    "    pdb_struct = get_pdb_structure(pdb_id)  # Returns the BioPython structure of the PDB file\n",
    "    ppb = PPBuilder()\n",
    "    chain_dist_matrix = {}\n",
    "    for model in pdb_struct:\n",
    "        for chain in model:\n",
    "            if chain.id == full_pdb_id[4] and partial:\n",
    "                residues = []\n",
    "                for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                    for residue in pp:\n",
    "                        residues.append(residue)\n",
    "                lng = len(residues)\n",
    "                dist_matrix = np.zeros((lng, lng), np.float)\n",
    "                for i in range(0, lng):\n",
    "                    for j in range(i, lng):\n",
    "                        try:\n",
    "                            diff_vector  = residues[i]['C'].coord - residues[j]['C'].coord\n",
    "                        except Exception:\n",
    "                            # This just gets the average value around the unknown residue distance\n",
    "                            count = 0\n",
    "                            sums = 0\n",
    "                            if j - 1 > 0:\n",
    "                                count += 1\n",
    "                                sums += dist_matrix[i][j - 1]\n",
    "                            if i - 1 > 0:\n",
    "                                count += 1\n",
    "                                sums += dist_matrix[i - 1][j]\n",
    "                            if count > 0:\n",
    "                                diff_vector = sums / count\n",
    "                            else:\n",
    "                                diff_vector = 0\n",
    "                        dist = np.sqrt(np.sum(diff_vector * diff_vector))\n",
    "                        dist_matrix[i][j] = dist\n",
    "                        dist_matrix[j][i] = dist\n",
    "                    chain_dist_matrix = dist_matrix\n",
    "        \n",
    "    if partial:\n",
    "        start_pos, end_pos, bool_arr = get_seq_pos(full_pdb_id)\n",
    "        if start_pos == -1:\n",
    "            return pd.DataFrame(chain_dist_matrix[bool_arr]).iloc[:, bool_arr]\n",
    "        return pd.DataFrame(chain_dist_matrix[start_pos:end_pos]).iloc[:, start_pos:end_pos]\n",
    "    else:\n",
    "        return chain_dist_matrix\n",
    "    \n",
    "    \n",
    "# gets alignment of given sequence on top of the original sequence\n",
    "# from a fasta file downloaded from the cloud repository\n",
    "def get_seq_alignment(full_pdb_id):\n",
    "    seq = ''\n",
    "    for record in SeqIO.parse(os.path.join(fasta_path, f'{full_pdb_id}.fasta'), \"fasta\"):\n",
    "        seq = record.seq\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    #print(pdb_id)\n",
    "    get_pdb_file(pdb_id)\n",
    "    struct = get_pdb_structure(pdb_id)\n",
    "    ppb = PPBuilder()\n",
    "    for model in struct:\n",
    "        for chain in model:\n",
    "            if chain.id == full_pdb_id[4]:\n",
    "                ground_seq = ''\n",
    "                for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                    ground_seq += pp.get_sequence()\n",
    "                alignments = pairwise2.align.globalxx(ground_seq, str(seq))\n",
    "                al_seq = alignments[0][1]\n",
    "                al_seq_bool = [False if a is '-' else True for a in al_seq]\n",
    "                return al_seq_bool\n",
    "\n",
    "            \n",
    "def get_seq_pos(full_pdb_id):\n",
    "    seq = ''\n",
    "    for record in SeqIO.parse(os.path.join(fasta_path, f'{full_pdb_id}.fasta'), \"fasta\"):\n",
    "        seq = str(record.seq)\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    get_pdb_file(pdb_id)\n",
    "    struct = get_pdb_structure(pdb_id)\n",
    "    ppb = PPBuilder()\n",
    "    for model in struct:\n",
    "        for chain in model:\n",
    "            if chain.id == full_pdb_id[4]:\n",
    "                ground_seq = ''\n",
    "                for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                    ground_seq += pp.get_sequence()\n",
    "                start_pos = ground_seq.find(seq)\n",
    "                if start_pos == -1:\n",
    "                    alignments = pairwise2.align.globalxx(ground_seq, seq)\n",
    "                    al_seq = alignments[0][1]\n",
    "                    bool_seq = [False if a == '-' else True for a in al_seq]\n",
    "                    #start_pos = largest_substring_pos(ground_seq, seq)\n",
    "                    #start_pos = int(alignments[0][2] - 1)\n",
    "                    #end_pos = start_pos + len(al_seq)\n",
    "                    return (-1, -1, bool_seq)\n",
    "                else:\n",
    "                    end_pos = start_pos + len(seq)\n",
    "                    return (start_pos, end_pos, '')\n",
    "\n",
    "\n",
    "            \n",
    "            \n",
    "# calls dssp api to get dssp file\n",
    "def pdb_id_to_dssp_file(pdb_id, replace=False):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    rest_url = 'http://www.cmbi.umcn.nl/xssp/'\n",
    "    # Read the pdb id data into a variable\n",
    "    data = {'data': pdb_id}\n",
    "\n",
    "    # Send a request to the server to create hssp data from the pdb file data.\n",
    "    # If an error occurs, an exception is raised and the program exits. If the\n",
    "    # request is successful, the id of the job running on the server is\n",
    "    # returned.\n",
    "    url_create = f'{rest_url}api/create/pdb_id/dssp/'\n",
    "    r = requests.post(url_create, data=data)\n",
    "    r.raise_for_status()\n",
    "\n",
    "    job_id = json.loads(r.text)['id']\n",
    "    #print(f'Job submitted successfully. Id is: {job_id}')\n",
    "\n",
    "    # Loop until the job running on the server has finished, either successfully\n",
    "    # or due to an error.\n",
    "    ready = False\n",
    "    while not ready:\n",
    "        # Check the status of the running job. If an error occurs an exception\n",
    "        # is raised and the program exits. If the request is successful, the\n",
    "        # status is returned.\n",
    "        url_status = f'{rest_url}api/status/pdb_id/dssp/{job_id}/'\n",
    "        r = requests.get(url_status)\n",
    "        r.raise_for_status()\n",
    "\n",
    "        status = json.loads(r.text)['status']\n",
    "        #print(f'Job status is: {status}')\n",
    "\n",
    "        # If the status equals SUCCESS, exit out of the loop by changing the\n",
    "        # condition ready. This causes the code to drop into the `else` block\n",
    "        # below.\n",
    "        #\n",
    "        # If the status equals either FAILURE or REVOKED, an exception is raised\n",
    "        # containing the error message. The program exits.\n",
    "        #\n",
    "        # Otherwise, wait for five seconds and start at the beginning of the\n",
    "        # loop again.\n",
    "        if status == 'SUCCESS':\n",
    "            ready = True\n",
    "        elif status in ['FAILURE', 'REVOKED']:\n",
    "            raise Exception(json.loads(r.text)['message'])\n",
    "        else:\n",
    "            time.sleep(5)\n",
    "    else:\n",
    "        # Requests the result of the job. If an error occurs an exception is\n",
    "        # raised and the program exits. If the request is successful, the result\n",
    "        # is returned.\n",
    "        url_result = f'{rest_url}api/result/pdb_id/dssp/{job_id}/'\n",
    "        r = requests.get(url_result)\n",
    "        r.raise_for_status()\n",
    "        result = json.loads(r.text)['result']\n",
    "        try:\n",
    "            file_path = os.path.join(dssp_path, f'{pdb_id}.dssp')\n",
    "            if os.path.isfile(file_path) and not replace:\n",
    "                return True\n",
    "            #print(file_path)\n",
    "            file = open(file_path, \"w\")\n",
    "            file.write(result)\n",
    "            file.close()\n",
    "        except Exception as e: \n",
    "            print(e)\n",
    "            return False\n",
    "        # Return the result to the caller, which prints it to the screen.\n",
    "        return True\n",
    "\n",
    "# uses api call file to generate secondary structure, phi/psi, solvent-accessibility stuff\n",
    "# because of indexing stuff, it crashes if we use partial = True\n",
    "# ask aaron about that\n",
    "def get_ground_truth_api(full_pdb_id, partial=False):\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    get_pdb_file(pdb_id)\n",
    "    struct = get_pdb_structure(pdb_id)\n",
    "    ppb = PPBuilder()\n",
    "    struct_info = []\n",
    "    for model in struct:\n",
    "        file = os.path.join(pdb_path, f'{pdb_id}.pdb')\n",
    "        pdb_id_to_dssp_file(pdb_id)\n",
    "        dssp = DSSP(model=model, in_file=os.path.join(dssp_path, f'{pdb_id}.dssp'), file_type='DSSP')\n",
    "        seq_count = 0\n",
    "        #dssp = DSSP(model=model, in_file=file)\n",
    "        for chain in model:\n",
    "            if chain.id == full_pdb_id[4] or not partial:\n",
    "                residues = []\n",
    "                seq = ''\n",
    "                for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                    seq += pp.get_sequence()\n",
    "                    for residue in pp:\n",
    "                        residues.append(residue)\n",
    "                for i, residue in enumerate(residues):\n",
    "                    try:\n",
    "                        key = list(dssp.keys())[seq_count]   \n",
    "                        dssp_info = dssp[key]\n",
    "                        amino_acid = dssp_info[1]\n",
    "                        sec_struct = dssp_info[2]\n",
    "                        solv_acc = dssp_info[3]\n",
    "                        phi = dssp_info[4]\n",
    "                        psi = dssp_info[5]\n",
    "                    except Exception:\n",
    "                        # DSSP didn't capture this amino acid\n",
    "                        #print(full_pdb_id, seq_count, len(dssp), len(seq), residue)\n",
    "                        amino_acid = seq[i]\n",
    "                        sec_struct = '-'\n",
    "                        solv_acc = 0\n",
    "                        phi = 360\n",
    "                        psi = 360\n",
    "                    seq_count += 1\n",
    "                    # Keys 6 through 13 is bonding energy / relidx (no clue what this is)\n",
    "                    struct_info.append([model.full_id[0], model.full_id[1], chain.id,\n",
    "                        residue.resname, residue.id[1], amino_acid, sec_struct, solv_acc, phi, psi])\n",
    "    info_df = pd.DataFrame(struct_info, \n",
    "        columns=['Model_Name', 'Model_ID', 'Chain', 'Residue_Name',\n",
    "        'Residue_ID', 'Amino_Acid', 'Secondary_Structure', 'Solvent_Accessability', \n",
    "        'Phi', 'Psi'])\n",
    "    if partial:    \n",
    "        start_pos, end_pos, bool_arr = get_seq_pos(full_pdb_id)\n",
    "        if start_pos == -1:\n",
    "            return info_df[bool_arr]\n",
    "        return info_df[start_pos:end_pos]\n",
    "    else:\n",
    "        return info_df\n",
    "\n",
    "\n",
    "    \n",
    "# the following two functions download the files from the api\n",
    "# to create a cache on the disk\n",
    "# speeds things up immensely, only need to run once though\n",
    "def get_ground_truth_files(full_pdb_id, replace=False, partial=False):\n",
    "    bad_list = ['4osnA0']\n",
    "    if full_pdb_id in bad_list:\n",
    "        return 'Bad File'\n",
    "    if partial:\n",
    "        file_path = os.path.join(protein_path, full_pdb_id + '.csv')\n",
    "    else:\n",
    "        file_path = os.path.join(protein_path, full_pdb_id[0:4] + '.csv')\n",
    "        \n",
    "    if os.path.isfile(file_path) and not replace:\n",
    "        return True\n",
    "    \n",
    "    df = get_ground_truth(full_pdb_id, partial)\n",
    "    file_path = ''\n",
    "    try:\n",
    "        df.to_csv(file_path)\n",
    "    except Exception:\n",
    "        return False\n",
    "    return True\n",
    "    \n",
    "def get_cb_distance_files(full_pdb_id, replace=False, partial=False):\n",
    "    bad_list = ['4osnA0']\n",
    "    if full_pdb_id in bad_list:\n",
    "        return 'Bad File'\n",
    "    file_path = ''\n",
    "    if partial:\n",
    "        file_path = os.path.join(distances_path, full_pdb_id + '.csv')\n",
    "        if os.path.isfile(file_path) and not replace:\n",
    "            return True\n",
    "        try:\n",
    "            df = get_cb_distances(full_pdb_id, partial)\n",
    "            df.to_csv(file_path)\n",
    "        except Exception:\n",
    "            return False\n",
    "        return True\n",
    "    else:\n",
    "        df = get_cb_distances(full_pdb_id, partial)\n",
    "        keys = df.keys\n",
    "        for key in keys:\n",
    "            file_path = os.path.join(distances_path, full_pdb_id[0:4] + key + '0.csv')\n",
    "            if os.path.isfile(file_path) and not replace:\n",
    "                return True\n",
    "            try:\n",
    "                df[key].to_csv(file_path)\n",
    "            except Exception:\n",
    "                return False\n",
    "        return True\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_25=pandas.read_csv(\"datasets/protein_all_25_12_no_na.csv\")\n",
    "dataset_10=pandas.read_csv(\"datasets/protein_all_10_5_no_na.csv\")\n",
    "dataset_5=pandas.read_csv(\"datasets/protein_all_5_3_no_na.csv\")\n",
    "\n",
    "fix_chip_start_stop(dataset_25,12,25)\n",
    "fix_chip_start_stop(dataset_10,5,10)\n",
    "fix_chip_start_stop(dataset_5,3,5)\n",
    "\n",
    "\n",
    "pdb_list, length_dict, aa_seq, input_features = np.load(\"datasets/testset-input-features.npy\",allow_pickle=True)\n",
    "pdb_list_y, distance_maps_cb = np.load(\"datasets/testset-distance-maps-cb.npy\",encoding=\"latin1\",allow_pickle=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions\n",
    "def get_rows_from_segmentation(seg, dataset):\n",
    "    with open(seg_dir+seg) as jsonf:\n",
    "        a=json.load(jsonf)\n",
    "        df=dataset.iloc[[int(m) for m in a['rows']],:]\n",
    "        return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def draw_ss_predicts(index,ss):\n",
    "    \n",
    "#     H, G and I to H; E to E; the rest to C\n",
    "    ss_dict={\n",
    "        \"H\":0,\n",
    "        \"G\":0,\n",
    "        \"I\":0,\n",
    "        \"E\":1,\n",
    "        \"B\":2,\n",
    "        \"T\":2,\n",
    "        \"S\":2             \n",
    "            }\n",
    "    \n",
    "    aa_length=length_dict[pdb_list[index]]\n",
    "    \n",
    "    plt.figure(figsize=(15,10))\n",
    "\n",
    "    df25=dataset_25[dataset_25[\"protein_id\"]==dataset_25[\"protein_id\"].unique()[index]] \n",
    "    df10=dataset_10[dataset_10[\"protein_id\"]==dataset_10[\"protein_id\"].unique()[index]] \n",
    "    df5=dataset_5[dataset_5[\"protein_id\"]==dataset_5[\"protein_id\"].unique()[index]] \n",
    "\n",
    "    plt.subplot(4,1,1)\n",
    "    y=ground_truth_window_avg(df5)[\"ground_truth_\"+ss]\n",
    "    x=range(0,len(y))\n",
    "    plt.bar(x,y,1)\n",
    "    plt.xlim((0,aa_length))    \n",
    "\n",
    "    plt.subplot(4,1,2)\n",
    "    y=ground_truth_window_avg(df10)[\"ground_truth_\"+ss]\n",
    "    x=range(0,len(y))\n",
    "    plt.bar(x,y,1)\n",
    "    plt.xlim((0,aa_length))\n",
    "\n",
    "    plt.subplot(4,1,3)\n",
    "    y=ground_truth_window_avg(df25)[\"ground_truth_\"+ss]\n",
    "    x=range(0,len(y))\n",
    "    plt.bar(x,y,1)\n",
    "    plt.xlim((0,aa_length))\n",
    "\n",
    "    plt.subplot(4,1,4)\n",
    "    y=[1-n for n in input_features[pdb_list[index]][ss_dict[ss]].astype(np.float)]\n",
    "    x=range(0,len(y))\n",
    "    plt.bar(x,y,1)\n",
    "    plt.xlim((0,aa_length))\n",
    "\n",
    "    \n",
    "    \n",
    "def get_inv_psipred_from_ss(protein):\n",
    "#     H to H; E to E; the rest to C\n",
    "    ss_dict={\n",
    "        \"H\":0,\n",
    "        \"G\":0,\n",
    "        \"I\":0,\n",
    "        \"E\":1,\n",
    "        \"B\":1,\n",
    "        \"T\":2,\n",
    "        \"S\":2,\n",
    "        \"-\":2\n",
    "            }\n",
    "    \n",
    "    # I subtract it from 1 to make it on the same scale as the other predictions\n",
    "    # (i.e. high means more likely to be X)\n",
    "    \n",
    "    return [1-n for n in input_features[protein][ss_dict[ss]].astype(np.float)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"HBEGITS\"\n",
    "# draw_ss_predicts(1,\"H\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_reduced_ss(protein_id):\n",
    "    ss_file=os.path.join(data_path,\"ss\",protein_id+\".ss\")\n",
    "    sf=open(ss_file).read()\n",
    "    return(list(sf.split(\"\\n\")[1]))\n",
    "   \n",
    "# get_reduced_ss(\"1bebA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ground_truth_window_avg(df):\n",
    "    aa_length=max(df[\"stop\"])\n",
    "    mean_ss=[]\n",
    "    gt=[\"ground_truth_\"+ss for ss in \"HBEGITS\"]\n",
    "\n",
    "    for aa in range(0,aa_length):\n",
    "        chip=df[(df[\"start\"]<=aa) & (df[\"stop\"]>aa)]\n",
    "        mean_ss.append(chip[gt].mean(0))\n",
    "    gt_avg=pandas.DataFrame(mean_ss)\n",
    "    \n",
    "    return gt_avg\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for each protein:\n",
    "#     for chip in protein\n",
    "#         get avg ss's for chip\n",
    "#     get avg ss for each aa\n",
    "#     get psipred for each aa\n",
    "#     present as bar graph\n",
    "\n",
    "#     get gt for protein !!!!\n",
    "\n",
    "#     get thd_rmse for protein\n",
    "#     get psipred_rmse for protein\n",
    "#     put rmse's in vectors\n",
    "\n",
    "# graph rmse's\n",
    "\n",
    "# the following is pseudocode\n",
    "\n",
    "def rmse(a,b):\n",
    "    return np.sqrt(np.mean((a-b)**2))\n",
    "\n",
    "def eval_test_data_new(leaf_df,dataset):\n",
    "    output_df=pandas.DataFrame()\n",
    "    \n",
    "    for protein in pandas.unique(leaf_df['protein_id']):\n",
    "        print(protein)\n",
    "        mean_ss=[]\n",
    "        gt_names=[\"pred_gt_\"+ss for ss in \"HBEGITS\"]\n",
    "\n",
    "        # this part is avg'ing across the wrong stuff\n",
    "        # all the values end up the same\n",
    "        \n",
    "        for aa in range(length_dict[protein]):\n",
    "            chip=leaf_df[(leaf_df[\"start\"]<=aa) & (leaf_df[\"stop\"]>aa)]\n",
    "            mean_ss.append([protein]+list(chip[gt_names].mean(0)))\n",
    "        gt_avg=pandas.DataFrame(mean_ss)\n",
    "        gt_avg.columns=[\"protein_id\"]+[\"pred_ss_\"+ss for ss in \"HBEGITS\"]\n",
    "        \n",
    "        \n",
    "        ssgt=[]\n",
    "        \n",
    "        for sec_s in \"HBEGITS\":\n",
    "        # get gt for protein\n",
    "            ss=get_ground_truth_api(protein)['Secondary_Structure']\n",
    "            gt=[int(x) for x in ss==sec_s]\n",
    "            gt_avg[\"ground_truth_\"+sec_s]=gt\n",
    "            \n",
    "        gt_avg['psipred_helix']=input_features[protein][0].astype(np.float)\n",
    "        gt_avg['psipred_sheet']=input_features[protein][1].astype(np.float)\n",
    "        gt_avg['psipred_coil']=input_features[protein][2].astype(np.float)\n",
    "        \n",
    "        output_df.append(gt_avg)\n",
    "    return output_df\n",
    "\n",
    "    \n",
    "def eval_test_data_reduced(leaf_df,dataset):\n",
    "    output_df=pandas.DataFrame()\n",
    "    \n",
    "    for protein in pandas.unique(leaf_df['protein_id']):\n",
    "        \n",
    "        df_prot=leaf_df[leaf_df['protein_id']==protein]\n",
    "        \n",
    "        print(protein)\n",
    "        mean_ss=[]\n",
    "        gt_names=[\"thd_pred_helix\",\"thd_pred_sheet\",\"thd_pred_coil\"]\n",
    "\n",
    "        # this part is avg'ing across the wrong stuff\n",
    "        # all the values end up the same\n",
    "        \n",
    "        for aa in range(length_dict[protein]):\n",
    "            chip=df_prot[(df_prot[\"start\"]<=aa) & (df_prot[\"stop\"]>aa)]\n",
    "            mean_ss.append([protein,aa]+list(chip[gt_names].mean(0)))\n",
    "        gt_avg=pandas.DataFrame(mean_ss)\n",
    "        gt_avg.columns=[\"protein_id\",\"aa_index\"]+[\"pred_ss_\"+ss for ss in \"HEC\"]\n",
    "        \n",
    "        \n",
    "        ssgt=[]\n",
    "        \n",
    "        for sec_s in \"HEC\":\n",
    "        # get gt for protein\n",
    "            ss=get_reduced_ss(protein)\n",
    "            gt=[int(x==sec_s) for x in ss]\n",
    "            gt_avg[\"ground_truth_\"+sec_s]=gt\n",
    "            \n",
    "        gt_avg['psipred_helix']=input_features[protein][0].astype(np.float)\n",
    "        gt_avg['psipred_sheet']=input_features[protein][1].astype(np.float)\n",
    "        gt_avg['psipred_coil']=input_features[protein][2].astype(np.float)\n",
    "        \n",
    "        gt_avg['inv_psipred_helix']=1-input_features[protein][0].astype(np.float)\n",
    "        gt_avg['inv_psipred_sheet']=1-input_features[protein][1].astype(np.float)\n",
    "        gt_avg['inv_psipred_coil']=1-input_features[protein][2].astype(np.float)\n",
    " \n",
    "        \n",
    "        output_df=output_df.append(gt_avg)\n",
    "        \n",
    "    return output_df\n",
    "\n",
    "    \n",
    "    \n",
    "def seg_rows_formatted(thd_name,file_id,dataset):\n",
    "    seg=\"GROUPDATA_\"+thd_name+\" \"+file_id+\".json\"\n",
    "    return get_rows_from_segmentation(thd_name+\"/\"+seg,dataset)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_pred_gt(leafdf,dataset):\n",
    "\n",
    "    gt=[\"ground_truth_\"+ss for ss in \"HBEGITS\"]\n",
    "\n",
    "    ss_pred=pandas.DataFrame()\n",
    "\n",
    "    for index, row in leafdf.iterrows():\n",
    "        ssp=seg_rows_formatted(row['thd_id'],row['file_id'],dataset)[gt].mean(0)\n",
    "        ss_pred=ss_pred.append(ssp, ignore_index=True)\n",
    "        \n",
    "    ss_pred.columns=[\"pred_gt_\"+ss for ss in \"HBEGITS\"]\n",
    "    return pandas.concat([leafdf,ss_pred],axis=1)\n",
    "\n",
    "def add_pred_condensed(leafdf,dataset):\n",
    "    \n",
    "    gt=[\"ground_truth_\"+ss for ss in \"HBEGITS\"]\n",
    "\n",
    "    ss_pred=pandas.DataFrame()\n",
    "\n",
    "    for index, row in leafdf.iterrows():\n",
    "        ssp=seg_rows_formatted(row['thd_id'],row['file_id'],dataset)[gt]\n",
    "        \n",
    "        helix=[\"ground_truth_\"+ss for ss in \"HGI\"]\n",
    "        sheet=[\"ground_truth_\"+ss for ss in \"EB\"]\n",
    "        coil=[\"ground_truth_\"+ss for ss in \"TS\"]\n",
    "\n",
    "        hlx=ssp[helix].sum(axis=1)\n",
    "        sh=ssp[sheet].sum(axis=1)\n",
    "        cl=1-(hlx+sh)\n",
    "\n",
    "        pred=pandas.concat([hlx,sh,cl],axis=1)\n",
    "\n",
    "        ss_pred=ss_pred.append(pred.mean(axis=0),ignore_index=True)\n",
    "    ss_pred.columns=[\"thd_pred_helix\",\"thd_pred_sheet\",\"thd_pred_coil\"]\n",
    "    return pandas.concat([leafdf,ss_pred],axis=1)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "thd_leafs=add_pred_condensed(pandas.read_csv(\"datasets/test6_sequences_pred.csv\"),dataset_5)\n",
    "thd_leafs.to_csv(\"data/thd_leafs.csv\")\n",
    "\n",
    "prediction_df=eval_test_data_reduced(thd_leafs,dataset_5)\n",
    "prediction_df.to_csv(\"data/reduced_predict_fixed.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this gives us a table of the accuracy results\n",
    "\n",
    "def pred_accuracy(pred_df):\n",
    "\n",
    "    rmse_df=pandas.DataFrame()\n",
    "\n",
    "    psipred=[\"inv_psipred_helix\",\"inv_psipred_sheet\",\"inv_psipred_coil\"]\n",
    "\n",
    "    results_df=pandas.DataFrame()\n",
    "\n",
    "    for ss in range(3):\n",
    "\n",
    "        pred=\"pred_ss_\"+\"HEC\"[ss]\n",
    "        gt=\"ground_truth_\"+\"HEC\"[ss]\n",
    "        psi=psipred[ss]\n",
    "\n",
    "        rmse_pred=rmse(pred_df[pred],pred_df[gt])\n",
    "        acc_pred=1-np.mean(abs(pred_df[pred]-pred_df[gt]))\n",
    "\n",
    "        rmse_psi=rmse(pred_df[psi],pred_df[gt])\n",
    "        acc_psi=1-np.mean(abs(pred_df[psi]-pred_df[gt]))\n",
    "\n",
    "        row=[[\"helix\",\"sheet\",\"coil\"][ss]]+[rmse_pred,acc_pred,rmse_psi,acc_psi]\n",
    "        row=pandas.Series(row)\n",
    "        results_df=results_df.append(row,ignore_index=True)\n",
    "\n",
    "    columns=[\"structure\"]\n",
    "    columns+=[\"thd_pred_rmse\",\"thd_pred_accuracy\"]\n",
    "    columns+=[\"psipred_rmse\",\"psipred_accuracy\"]\n",
    "\n",
    "    results_df.columns=columns\n",
    "    \n",
    "    return results_df\n",
    "\n",
    "    \n",
    "# see sequence graphed on all classes\n",
    "#      because ryan wants to know how to threshhold it\n",
    "# accuracy for the aa that are helices\n",
    "# add a none % column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# accuracy for the aa that ARE x\n",
    "\n",
    "def pred_accuracy_signal(pred_df):\n",
    "\n",
    "    rmse_df=pandas.DataFrame()\n",
    "\n",
    "    psipred=[\"inv_psipred_helix\",\"inv_psipred_sheet\",\"inv_psipred_coil\"]\n",
    "\n",
    "    results_df=pandas.DataFrame()\n",
    "\n",
    "    for ss in range(3):\n",
    "\n",
    "        pdf2=pred_df[pred_df['ground_truth_'+\"HEC\"[ss]]==1]\n",
    "\n",
    "        pred=\"pred_ss_\"+\"HEC\"[ss]\n",
    "        gt=\"ground_truth_\"+\"HEC\"[ss]\n",
    "        psi=psipred[ss]\n",
    "\n",
    "        rmse_pred=rmse(pdf2[pred],pdf2[gt])\n",
    "        acc_pred=1-np.mean(abs(pdf2[pred]-pdf2[gt]))\n",
    "\n",
    "        rmse_psi=rmse(pdf2[psi],pdf2[gt])\n",
    "        acc_psi=1-np.mean(abs(pdf2[psi]-pdf2[gt]))\n",
    "\n",
    "        row=[[\"helix\",\"sheet\",\"coil\"][ss]]+[rmse_pred,acc_pred,rmse_psi,acc_psi]\n",
    "        row=pandas.Series(row)\n",
    "        results_df=results_df.append(row,ignore_index=True)\n",
    "\n",
    "    columns=[\"structure_signal\"]\n",
    "    columns+=[\"thd_pred_rmse_signal\",\"thd_pred_accuracy_signal\"]\n",
    "    columns+=[\"psipred_rmse_signal\",\"psipred_accuracy_signal\"]\n",
    "\n",
    "    results_df.columns=columns\n",
    "\n",
    "    return results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_id_0</th>\n",
       "      <th>file_id_1</th>\n",
       "      <th>file_id_2</th>\n",
       "      <th>file_id_3</th>\n",
       "      <th>file_id_4</th>\n",
       "      <th>file_id_5</th>\n",
       "      <th>file_id_6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>28.1.30</td>\n",
       "      <td>20.31.60</td>\n",
       "      <td>20.31.60</td>\n",
       "      <td>20.31.60</td>\n",
       "      <td>28.1.30</td>\n",
       "      <td>20.31.42</td>\n",
       "      <td>20.31.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>27.1.21</td>\n",
       "      <td>29.0.11</td>\n",
       "      <td>29.0.11</td>\n",
       "      <td>29.0.11</td>\n",
       "      <td>17.0.0</td>\n",
       "      <td>30.0.74</td>\n",
       "      <td>30.0.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15.1.19</td>\n",
       "      <td>27.1.12</td>\n",
       "      <td>15.1.19</td>\n",
       "      <td>27.1.12</td>\n",
       "      <td>30.0.47</td>\n",
       "      <td>27.1.12</td>\n",
       "      <td>27.1.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29.0.7</td>\n",
       "      <td>28.1.6</td>\n",
       "      <td>28.1.6</td>\n",
       "      <td>28.1.6</td>\n",
       "      <td>29.0.7</td>\n",
       "      <td>28.1.6</td>\n",
       "      <td>28.1.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16.0.0</td>\n",
       "      <td>27.1.11</td>\n",
       "      <td>27.1.11</td>\n",
       "      <td>27.1.11</td>\n",
       "      <td>16.0.0</td>\n",
       "      <td>23.1.63</td>\n",
       "      <td>23.1.63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>28.1.23</td>\n",
       "      <td>23.1.19</td>\n",
       "      <td>23.1.19</td>\n",
       "      <td>23.1.19</td>\n",
       "      <td>17.0.0</td>\n",
       "      <td>14.29.17</td>\n",
       "      <td>14.29.17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>30.0.80</td>\n",
       "      <td>27.1.11</td>\n",
       "      <td>27.1.11</td>\n",
       "      <td>27.1.11</td>\n",
       "      <td>30.0.24</td>\n",
       "      <td>27.1.11</td>\n",
       "      <td>27.1.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>9.0.0</td>\n",
       "      <td>30.0.84</td>\n",
       "      <td>30.0.84</td>\n",
       "      <td>30.0.84</td>\n",
       "      <td>9.0.0</td>\n",
       "      <td>30.0.84</td>\n",
       "      <td>30.0.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>16.1.11</td>\n",
       "      <td>8.0.0</td>\n",
       "      <td>16.1.11</td>\n",
       "      <td>8.0.0</td>\n",
       "      <td>2.0.0</td>\n",
       "      <td>8.0.0</td>\n",
       "      <td>8.0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>30.0.80</td>\n",
       "      <td>30.0.61</td>\n",
       "      <td>30.0.61</td>\n",
       "      <td>30.0.61</td>\n",
       "      <td>15.0.0</td>\n",
       "      <td>30.0.104</td>\n",
       "      <td>30.0.104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1.0.0</td>\n",
       "      <td>6.0.0</td>\n",
       "      <td>6.0.0</td>\n",
       "      <td>6.0.0</td>\n",
       "      <td>1.0.0</td>\n",
       "      <td>23.1.57</td>\n",
       "      <td>23.1.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>30.0.92</td>\n",
       "      <td>18.4.9</td>\n",
       "      <td>18.4.9</td>\n",
       "      <td>18.4.9</td>\n",
       "      <td>30.0.75</td>\n",
       "      <td>15.1.20</td>\n",
       "      <td>15.1.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>29.0.15</td>\n",
       "      <td>23.1.53</td>\n",
       "      <td>23.1.53</td>\n",
       "      <td>23.1.53</td>\n",
       "      <td>30.0.105</td>\n",
       "      <td>20.31.44</td>\n",
       "      <td>20.31.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>27.1.35</td>\n",
       "      <td>23.1.21</td>\n",
       "      <td>23.1.21</td>\n",
       "      <td>23.1.21</td>\n",
       "      <td>22.0.0</td>\n",
       "      <td>23.1.23</td>\n",
       "      <td>23.1.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>30.0.35</td>\n",
       "      <td>27.1.17</td>\n",
       "      <td>27.1.17</td>\n",
       "      <td>27.1.17</td>\n",
       "      <td>14.0.1</td>\n",
       "      <td>27.1.10</td>\n",
       "      <td>14.0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>28.1.49</td>\n",
       "      <td>16.1.16</td>\n",
       "      <td>28.1.49</td>\n",
       "      <td>16.1.16</td>\n",
       "      <td>30.0.80</td>\n",
       "      <td>16.1.16</td>\n",
       "      <td>16.1.16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>23.1.64</td>\n",
       "      <td>19.5.32</td>\n",
       "      <td>23.1.64</td>\n",
       "      <td>19.5.32</td>\n",
       "      <td>23.1.64</td>\n",
       "      <td>23.1.53</td>\n",
       "      <td>23.1.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>30.0.80</td>\n",
       "      <td>23.1.37</td>\n",
       "      <td>30.0.80</td>\n",
       "      <td>23.1.37</td>\n",
       "      <td>1.0.0</td>\n",
       "      <td>23.1.37</td>\n",
       "      <td>23.1.37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>23.1.12</td>\n",
       "      <td>18.4.9</td>\n",
       "      <td>23.1.12</td>\n",
       "      <td>18.4.9</td>\n",
       "      <td>0.0.0</td>\n",
       "      <td>18.4.9</td>\n",
       "      <td>18.4.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>0.0.0</td>\n",
       "      <td>30.0.55</td>\n",
       "      <td>30.0.55</td>\n",
       "      <td>30.0.55</td>\n",
       "      <td>13.0.20</td>\n",
       "      <td>30.0.32</td>\n",
       "      <td>13.0.20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>0.0.0</td>\n",
       "      <td>23.1.29</td>\n",
       "      <td>23.1.29</td>\n",
       "      <td>23.1.29</td>\n",
       "      <td>0.0.0</td>\n",
       "      <td>23.0.3</td>\n",
       "      <td>23.0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>16.31.33</td>\n",
       "      <td>23.1.53</td>\n",
       "      <td>16.31.33</td>\n",
       "      <td>23.1.53</td>\n",
       "      <td>30.0.36</td>\n",
       "      <td>23.1.53</td>\n",
       "      <td>23.1.53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>30.0.80</td>\n",
       "      <td>20.31.24</td>\n",
       "      <td>20.31.24</td>\n",
       "      <td>20.31.24</td>\n",
       "      <td>30.0.80</td>\n",
       "      <td>20.31.34</td>\n",
       "      <td>20.31.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>27.1.20</td>\n",
       "      <td>20.31.8</td>\n",
       "      <td>20.31.8</td>\n",
       "      <td>20.31.8</td>\n",
       "      <td>30.0.24</td>\n",
       "      <td>20.31.8</td>\n",
       "      <td>20.31.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>16.1.25</td>\n",
       "      <td>28.1.38</td>\n",
       "      <td>28.1.38</td>\n",
       "      <td>28.1.38</td>\n",
       "      <td>16.1.25</td>\n",
       "      <td>20.31.38</td>\n",
       "      <td>20.31.38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>30.0.80</td>\n",
       "      <td>19.1.19</td>\n",
       "      <td>19.1.19</td>\n",
       "      <td>19.1.19</td>\n",
       "      <td>30.0.14</td>\n",
       "      <td>20.31.62</td>\n",
       "      <td>20.31.62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>23.0.0</td>\n",
       "      <td>27.1.18</td>\n",
       "      <td>27.1.18</td>\n",
       "      <td>27.1.18</td>\n",
       "      <td>23.0.0</td>\n",
       "      <td>27.1.9</td>\n",
       "      <td>27.1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>13.0.10</td>\n",
       "      <td>24.0.2</td>\n",
       "      <td>13.0.10</td>\n",
       "      <td>24.0.2</td>\n",
       "      <td>30.0.80</td>\n",
       "      <td>25.1.6</td>\n",
       "      <td>25.1.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>28.1.36</td>\n",
       "      <td>20.31.44</td>\n",
       "      <td>28.1.36</td>\n",
       "      <td>20.31.44</td>\n",
       "      <td>25.0.0</td>\n",
       "      <td>18.4.5</td>\n",
       "      <td>18.4.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>30.0.80</td>\n",
       "      <td>28.1.11</td>\n",
       "      <td>30.0.80</td>\n",
       "      <td>28.1.11</td>\n",
       "      <td>30.0.80</td>\n",
       "      <td>23.1.53</td>\n",
       "      <td>30.0.80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1523</th>\n",
       "      <td>18.4.14</td>\n",
       "      <td>3.0.0</td>\n",
       "      <td>3.0.0</td>\n",
       "      <td>3.0.0</td>\n",
       "      <td>18.4.12</td>\n",
       "      <td>29.0.3</td>\n",
       "      <td>29.0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1524</th>\n",
       "      <td>12.0.0</td>\n",
       "      <td>12.0.0</td>\n",
       "      <td>12.0.0</td>\n",
       "      <td>12.0.0</td>\n",
       "      <td>13.0.10</td>\n",
       "      <td>30.0.79</td>\n",
       "      <td>13.0.10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>20.31.25</td>\n",
       "      <td>5.0.0</td>\n",
       "      <td>20.31.25</td>\n",
       "      <td>5.0.0</td>\n",
       "      <td>22.0.0</td>\n",
       "      <td>18.4.26</td>\n",
       "      <td>18.4.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>30.0.21</td>\n",
       "      <td>23.0.3</td>\n",
       "      <td>23.0.3</td>\n",
       "      <td>23.0.3</td>\n",
       "      <td>27.1.23</td>\n",
       "      <td>23.0.3</td>\n",
       "      <td>23.0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1527</th>\n",
       "      <td>30.0.105</td>\n",
       "      <td>27.1.17</td>\n",
       "      <td>27.1.17</td>\n",
       "      <td>27.1.17</td>\n",
       "      <td>30.0.105</td>\n",
       "      <td>16.1.12</td>\n",
       "      <td>16.1.12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1528</th>\n",
       "      <td>23.0.1</td>\n",
       "      <td>16.1.20</td>\n",
       "      <td>23.0.1</td>\n",
       "      <td>16.1.20</td>\n",
       "      <td>22.0.0</td>\n",
       "      <td>23.1.60</td>\n",
       "      <td>23.1.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1529</th>\n",
       "      <td>16.31.33</td>\n",
       "      <td>28.1.42</td>\n",
       "      <td>16.31.33</td>\n",
       "      <td>28.1.42</td>\n",
       "      <td>30.0.80</td>\n",
       "      <td>28.1.42</td>\n",
       "      <td>28.1.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1530</th>\n",
       "      <td>30.0.99</td>\n",
       "      <td>27.1.11</td>\n",
       "      <td>30.0.99</td>\n",
       "      <td>27.1.11</td>\n",
       "      <td>30.0.21</td>\n",
       "      <td>27.1.9</td>\n",
       "      <td>27.1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1531</th>\n",
       "      <td>30.0.97</td>\n",
       "      <td>24.0.2</td>\n",
       "      <td>24.0.2</td>\n",
       "      <td>24.0.2</td>\n",
       "      <td>30.0.97</td>\n",
       "      <td>30.0.90</td>\n",
       "      <td>30.0.90</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1532</th>\n",
       "      <td>30.0.36</td>\n",
       "      <td>26.0.3</td>\n",
       "      <td>26.0.3</td>\n",
       "      <td>26.0.3</td>\n",
       "      <td>30.0.18</td>\n",
       "      <td>23.1.26</td>\n",
       "      <td>23.1.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1533</th>\n",
       "      <td>17.0.0</td>\n",
       "      <td>16.1.5</td>\n",
       "      <td>16.1.5</td>\n",
       "      <td>16.1.5</td>\n",
       "      <td>1.0.0</td>\n",
       "      <td>16.1.4</td>\n",
       "      <td>16.1.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1534</th>\n",
       "      <td>30.0.3</td>\n",
       "      <td>15.16.42</td>\n",
       "      <td>15.16.42</td>\n",
       "      <td>15.16.42</td>\n",
       "      <td>0.0.0</td>\n",
       "      <td>30.0.79</td>\n",
       "      <td>30.0.79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1535</th>\n",
       "      <td>0.0.0</td>\n",
       "      <td>23.1.26</td>\n",
       "      <td>0.0.0</td>\n",
       "      <td>23.1.26</td>\n",
       "      <td>0.0.0</td>\n",
       "      <td>23.1.26</td>\n",
       "      <td>23.1.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1536</th>\n",
       "      <td>16.31.33</td>\n",
       "      <td>14.29.17</td>\n",
       "      <td>16.31.33</td>\n",
       "      <td>14.29.17</td>\n",
       "      <td>16.31.33</td>\n",
       "      <td>14.29.17</td>\n",
       "      <td>16.31.33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1537</th>\n",
       "      <td>30.0.14</td>\n",
       "      <td>20.31.44</td>\n",
       "      <td>20.31.44</td>\n",
       "      <td>20.31.44</td>\n",
       "      <td>2.0.0</td>\n",
       "      <td>28.1.45</td>\n",
       "      <td>28.1.45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1538</th>\n",
       "      <td>30.0.27</td>\n",
       "      <td>23.1.47</td>\n",
       "      <td>23.1.47</td>\n",
       "      <td>23.1.47</td>\n",
       "      <td>30.0.27</td>\n",
       "      <td>28.1.40</td>\n",
       "      <td>28.1.40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1539</th>\n",
       "      <td>1.0.0</td>\n",
       "      <td>29.0.15</td>\n",
       "      <td>29.0.15</td>\n",
       "      <td>29.0.15</td>\n",
       "      <td>1.0.0</td>\n",
       "      <td>29.0.15</td>\n",
       "      <td>29.0.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1540</th>\n",
       "      <td>30.0.60</td>\n",
       "      <td>27.1.12</td>\n",
       "      <td>30.0.60</td>\n",
       "      <td>27.1.12</td>\n",
       "      <td>30.0.60</td>\n",
       "      <td>20.31.66</td>\n",
       "      <td>30.0.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1541</th>\n",
       "      <td>23.1.42</td>\n",
       "      <td>20.31.30</td>\n",
       "      <td>20.31.30</td>\n",
       "      <td>20.31.30</td>\n",
       "      <td>23.1.42</td>\n",
       "      <td>18.4.8</td>\n",
       "      <td>18.4.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1542</th>\n",
       "      <td>30.0.80</td>\n",
       "      <td>13.0.2</td>\n",
       "      <td>13.0.2</td>\n",
       "      <td>13.0.2</td>\n",
       "      <td>1.0.0</td>\n",
       "      <td>13.0.2</td>\n",
       "      <td>13.0.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1543</th>\n",
       "      <td>27.1.21</td>\n",
       "      <td>23.1.19</td>\n",
       "      <td>23.1.19</td>\n",
       "      <td>23.1.19</td>\n",
       "      <td>15.0.0</td>\n",
       "      <td>19.5.34</td>\n",
       "      <td>19.5.34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1544</th>\n",
       "      <td>30.0.40</td>\n",
       "      <td>20.31.30</td>\n",
       "      <td>20.31.30</td>\n",
       "      <td>20.31.30</td>\n",
       "      <td>13.0.0</td>\n",
       "      <td>20.31.2</td>\n",
       "      <td>20.31.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1545</th>\n",
       "      <td>27.1.20</td>\n",
       "      <td>20.31.42</td>\n",
       "      <td>20.31.42</td>\n",
       "      <td>20.31.42</td>\n",
       "      <td>28.1.24</td>\n",
       "      <td>20.31.42</td>\n",
       "      <td>20.31.42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1546</th>\n",
       "      <td>27.1.21</td>\n",
       "      <td>29.0.25</td>\n",
       "      <td>27.1.21</td>\n",
       "      <td>29.0.25</td>\n",
       "      <td>30.0.18</td>\n",
       "      <td>29.0.25</td>\n",
       "      <td>29.0.25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1547</th>\n",
       "      <td>16.31.33</td>\n",
       "      <td>20.31.30</td>\n",
       "      <td>20.31.30</td>\n",
       "      <td>20.31.30</td>\n",
       "      <td>0.0.0</td>\n",
       "      <td>20.31.30</td>\n",
       "      <td>20.31.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1548</th>\n",
       "      <td>27.0.3</td>\n",
       "      <td>16.1.18</td>\n",
       "      <td>16.1.18</td>\n",
       "      <td>16.1.18</td>\n",
       "      <td>30.0.80</td>\n",
       "      <td>23.1.46</td>\n",
       "      <td>23.1.46</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1549</th>\n",
       "      <td>28.1.23</td>\n",
       "      <td>28.1.31</td>\n",
       "      <td>28.1.31</td>\n",
       "      <td>28.1.31</td>\n",
       "      <td>22.0.0</td>\n",
       "      <td>20.31.39</td>\n",
       "      <td>20.31.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1550</th>\n",
       "      <td>14.0.9</td>\n",
       "      <td>20.31.72</td>\n",
       "      <td>14.0.9</td>\n",
       "      <td>20.31.72</td>\n",
       "      <td>30.0.14</td>\n",
       "      <td>20.31.72</td>\n",
       "      <td>20.31.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1551</th>\n",
       "      <td>28.1.40</td>\n",
       "      <td>27.1.12</td>\n",
       "      <td>27.1.12</td>\n",
       "      <td>27.1.12</td>\n",
       "      <td>0.0.0</td>\n",
       "      <td>20.31.30</td>\n",
       "      <td>20.31.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1552</th>\n",
       "      <td>1.0.0</td>\n",
       "      <td>29.0.12</td>\n",
       "      <td>29.0.12</td>\n",
       "      <td>29.0.12</td>\n",
       "      <td>0.0.0</td>\n",
       "      <td>28.1.6</td>\n",
       "      <td>28.1.6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1553 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     file_id_0 file_id_1 file_id_2 file_id_3 file_id_4 file_id_5 file_id_6\n",
       "0      28.1.30  20.31.60  20.31.60  20.31.60   28.1.30  20.31.42  20.31.42\n",
       "1      27.1.21   29.0.11   29.0.11   29.0.11    17.0.0   30.0.74   30.0.74\n",
       "2      15.1.19   27.1.12   15.1.19   27.1.12   30.0.47   27.1.12   27.1.12\n",
       "3       29.0.7    28.1.6    28.1.6    28.1.6    29.0.7    28.1.6    28.1.6\n",
       "4       16.0.0   27.1.11   27.1.11   27.1.11    16.0.0   23.1.63   23.1.63\n",
       "5      28.1.23   23.1.19   23.1.19   23.1.19    17.0.0  14.29.17  14.29.17\n",
       "6      30.0.80   27.1.11   27.1.11   27.1.11   30.0.24   27.1.11   27.1.11\n",
       "7        9.0.0   30.0.84   30.0.84   30.0.84     9.0.0   30.0.84   30.0.84\n",
       "8      16.1.11     8.0.0   16.1.11     8.0.0     2.0.0     8.0.0     8.0.0\n",
       "9      30.0.80   30.0.61   30.0.61   30.0.61    15.0.0  30.0.104  30.0.104\n",
       "10       1.0.0     6.0.0     6.0.0     6.0.0     1.0.0   23.1.57   23.1.57\n",
       "11     30.0.92    18.4.9    18.4.9    18.4.9   30.0.75   15.1.20   15.1.20\n",
       "12     29.0.15   23.1.53   23.1.53   23.1.53  30.0.105  20.31.44  20.31.44\n",
       "13     27.1.35   23.1.21   23.1.21   23.1.21    22.0.0   23.1.23   23.1.23\n",
       "14     30.0.35   27.1.17   27.1.17   27.1.17    14.0.1   27.1.10    14.0.1\n",
       "15     28.1.49   16.1.16   28.1.49   16.1.16   30.0.80   16.1.16   16.1.16\n",
       "16     23.1.64   19.5.32   23.1.64   19.5.32   23.1.64   23.1.53   23.1.53\n",
       "17     30.0.80   23.1.37   30.0.80   23.1.37     1.0.0   23.1.37   23.1.37\n",
       "18     23.1.12    18.4.9   23.1.12    18.4.9     0.0.0    18.4.9    18.4.9\n",
       "19       0.0.0   30.0.55   30.0.55   30.0.55   13.0.20   30.0.32   13.0.20\n",
       "20       0.0.0   23.1.29   23.1.29   23.1.29     0.0.0    23.0.3    23.0.3\n",
       "21    16.31.33   23.1.53  16.31.33   23.1.53   30.0.36   23.1.53   23.1.53\n",
       "22     30.0.80  20.31.24  20.31.24  20.31.24   30.0.80  20.31.34  20.31.34\n",
       "23     27.1.20   20.31.8   20.31.8   20.31.8   30.0.24   20.31.8   20.31.8\n",
       "24     16.1.25   28.1.38   28.1.38   28.1.38   16.1.25  20.31.38  20.31.38\n",
       "25     30.0.80   19.1.19   19.1.19   19.1.19   30.0.14  20.31.62  20.31.62\n",
       "26      23.0.0   27.1.18   27.1.18   27.1.18    23.0.0    27.1.9    27.1.9\n",
       "27     13.0.10    24.0.2   13.0.10    24.0.2   30.0.80    25.1.6    25.1.6\n",
       "28     28.1.36  20.31.44   28.1.36  20.31.44    25.0.0    18.4.5    18.4.5\n",
       "29     30.0.80   28.1.11   30.0.80   28.1.11   30.0.80   23.1.53   30.0.80\n",
       "...        ...       ...       ...       ...       ...       ...       ...\n",
       "1523   18.4.14     3.0.0     3.0.0     3.0.0   18.4.12    29.0.3    29.0.3\n",
       "1524    12.0.0    12.0.0    12.0.0    12.0.0   13.0.10   30.0.79   13.0.10\n",
       "1525  20.31.25     5.0.0  20.31.25     5.0.0    22.0.0   18.4.26   18.4.26\n",
       "1526   30.0.21    23.0.3    23.0.3    23.0.3   27.1.23    23.0.3    23.0.3\n",
       "1527  30.0.105   27.1.17   27.1.17   27.1.17  30.0.105   16.1.12   16.1.12\n",
       "1528    23.0.1   16.1.20    23.0.1   16.1.20    22.0.0   23.1.60   23.1.60\n",
       "1529  16.31.33   28.1.42  16.31.33   28.1.42   30.0.80   28.1.42   28.1.42\n",
       "1530   30.0.99   27.1.11   30.0.99   27.1.11   30.0.21    27.1.9    27.1.9\n",
       "1531   30.0.97    24.0.2    24.0.2    24.0.2   30.0.97   30.0.90   30.0.90\n",
       "1532   30.0.36    26.0.3    26.0.3    26.0.3   30.0.18   23.1.26   23.1.26\n",
       "1533    17.0.0    16.1.5    16.1.5    16.1.5     1.0.0    16.1.4    16.1.4\n",
       "1534    30.0.3  15.16.42  15.16.42  15.16.42     0.0.0   30.0.79   30.0.79\n",
       "1535     0.0.0   23.1.26     0.0.0   23.1.26     0.0.0   23.1.26   23.1.26\n",
       "1536  16.31.33  14.29.17  16.31.33  14.29.17  16.31.33  14.29.17  16.31.33\n",
       "1537   30.0.14  20.31.44  20.31.44  20.31.44     2.0.0   28.1.45   28.1.45\n",
       "1538   30.0.27   23.1.47   23.1.47   23.1.47   30.0.27   28.1.40   28.1.40\n",
       "1539     1.0.0   29.0.15   29.0.15   29.0.15     1.0.0   29.0.15   29.0.15\n",
       "1540   30.0.60   27.1.12   30.0.60   27.1.12   30.0.60  20.31.66   30.0.60\n",
       "1541   23.1.42  20.31.30  20.31.30  20.31.30   23.1.42    18.4.8    18.4.8\n",
       "1542   30.0.80    13.0.2    13.0.2    13.0.2     1.0.0    13.0.2    13.0.2\n",
       "1543   27.1.21   23.1.19   23.1.19   23.1.19    15.0.0   19.5.34   19.5.34\n",
       "1544   30.0.40  20.31.30  20.31.30  20.31.30    13.0.0   20.31.2   20.31.2\n",
       "1545   27.1.20  20.31.42  20.31.42  20.31.42   28.1.24  20.31.42  20.31.42\n",
       "1546   27.1.21   29.0.25   27.1.21   29.0.25   30.0.18   29.0.25   29.0.25\n",
       "1547  16.31.33  20.31.30  20.31.30  20.31.30     0.0.0  20.31.30  20.31.30\n",
       "1548    27.0.3   16.1.18   16.1.18   16.1.18   30.0.80   23.1.46   23.1.46\n",
       "1549   28.1.23   28.1.31   28.1.31   28.1.31    22.0.0  20.31.39  20.31.39\n",
       "1550    14.0.9  20.31.72    14.0.9  20.31.72   30.0.14  20.31.72  20.31.72\n",
       "1551   28.1.40   27.1.12   27.1.12   27.1.12     0.0.0  20.31.30  20.31.30\n",
       "1552     1.0.0   29.0.12   29.0.12   29.0.12     0.0.0    28.1.6    28.1.6\n",
       "\n",
       "[1553 rows x 7 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m_strat[[\"file_id_\"+str(x) for x in range(7)]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "m_strat = pandas.read_csv(os.path.join(data_path,\"Cor_distance_AllofSourceData_NHLOne_NHLTwo_test6_seq_pred.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "testing strat 0\n",
      "1a70A\n",
      "1hh8A\n",
      "1gmiA\n",
      "1atzA\n",
      "3borA\n",
      "1fcyA\n",
      "1jfuA\n",
      "1g9oA\n",
      "1tqhA\n",
      "1mugA\n",
      "testing strat 1\n",
      "1a70A\n",
      "1hh8A\n",
      "1gmiA\n",
      "1atzA\n",
      "3borA\n",
      "1fcyA\n",
      "1jfuA\n",
      "1g9oA\n",
      "1tqhA\n",
      "1mugA\n",
      "testing strat 2\n",
      "1a70A\n",
      "1hh8A\n",
      "1gmiA\n",
      "1atzA\n",
      "3borA\n",
      "1fcyA\n",
      "1jfuA\n",
      "1g9oA\n",
      "1tqhA\n",
      "1mugA\n",
      "testing strat 3\n",
      "1a70A\n",
      "1hh8A\n",
      "1gmiA\n",
      "1atzA\n",
      "3borA\n",
      "1fcyA\n",
      "1jfuA\n",
      "1g9oA\n",
      "1tqhA\n",
      "1mugA\n",
      "testing strat 4\n",
      "1a70A\n",
      "1hh8A\n",
      "1gmiA\n",
      "1atzA\n",
      "3borA\n",
      "1fcyA\n",
      "1jfuA\n",
      "1g9oA\n",
      "1tqhA\n",
      "1mugA\n",
      "testing strat 5\n",
      "1a70A\n",
      "1hh8A\n",
      "1gmiA\n",
      "1atzA\n",
      "3borA\n",
      "1fcyA\n",
      "1jfuA\n",
      "1g9oA\n",
      "1tqhA\n",
      "1mugA\n",
      "testing strat 6\n",
      "1a70A\n",
      "1hh8A\n",
      "1gmiA\n",
      "1atzA\n",
      "3borA\n",
      "1fcyA\n",
      "1jfuA\n",
      "1g9oA\n",
      "1tqhA\n",
      "1mugA\n"
     ]
    }
   ],
   "source": [
    "strat_test=pandas.DataFrame()\n",
    "\n",
    "for strat in range(7):\n",
    "    print(\"testing strat \"+str(strat))\n",
    "    \n",
    "    cur_strat=m_strat.copy()\n",
    "    cur_strat[\"file_id\"]=cur_strat[\"file_id_\"+str(strat)]\n",
    "    \n",
    "    cur_strat=add_pred_condensed(cur_strat,dataset_5)\n",
    "    \n",
    "    pred_df=eval_test_data_reduced(cur_strat,dataset_5)\n",
    "    \n",
    "    pa=pred_accuracy(pred_df)\n",
    "    \n",
    "    pas=pred_accuracy_signal(pred_df)\n",
    "        \n",
    "    strat_name=pandas.DataFrame([strat]*len(pa))\n",
    "    strat_name.columns=['strat']\n",
    "    \n",
    "    strat_result=pandas.concat([strat_name,pa,pas],axis=1)\n",
    "    \n",
    "    strat_test=strat_test.append(strat_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x17edaa79668>"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAEiCAYAAADptCm5AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAa2klEQVR4nO3de5RU5Z3u8e9DK14i6hh6cpSLTc6gJyIXpcXxHkUFRwdnjCbgUXFMZKIyJlGyxEQZxgyZqCw9iUNWBC+jRoOoUdtIRBN0Jl6HS1BsiYrYSofoICgaBLn4O39UgUVT3b272VXVvfv5rMWi9rvf2vtXQD/18tbebykiMDOzzq9bpQswM7N0ONDNzDLCgW5mlhEOdDOzjHCgm5llhAPdzCwjdqrUiXv27Bk1NTWVOr2ZWae0YMGC9yKiuti+igV6TU0N8+fPr9Tpzcw6JUlvNbfPUy5mZhnhQDczywgHuplZRjjQzcwywoFuZpYRDnQzs4xIFOiSRkp6VdJSSROL7L9R0qL8r9ckfZB+qWZm1pJWr0OXVAVMA04CGoF5kuoi4pUtfSLiOwX9/wk4pAS1mplZC5LcWDQMWBoRywAkzQROB15ppv8Y4J/TKc/MrBOZvFcL+9aU/PRJplx6AcsLthvzbduRtD/QD5i746WZmVlbJAl0FWlr7nvrRgP3R8TmogeSxkmaL2n+ypUrk9ZoZmYJJAn0RqBPwXZvYEUzfUcDv2juQBExPSJqI6K2urro2jJmZtZOSQJ9HtBfUj9J3cmFdl3TTpIOBP4CeC7dEs3MLIlWAz0iNgHjgTnAEmBWRNRLukbSqIKuY4CZEdHcdIyZmZVQouVzI2I2MLtJ26Qm25PTK8vMzNrKd4qamWWEA93MLCMc6GZmGVGxr6CznIF3DGx23+Kxi8tYSTb4z9O6Mge6JeKgNOv4POViZpYRDnQzs4zwlIuZNctTbZ2LA93MOj2/8eR4ysXMLCMc6GZmGeFANzPLCAe6mVlGONDNzDLCgW5mlhEOdDOzjHCgm5llhAPdzCwjHOhmZhnhQDczy4hEgS5ppKRXJS2VNLGZPl+V9Iqkekn3pFummZm1ptXFuSRVAdOAk4BGYJ6kuoh4paBPf+BK4KiIeF/SX5aqYDMzKy7JCH0YsDQilkXEBmAmcHqTPhcC0yLifYCI+J90yzQzs9YkCfRewPKC7cZ8W6EDgAMkPSPpeUkj0yrQzMySSbIeuoq0RZHj9Ae+DPQGfifp4Ij4YJsDSeOAcQB9+/Ztc7FmZta8JCP0RqBPwXZvYEWRPg9HxMaIeBN4lVzAbyMipkdEbUTUVldXt7dmMzMrIskIfR7QX1I/4I/AaODsJn0eAsYA/yGpJ7kpmGVpFtpW/gYTM+tqWh2hR8QmYDwwB1gCzIqIeknXSBqV7zYHWCXpFeBJ4LsRsapURZuZ2fYSfadoRMwGZjdpm1TwOIDL8r/MzKwCfKeomVlGONDNzDLCgW5mlhEOdDOzjHCgm5llhAPdzCwjHOhmZhnhQDczy4hENxaZmdmOKcdyJA50s1KZvFcL+9aUr47WtFRnP6+K2pk40M0qwIvHtYPfeFrV+QO9ub9k/wWbWRfT+QPd0uMRUGbVTHy02X0Nu5axECspB3o5OCjT5T/PzPIbz47xZYtmZhnhEbrZDvCI0joSj9DNzDLCI/QuxiNKs+zqFIHuEDIza52nXMzMMqJTjNA7A/8vIl3+8zRru0QjdEkjJb0qaamkiUX2ny9ppaRF+V/fSL9UMzNrSasjdElVwDTgJKARmCepLiJeadL13ogYX4IazcwsgSQj9GHA0ohYFhEbgJnA6aUty8zM2ipJoPcClhdsN+bbmvqKpJck3S+pTyrVmZlZYkkCXUXaosn2I0BNRAwCfgPcUfRA0jhJ8yXNX7lyZdsqNTOzFiW5yqURKBxx9wZWFHaIiFUFmzOAa4sdKCKmA9MBamtrm74pmJl1eB35CqwkI/R5QH9J/SR1B0YDdYUdJO1bsDkKWJJeiWZmlkSrI/SI2CRpPDAHqAJui4h6SdcA8yOiDrhU0ihgE7AaOL+ENZuZWRGJbiyKiNnA7CZtkwoeXwlcmW5pZmbWFr7138wsIxzoZmYZ4UA3M8sIB7qZWUY40M3MMsKBbmaWEQ50M7OMcKCbmWWEA93MLCMc6GZmGeFANzPLCAe6mVlGONDNzDLCgW5mlhEOdDOzjHCgm5llhAPdzCwjHOhmZhnhQDczywgHuplZRiQKdEkjJb0qaamkiS30O1NSSKpNr0QzM0ui1UCXVAVMA04BDgLGSDqoSL8ewKXAC2kXaWZmrUsyQh8GLI2IZRGxAZgJnF6k3w+A64D1KdZnZmYJJQn0XsDygu3GfNtWkg4B+kTEr1KszczM2iBJoKtIW2zdKXUDbgQub/VA0jhJ8yXNX7lyZfIqzcysVUkCvRHoU7DdG1hRsN0DOBh4SlID8NdAXbEPRiNiekTURkRtdXV1+6s2M7PtJAn0eUB/Sf0kdQdGA3VbdkbEmojoGRE1EVEDPA+Mioj5JanYzMyKajXQI2ITMB6YAywBZkVEvaRrJI0qdYFmZpbMTkk6RcRsYHaTtknN9P3yjpdlZmZt5TtFzcwywoFuZpYRDnQzs4xwoJuZZYQD3cwsIxzoZmYZ4UA3M8sIB7qZWUY40M3MMsKBbmaWEQ50M7OMcKCbmWWEA93MLCMc6GZmGeFANzPLCAe6mVlGONDNzDLCgW5mlhEOdDOzjHCgm5llhAPdzCwjEgW6pJGSXpW0VNLEIvu/KWmxpEWSnpZ0UPqlmplZS1oNdElVwDTgFOAgYEyRwL4nIgZGxBDgOuCG1Cs1M7MWJRmhDwOWRsSyiNgAzAROL+wQER8WbH4OiPRKNDOzJHZK0KcXsLxguxE4vGknSZcAlwHdgROKHUjSOGAcQN++fdtaq5mZtSDJCF1F2rYbgUfEtIj438AVwFXFDhQR0yOiNiJqq6ur21apmZm1KEmgNwJ9CrZ7Ayta6D8T+LsdKcrMzNouSaDPA/pL6iepOzAaqCvsIKl/weapwOvplWhmZkm0OoceEZskjQfmAFXAbRFRL+kaYH5E1AHjJZ0IbATeB8aWsmgzM9tekg9FiYjZwOwmbZMKHn8r5brMzKyNfKeomVlGONDNzDLCgW5mlhEOdDOzjHCgm5llhAPdzCwjHOhmZhnhQDczywgHuplZRjjQzcwywoFuZpYRDnQzs4xwoJuZZYQD3cwsIxzoZmYZ4UA3M8sIB7qZWUY40M3MMsKBbmaWEQ50M7OMSBTokkZKelXSUkkTi+y/TNIrkl6S9FtJ+6dfqpmZtaTVQJdUBUwDTgEOAsZIOqhJt98DtRExCLgfuC7tQs3MrGVJRujDgKURsSwiNgAzgdMLO0TEkxHxcX7zeaB3umWamVlrkgR6L2B5wXZjvq05Xwd+vSNFmZlZ2+2UoI+KtEXRjtI5QC1wXDP7xwHjAPr27ZuwRDMzSyLJCL0R6FOw3RtY0bSTpBOB7wOjIuKTYgeKiOkRURsRtdXV1e2p18zMmpEk0OcB/SX1k9QdGA3UFXaQdAhwM7kw/5/0yzQzs9a0GugRsQkYD8wBlgCzIqJe0jWSRuW7XQ/sAdwnaZGkumYOZ2ZmJZJkDp2ImA3MbtI2qeDxiSnXZWZmbeQ7Rc3MMiLRCL1cNm7cSGNjI+vXr9+mfcaofZt9zhLNKtr+/3Zq/qUtWbKkfQW2oD01Qq7OIFi+bjkz3p7BR5s/Sr02M+saOlSgNzY20qNHD2pqapA+u1pyY+MHzT7nS92KXVUJn3bv3vxzen6p/UU2oz01Qq7OiGCfj/bhQi7khjdvSL02M+saOtSUy/r16/n85z+/TZh3BZLo3qM7fXbr03pnM7NmdKhAB7pcmG8hCRW9h8vMLJkOF+idyc9+9jPuvPNOAK7+zsU88ejDFa7IzLqyDjWH3lTNxEdTPd6jE2pSPd43v/nNVI9nZrYjPEIv4s4772TQoEEMHjyYc889l7feeovhw4czaNAghg8fzttvvw3A5MmTmTp1aoWrNTPL6dAj9Eqor69nypQpPPPMM/Ts2ZPVq1czduxYzjvvPMaOHcttt93GpZdeykMPPVTpUs3MtuERehNz587lzDPPpGfPngDss88+PPfcc5x99tkAnHvuuTz99NOVLNHMrCgHehMR0eqVNl31Shwz69gc6E0MHz6cWbNmsWrVKgBWr17NkUceycyZMwG4++67OfrooytZoplZUZ5Db2LAgAF8//vf57jjjqOqqopDDjmEn/zkJ1xwwQVcf/31VFdXc/vtt1e6TDOz7XToQG/40akAvNTCbfWDur1ZtL2+hVv/WzN27FjGjh27TdvcuXO36zd58uStj39w40/bfT4zszR4ysXMLCMc6GZmGeFANzPLCAe6mVlGONDNzDLCgW5mlhGJAl3SSEmvSloqaWKR/cdKWihpk6Qz0y+zfBoaGjj44IMT9y9coGva1B/y/O+eKlFlZmYta/U6dElVwDTgJKARmCepLiJeKej2NnA+MCHV6ibvBcCgdjx1QJG2+vHP7lA5rblkwvdKenwzs5YkGaEPA5ZGxLKI2ADMBE4v7BARDRHxEvBpCWosu82bN3PhhRcyYMAATj75ZNatW8cbb7zByJEjGTp0KMcccwx/+MMftnveli+5+OjDNYw67jAa3ngdgDEXX8mMu39Z7pdhZl1MkkDvBSwv2G7Mt2XW66+/ziWXXEJ9fT177703DzzwAOPGjeOmm25iwYIFTJ06lYsvvrjZ5/fYcy+u/MF1XH3ZJfz64Qd4f82HXPh/zyjjKzCzrijJrf/FlhaM9pxM0jhgHEDfvn3bc4iy6NevH0OGDAFg6NChNDQ08Oyzz3LWWWdt7fPJJ5+0eIwjjj2eJx59mH+76ru8/MQ9Ja3XzAySBXojUPh19L2BFe05WURMB6YD1NbWtutNoRx22WWXrY+rqqp499132XvvvVm0aFHiY3z66acse/1Vdtl1N1Z/sIbe+32hFKWamW2VZMplHtBfUj9J3YHRQF1py+pY9txzT/r168d9990H5NZMf/HFF1t8zl0zfsoX+x/Itf8+gwsu/xc2btxYjlLNrAtrNdAjYhMwHpgDLAFmRUS9pGskjQKQdJikRuAs4GZJ9aUsuhLuvvtubr31VgYPHsyAAQN4+OGHm+3bsGwpD868i8uv/gGHHn4kxx5+KP/641vKWK2ZdUWJls+NiNnA7CZtkwoezyM3FZOuyWuA8i6fW1NTw8svv7x1e8KEz67EfOyxx7YvMb+E7kuNH2yzhO5DT76w9fENky9vVy1mZm3hO0XNzDLCgW5mlhEOdDOzjHCgm5llhAPdzCwjHOhmZhnhQE+gpqaG9957b4eP07B8Bfc8+OsUKjIz216i69ArZeAdA1M93sxTZ6Z6vLbaEuhn//0pFa3DzLLJI/Qm1q5dy6mnnsrgwYM5+OCDuffeewG46aabOPTQQxk4cODWpXPXrl3LBRdcwGGHHcZXRx7Lk3Ny915t3ryZG/71as4+9QTOPOkobr7rfgAm/vAn/O6/FzHkpNHcOP3nlXmBZpZZDvQmHnvsMfbbbz9efPFFXn75ZUaOHAlAz549WbhwIRdddNHWbyiaMmUKJ5xwAvPmzeOWex/hhimT+PjjtTw48y726LEX9zw6l3t+NZcZ9zzIm2//kR9971KOGTaERU/M5DvjzqnkyzSzDOrQUy6VMHDgQCZMmMAVV1zBaaedxjHHHAPAGWfk1jMfOnQov/xl7ssqHn/8cerq6pg6dSrrN25mwyfreeePjTz3X0/y2pJ6fjM7t97LJx+t4fU336b7zjtX5kWZWZfgQG/igAMOYMGCBcyePZsrr7ySk08+GfhsSd2qqio2bdoE5FZdfOCBBzjwwAO3WW8mIph4zbUc9eXhwGfrzTz17PxyvhQz62I85dLEihUr2H333TnnnHOYMGECCxcubLbviBEjuOmmm4jILe2+5OWXADjyuBO4767bti6Z+9obb7H243X02GN3Plr7celfhJl1SQ70JhYvXsywYcMYMmQIU6ZM4aqrrmq279VXX83GjRsZNGgQZww/gmlTpwBwxpjz+GL/Axl9ynGcMfwI/vGKKWzatIlBX+rPTlVVDD7xa/5Q1MxS16GnXBaPXQyUd/ncESNGMGLEiG3aGhoatj6ura3lqaeeAmC33Xbj5ptv3q7Gbt26cenESVw6cdJ2Nf521s3tqsvMrDUeoZuZZYQD3cwsIxzoZmYZ0eECfcsVI11NRBB0zdduZunoUIG+6667smrVqi4X6hHBho82sHzd8kqXYmadWIe6yqV37940NjaycuXKbdrffX9ds89ZopVF29/ZqfmX1m1l+u9j7akRcnUGwfJ1y5nx9ozU6zKzriNRoEsaCfwYqAJuiYgfNdm/C3AnMBRYBXwtIhraWszOO+9Mv379tms/ZeKjzT6nYdezi7Z/tV/fZp+z5XLINLWnRmi5TjOztmh1qCqpCpgGnAIcBIyRdFCTbl8H3o+IvwJuBK5Nu1AzM2tZkrmHYcDSiFgWERuAmcDpTfqcDtyRf3w/MFyS0ivTzMxakyTQewGFn9Y15tuK9omITcAa4PNpFGhmZsmotStKJJ0FjIiIb+S3zwWGRcQ/FfSpz/dpzG+/ke+zqsmxxgHj8psHAq+m9ULyegI7/l1xpec609UZ6uwMNYLrTFsp6tw/IqqL7UjyoWgj0Kdguzewopk+jZJ2AvYCVjc9UERMB6Ynqbg9JM2PiNpSHT8trjNdnaHOzlAjuM60lbvOJFMu84D+kvpJ6g6MBuqa9KkDxuYfnwnMja52MbmZWYW1OkKPiE2SxgNzyF22eFtE1Eu6BpgfEXXArcBdkpaSG5mPLmXRZma2vUTXoUfEbGB2k7ZJBY/XA2elW1q7lGw6J2WuM12doc7OUCO4zrSVtc5WPxQ1M7POoUOt5WJmZu3nQDczy4gOtThXW0n6P+TuUu0FBLnLKesiYklFC+uE8n+WvYAXIuLPBe0jI+KxylW2LUnDgIiIefklKEYCf8h/ztNhSbozIs6rdB0tkXQ0uTvDX46IxytdzxaSDgeWRMSHknYDJgKHAq8AP4yINRUtsAPptHPokq4AxpBbiqAx39yb3BU2M5suINYRSfqHiLi9A9RxKXAJsAQYAnwrIh7O71sYEYdWsr4tJP0zuTWFdgKeAA4HngJOBOZExJTKVfcZSU0v6xVwPDAXICJGlb2oIiT9d0QMyz++kNy/gQeBk4FHOsrPUP7GxcH5K+6mAx+TX2Ik335GRQvMk9RiHRHxy5LX0IkD/TVgQERsbNLeHaiPiP6VqSw5SW9HRMWXW5S0GDgiIv4sqYbcD8tdEfFjSb+PiEMqWmBevs4hwC7AO0DvglHbCxExqKIF5klaSG70eAu5/zkK+AX5y3kj4j8rV91nCv9uJc0D/iYiVkr6HPB8RAysbIU5kpZExJfyj7cZYEhaFBFDKlfdZyRtGZz9JXAk+Tdwcm/mT5XjjaczT7l8CuwHvNWkfd/8vg5B0kvN7QK+UM5aWlC1ZZolIhokfRm4X9L+5OrsKDZFxGbgY0lvRMSHABGxTlKH+TsHaoFvAd8HvhsRiySt6yhBXqCbpL8g91maImIlQESslbSpsqVt4+WC/82+KKk2IuZLOgDY2NqTyyUi/gFA0q+AgyLiT/ntfcmtWFtynTnQvw38VtLrfLZ4WF/gr4DxFatqe18ARgDvN2kX8Gz5yynqHUlDImIRQH6kfhpwG9AhRml5GyTtHhEfk1t7HwBJe9GB3sQj4lPgRkn35X9/l475s7YXsIDcv8WQ9L8i4h1Je9Cx3si/AfxY0lXk1kV5TtJycj/336hoZcXVbAnzvHeBA8px4k475QIgqRu5D3F6kfsH2AjMy4/iOgRJtwK3R8TTRfbdExHNf/tFmUjqTW70+06RfUdFxDMVKGs7knaJiE+KtPcE9o2I9L+5JAWSTgWOiojvVbqWJCTtDnwhIt6sdC2FJPUAvkjuzbExIt6tcElFSfp3oD+5abYgN9W2tHBBw5KduzMHuplZRyTp74Fj85v/FREPluW8DnQzs3TlP3/qHxG/yf+PpyoiPir1eX1jkZlZivKXgN4P3Jxv6gU8VI5zO9DNzNJ1CXAUsOUqrNfJXcpYcg50M7N0fZL//mUA8l/6U5a5bQe6mVm6/lPS94DdJJ0E3Ac8Uo4T+0NRM7MU5S+n/jq5JRRE7suBbinHt7g50M3MUiKpCrgjIs6pxPk95WJmlpL8TY3V+TWlyq4j3o5sZtaZNQDP5FfdXLulMSJuKPWJHehmZulakf/VDehRzhN7Dt3MrAQk7Unuy1hKfofoFp5DNzNLkaTa/Nr9LwGLJb0oaWhrz0vl3B6hm5mlJ/8dCJdExO/y20cDPy3HF7B4hG5mlq6PtoQ5QH7p7LJMu3iEbmaWIkk3Arvz2XroXyP3BTcPAETEwpKd24FuZpYeSU+2sDsi4oSSnduBbmaWDb4O3cwsBZLOiYifS7qs2H7fWGRm1nl8Lv97WW8mKuQpFzOzjPBli2ZmKZJ0naQ9Je0s6beS3pNUltUXHehmZuk6OSI+BE4DGoEDgO+W48QOdDOzdO2c//1U4BcRsbpcJ/aHomZm6XpE0hJgPXCRpOr845Lzh6JmZimStBswHjgW2AAsIvcVdH8q+bkd6GZm6ZE0C/gQuDvfNAbYOyK+WvJzO9DNzNIj6cWIGNxaWyn4Q1Ezs3T9XtJfb9mQdDjwTDlO7BG6mVmK8h+IHgi8nW/qCywBPiW3OFfJ1kV3oJuZpUjS/i3tj4i3SnZuB7qZWTZ4Dt3MLCMc6GZmGeFAty5L0rcl7d6O550vab9S1GS2Ixzo1pV9m9x3P25HUlULzzsfcKBbh+O1XKxLkPQ5YBbQG6gC7iMXyk9Kei8ijpf0Z+AGYARwuaQTgL8FdgOeBf4R+ApQC9wtaR1wRESsK/sLMivCV7lYlyDpK8DIiLgwv70X8CJQGxHv5dsC+FpEzMpv77NlpTxJdwGzIuIRSU8BEyJifgVeilmzPOViXcVi4ERJ10o6JiLWFOmzGXigYPt4SS9IWgycAAwoR6Fm7eUpF+sSIuI1SUOBvwH+TdLjRbqtj4jNAJJ2BX5KbgS/XNJkYNeyFWzWDh6hW5eQvyrl44j4OTAVOBT4iOa/0HdLeL8naQ/gzIJ9LT3PrGI8QreuYiBwvaRPgY3ARcARwK8l/Skiji/sHBEfSJpBbqqmAZhXsPs/gJ/5Q1HraPyhqJlZRnjKxcwsIxzoZmYZ4UA3M8sIB7qZWUY40M3MMsKBbmaWEQ50M7OMcKCbmWXE/wdDdSv9UszSkgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAEiCAYAAADptCm5AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAbSElEQVR4nO3dfXRV9Z3v8ffXgAYUsRXGUSJN7h2wQJ7AANMqCqKAxdE1RUfBB9BKWoujtYNLVERKS4epVKpgi/hMR6Ugd5Qqg14qOhbRm4AgQqqiRsyiVQiIGEF5+N4/zkk8HE6SE9jnITuf11oszt77d/b+nkA++eW39/5tc3dERKT1OyrTBYiISDAU6CIiIaFAFxEJCQW6iEhIKNBFREJCgS4iEhLtMnXgLl26eH5+fqYOLyLSKq1evXqbu3dNtC1jgZ6fn09lZWWmDi8i0iqZ2YeNbdOQi4hISDQb6Gb2sJl9YmZvNbLdzOxeM9tkZm+aWb/gyxQRkeYk00N/FBjRxPbzgR7RP+XA7468LBERaalmx9Dd/X/MLL+JJhcB8z0yKcxrZnaCmZ3s7n9taTF79+6lpqaGPXv2tPStrV5ubi55eXm0b98+06WISCsVxEnRbsBHMcs10XUtDvSamho6depEfn4+ZhZAaa2Du1NbW0tNTQ0FBQWZLkdEWqkgToomSt6EUziaWbmZVZpZ5datWw/ZvmfPHk488cQ2FeYAZsaJJ57YJn8zEZHgBBHoNcCpMct5wJZEDd19nruXuXtZ164JL6Nsc2Fer61+bhEJThCBvgS4Knq1yz8COw9n/PxI/eY3v+GLL74IbH+//OUvA9uXiEg6WHMPuDCzJ4HBQBfgY+BOoD2Au8+1SNdyDpErYb4Arnb3Zu8YKisr8/gbi6qqqujVq1fLPwVf36jUpUuXQ7bt37+fnJycFu3vuOOO4/PPP2/Re/bt20e7dod/WuJIPr+IZLeix4oa3bZ+7Pqk92Nmq929LNG2Znvo7j7a3U929/bunufuD7n7XHefG93u7j7B3f+3uxclE+ZHqq6ujpEjR1JSUkJhYSE/+9nP2LJlC0OGDGHIkCFAJJCnTJnCwIEDWbVqFfn5+Wzbtg2AyspKBg8eDMDnn3/O1VdfTVFREcXFxSxevJhJkyaxe/duSktLufzyy6murqawsLDh+DNnzmTq1KkADB48mNtuu42zzz6be+65h61btzJq1Cj69+9P//79WblyZaq/HCIiQAZv/T8Sy5Yt45RTTuG5554DYOfOnTzyyCOsWLGioYdeV1dHYWEh06ZNa3JfP//5z+ncuTPr10d+Qu7YsYNRo0YxZ84c1q5dC0B1dXWT+/j00095+eWXARgzZgw33XQTZ555Jps3b2b48OFUVVUdyccVEUlKqwz0oqIiJk6cyC233MIFF1zAoEGDDmmTk5PDqFGjmt3X8uXLWbBgQcPyN77xjRbXc+mllx60v40bNzYsf/bZZ+zatYtOnTq1eL8iIi3RKgO9Z8+erF69mqVLl3LrrbcybNiwQ9rk5uYeNG7erl07Dhw4AHDQ5YHu3uwVJrHvjX8/wLHHHtvw+sCBA6xatYoOHTq07EOJiByhVjk515YtW+jYsSNXXHEFEydOZM2aNXTq1Ildu3Y1+p78/HxWr14NwOLFixvWDxs2jDlz5jQs79ixA4D27duzd+9eAE466SQ++eQTamtr+fLLL3n22WcbPU78/uqHbUREUq1VBvr69esZMGAApaWlTJ8+ncmTJ1NeXs7555/fcFI03p133smNN97IoEGDDuq5T548mR07dlBYWEhJSQkrVqwAoLy8nOLiYi6//HLat2/fcIL1ggsu4Nvf/najtd17771UVlZSXFxM7969mTt3brAfXkSkEc1etpgqQV+2GAZt/fOLhFlWXLYoIiKtgwJdRCQkFOgiIiGhQBcRCQkFuohISCjQRURCQoF+BObOncv8+fMBGDduHE899VSGKxKRtiyrb/3Pn/RcoPurnjEy0P396Ec/CnR/IiJHQj30BObPn09xcTElJSVceeWVfPjhhwwdOpTi4mKGDh3K5s2bAZg6dSozZ87McLUiIhFZ3UPPhA0bNjB9+nRWrlxJly5d2L59O2PHjuWqq65i7NixPPzww9xwww08/fTTmS5VROQg6qHHefHFF7n44osb5lX/5je/yapVqxgzZgwAV155JX/+858zWaKISEIK9DjJTKerBzqLSDZSoMcZOnQoCxcupLa2FoDt27fz3e9+t+EhGI8//jhnnnlmJksUEUlIY+hx+vTpw+23387ZZ59NTk4Offv25d577+Waa67hrrvuomvXrjzyyCOZLlNEMqSpq++CvpKupbI60DP1xRk7dixjx449aN2LL754SLv6B0UDPProoymuSkSkaRpyEREJCQW6iEhIKNBFREJCgS4iEhIKdBGRkMjqq1xERFqVqZ0b31bQPeWHVw89TnV1NYWFhUm3j52ga8qUKSxfvjxVpYmINCm7e+hN/bQ7rP3tDHZ/caZNm5bS/YuINEU99AT279/P+PHj6dOnD8OGDWP37t289957jBgxgtNPP51Bgwbxl7/85ZD31T/kYufOnZx22mm8/fbbAIwePZoHHngg3R9DRNoYBXoC7777LhMmTGDDhg2ccMIJLF68mPLycmbPns3q1auZOXMmP/7xjxt9f+fOnZkzZw7jxo1jwYIF7Nixg/Hjx6fxE4hIW5TdQy4ZUlBQQGlpKQCnn3461dXVvPrqq1xyySUNbb788ssm93HeeeexaNEiJkyYwLp161Jar4gIKNATOuaYYxpe5+Tk8PHHH3PCCSewdu3apPdx4MABqqqq6NChA9u3bycvLy8VpYqINEhqyMXMRpjZ22a2ycwmJdje3cxWmNkbZvammX0v+FIz5/jjj6egoIBFixYBkTnTm+t1z5o1i169evHkk09yzTXXsHfv3nSUKiJtWLOBbmY5wH3A+UBvYLSZ9Y5rNhlY6O59gcuA3wZdaKY9/vjjPPTQQ5SUlNCnTx+eeeaZRtu+8847PPjgg/z6179m0KBBnHXWWfziF79IY7Ui0hYlM+QyANjk7u8DmNkC4CJgY0wbB46Pvu4MbAmkuhRfZphIfn4+b731VsPyxIkTG14vW7bskPaNTaFbVVXV8Pruu+8OtkgRkQSSCfRuwEcxyzXAwLg2U4EXzOxfgWOBcwOpTkREkpbMGHqiB2h63PJo4FF3zwO+B/zezA7Zt5mVm1mlmVVu3bq15dWKiEijkgn0GuDUmOU8Dh1S+QGwEMDdVwG5QJf4Hbn7PHcvc/eyrl27Hl7FIiKSUDKBXgH0MLMCMzuayEnPJXFtNgNDAcysF5FAVxdcRCSNmg10d98HXA88D1QRuZplg5lNM7MLo83+DRhvZuuAJ4Fx7h4/LCMiIimU1I1F7r4UWBq3bkrM643AGcGWJiIiLaG5XJKQn5/Ptm3bjng/1dXVPPHEEwFUJCJyqKy+9b/osaJA97d+7PpA99dS9YE+ZsyYjNYhIuGkHnqcuro6Ro4cSUlJCYWFhfzhD38AYPbs2fTr14+ioqKGqXPr6uq45ppr6N+/P3379m24e3T//v3cfPPN9O/fn+LiYu6//34AJk2axCuvvEJpaSmzZs3KzAcUkdBSoMdZtmwZp5xyCuvWreOtt95ixIgRAHTp0oU1a9Zw3XXXNTyhaPr06ZxzzjlUVFSwYsUKbr75Zurq6njooYfo3LkzFRUVVFRU8MADD/DBBx8wY8YMBg0axNq1a7npppsy+TFFJIQU6HGKiopYvnw5t9xyC6+88gqdO0eemvT9738f+Ho6XYAXXniBGTNmUFpayuDBg9mzZw+bN2/mhRdeYP78+ZSWljJw4EBqa2t59913M/WRRKSNyOox9Ezo2bMnq1evZunSpdx6660MGzYM+HpK3ZycHPbt2wdEZl1cvHgxp5122kH7cHdmz57N8OHDD1r/0ksvpf4DiEibpR56nC1bttCxY0euuOIKJk6cyJo1axptO3z4cGbPnk39JfdvvPFGw/rf/e53DVPmvvPOO9TV1dGpUyd27dqV+g8hIm2SAj3O+vXrGTBgAKWlpUyfPp3Jkyc32vaOO+5g7969FBcXU1hYyB133AHAtddeS+/evenXrx+FhYX88Ic/ZN++fRQXF9OuXTtKSkp0UlREAmeZuqGzrKzMKysrD1pXVVVFr169MlJPNmjrn1+kNcif9Fyj26pzG78kuaige6PbWnJJtZmtdveyRNvUQxcRCQkFuohISCjQRURCIusCva1O0thWP7eIBCerAj03N5fa2to2F27uTm1tLbm5uZkuRURasay6sSgvL4+amhra4uPpcnNzycvLy3QZItKKZVWgt2/fnoKCgkyXISLSKmXVkIuIiBw+BbqISEgo0EVEQkKBLiISEgp0EZGQUKCLiISEAl1EJCQU6CIiIaFAFxEJCQW6iEhIKNBFREJCgS4iEhIKdBGRkFCgi4iEhAJdRCQkFOgiIiGhQBcRCYmkAt3MRpjZ22a2ycwmNdLmX8xso5ltMLMngi1TRESa0+wj6MwsB7gPOA+oASrMbIm7b4xp0wO4FTjD3XeY2d+lqmAREUksmR76AGCTu7/v7l8BC4CL4tqMB+5z9x0A7v5JsGWKiEhzkgn0bsBHMcs10XWxegI9zWylmb1mZiOCKlBERJLT7JALYAnWeYL99AAGA3nAK2ZW6O6fHrQjs3KgHKB79+4tLlZERBqXTA+9Bjg1ZjkP2JKgzTPuvtfdPwDeJhLwB3H3ee5e5u5lXbt2PdyaRUQkgWQCvQLoYWYFZnY0cBmwJK7N08AQADPrQmQI5v0gCxURkaY1G+juvg+4HngeqAIWuvsGM5tmZhdGmz0P1JrZRmAFcLO716aqaBEROVQyY+i4+1Jgady6KTGvHfhp9I+IiGSA7hQVEQkJBbqISEgo0EVEQkKBLiISEgp0EZGQUKCLiISEAl1EJCQU6CIiIaFAFxEJCQW6iEhIKNBFREJCgS4iEhIKdBGRkFCgi4iEhAJdRCQkFOgiIiGhQBcRCQkFuohISCjQRURCQoEuIhISCnQRkZBQoIuIhIQCXUQkJBToIiIhoUAXEQkJBbqISEgo0EVEQkKBLiISEgp0EZGQUKCLiISEAl1EJCQU6CIiIdEu0wWISIZN7dzEtp3pq0OOWFKBbmYjgHuAHOBBd5/RSLuLgUVAf3evDKxKkZApeqyo0W3rx65PYyUSJs0GupnlAPcB5wE1QIWZLXH3jXHtOgE3AK+notBslz/puUa3Vc8YmcZKwkFfT5GWS2YMfQCwyd3fd/evgAXARQna/Rz4FbAnwPpERCRJyQy5dAM+ilmuAQbGNjCzvsCp7v6smU0MsD4JmHq+IuGVTKBbgnXesNHsKGAWMK7ZHZmVA+UA3bt3T65CEWkz1OE4MskEeg1wasxyHrAlZrkTUAi8ZGYAfw8sMbML40+Muvs8YB5AWVmZIyKSLF2N06xkxtArgB5mVmBmRwOXAUvqN7r7Tnfv4u757p4PvAYcEuYiIpJazfbQ3X2fmV0PPE/kssWH3X2DmU0DKt19SdN7OHL6NSxN1AMKrSa/h3LTWIikVFLXobv7UmBp3LopjbQdfORliYhIS+lO0XRQzzdY+nqKJKS5XEREQkI9dJEjoPM7kk3UQxcRCYnQ9tA1+ZGItDXqoYuIhIQCXUQkJBToIiIhoUAXEQkJBbqISEgo0EVEQkKBLiISEgp0EZGQUKCLiISEAl1EJCQU6CIiIaFAFxEJCQW6iEhIKNBFREJCgS4iEhIKdBGRkFCgi4iEROt/YlFjT4Av6J7eOkREMkw9dBGRkGj9PXRJCz2jtW1qLf/uraXOVFMPXUQkJBToIiIhoUAXEQkJBbqISEgo0EVEQkKBLiISErpsMcN0uVWw9PWUtkyBLpIqjd3FDLqTWVIiqSEXMxthZm+b2SYzm5Rg+0/NbKOZvWlmfzKzbwVfqoiINKXZQDezHOA+4HygNzDazHrHNXsDKHP3YuAp4FdBFyoiIk1Lpoc+ANjk7u+7+1fAAuCi2AbuvsLdv4guvgbkBVumiIg0J5lA7wZ8FLNcE13XmB8A/30kRYmISMslc1LUEqzzhA3NrgDKgLMb2V4OlAN0766TQiIiQUqmh14DnBqznAdsiW9kZucCtwMXuvuXiXbk7vPcvczdy7p27Xo49YqISCOSCfQKoIeZFZjZ0cBlwJLYBmbWF7ifSJh/EnyZIiLSnGYD3d33AdcDzwNVwEJ332Bm08zswmizu4DjgEVmttbMljSyOxERSZGkbixy96XA0rh1U2JenxtwXSIi0kKay0VEJCQU6CIiIaFAFxEJCQW6iEhIKNBFREJCgS4iEhIKdBGRkFCgi4iEhAJdRCQkFOgiIiGhQBcRCQkFuohISCjQRURCQoEuIhISCnQRkZBQoIuIhIQCXUQkJBToIiIhoUAXEQkJBbqISEgo0EVEQkKBLiISEgp0EZGQUKCLiISEAl1EJCQU6CIiIaFAFxEJCQW6iEhIKNBFREJCgS4iEhIKdBGRkFCgi4iEhAJdRCQkFOgiIiGRVKCb2Qgze9vMNpnZpATbjzGzP0S3v25m+UEXKiIiTWs20M0sB7gPOB/oDYw2s95xzX4A7HD3fwBmAf8RdKEiItK0ZHroA4BN7v6+u38FLAAuimtzEfBY9PVTwFAzs+DKFBGR5iQT6N2Aj2KWa6LrErZx933ATuDEIAoUEZHkmLs33cDsEmC4u18bXb4SGODu/xrTZkO0TU10+b1om9q4fZUD5dHF04C3g/ogUV2AbQHvMxVUZ7BaQ52toUZQnUFLRZ3fcveuiTa0S+LNNcCpMct5wJZG2tSYWTugM7A9fkfuPg+Yl0zFh8PMKt29LFX7D4rqDFZrqLM11AiqM2jprjOZIZcKoIeZFZjZ0cBlwJK4NkuAsdHXFwMvenNdfxERCVSzPXR332dm1wPPAznAw+6+wcymAZXuvgR4CPi9mW0i0jO/LJVFi4jIoZIZcsHdlwJL49ZNiXm9B7gk2NIOS8qGcwKmOoPVGupsDTWC6gxaWuts9qSoiIi0Drr1X0QkJBToIiIhkdQYerYys28TuUu1G+BELqdc4u5VGS2sFYp+LbsBr7v75zHrR7j7ssxVdjAzGwC4u1dEp6AYAfwlep4na5nZfHe/KtN1NMXMziRyZ/hb7v5CpuupZ2YDgSp3/8zMOgCTgH7ARuCX7r4zowVmkVY7hm5mtwCjiUxFUBNdnUfkCpsF7j4jU7Uly8yudvdHsqCOG4AJQBVQCtzo7s9Et61x936ZrK+emd1JZE6hdsD/BQYCLwHnAs+7+/TMVfc1M4u/rNeAIcCLAO5+YdqLSsDM/p+7D4i+Hk/k/8B/AcOAP2bL91D0xsWS6BV384AviE4xEl3//YwWGGVmTdbh7v8n5TW04kB/B+jj7nvj1h8NbHD3HpmpLHlmttndu2dBHeuB77j759GZMp8Cfu/u95jZG+7eN6MFRkXrLAWOAf4G5MX02l539+KMFhhlZmuI9B4fJPKbowFPEr2c191fzlx1X4v9tzWzCuB77r7VzI4FXnP3osxWGGFmVe7eK/r6oA6Gma1199LMVfc1M6vvnP0d8F2iP8CJ/DB/KR0/eFrzkMsB4BTgw7j1J0e3ZQUze7OxTcBJ6aylCTn1wyzuXm1mg4GnzOxbROrMFvvcfT/whZm95+6fAbj7bjPLmn9zoAy4EbgduNnd15rZ7mwJ8hhHmdk3iJxLM3ffCuDudWa2L7OlHeStmN9m15lZmbtXmllPYG9zb04Xd78awMyeBXq7+1+jyycTmbE25VpzoP8E+JOZvcvXk4d1B/4BuD5jVR3qJGA4sCNuvQGvpr+chP5mZqXuvhYg2lO/AHgYyIpeWtRXZtbR3b8ATq9faWadyaIf4u5+AJhlZouif39Mdn6vdQZWE/m/6Gb29+7+NzM7juz6QX4tcI+ZTSYyL8oqM/uIyPf9tRmtLLH8+jCP+hjomY4Dt9ohFwAzO4rISZxuRP4D1gAV0V5cVjCzh4BH3P3PCbY94e5jMlBWfB15RHq/f0uw7Qx3X5mBsg5hZse4+5cJ1ncBTnb39Rkoq1lmNhI4w91vy3QtyTCzjsBJ7v5BpmuJZWadgP9F5Idjjbt/nOGSEjKzOUAPIsNsTmSobVPshIYpO3ZrDnQRkWxkZv8MnBVd/B93/6+0HFeBLiISrOj5px7uvjz6G0+Ou+9K9XF1Y5GISICil4A+BdwfXdUNeDodx1agi4gEawJwBlB/Fda7RC5lTDkFuohIsL6MPn8ZgOhDf9Iytq1AFxEJ1stmdhvQwczOAxYBf0zHgXVSVEQkQNHLqX9AZAoFI/JwoAfT8RQ3BbqISEDMLAd4zN2vyMTxNeQiIhKQ6E2NXaNzSqVdNt6OLCLSmlUDK6OzbtbVr3T3u1N9YAW6iEiwtkT/HAV0SueBNYYuIpICZnY8kYexpPwO0XoaQxcRCZCZlUXn7n8TWG9m68zs9ObeF8ix1UMXEQlO9BkIE9z9lejymcBv0/EAFvXQRUSCtas+zAGiU2enZdhFPXQRkQCZ2SygI1/Ph34pkQfcLAZw9zUpO7YCXUQkOGa2oonN7u7npOzYCnQRkXDQdegiIgEwsyvc/T/N7KeJtuvGIhGR1uPY6N9pvZkoloZcRERCQpctiogEyMx+ZWbHm1l7M/uTmW0zs7TMvqhAFxEJ1jB3/wy4AKgBegI3p+PACnQRkWC1j/49EnjS3ben68A6KSoiEqw/mlkVsAe4zsy6Rl+nnE6KiogEyMw6ANcDZwFfAWuJPILuryk/tgJdRCQ4ZrYQ+Ax4PLpqNHCCu/9Lyo+tQBcRCY6ZrXP3kubWpYJOioqIBOsNM/vH+gUzGwisTMeB1UMXEQlQ9IToacDm6KruQBVwgMjkXCmbF12BLiISIDP7VlPb3f3DlB1bgS4iEg4aQxcRCQkFuohISCjQpc0ys5+YWcfDeN84MzslFTWJHAkFurRlPyHy7MdDmFlOE+8bByjQJetoLhdpE8zsWGAhkAfkAIuIhPIKM9vm7kPM7HPgbmA48G9mdg7wT0AH4FXgh8AooAx43Mx2A99x991p/0AiCegqF2kTzGwUMMLdx0eXOwPrgDJ33xZd58Cl7r4wuvzN+pnyzOz3wEJ3/6OZvQRMdPfKDHwUkUZpyEXaivXAuWb2H2Y2yN13JmizH1gcszzEzF43s/XAOUCfdBQqcrg05CJtgru/Y2anA98D/t3MXkjQbI+77wcws1zgt0R68B+Z2VQgN20FixwG9dClTYhelfKFu/8nMBPoB+yi8Qf61of3NjM7Drg4ZltT7xPJGPXQpa0oAu4yswPAXuA64DvAf5vZX919SGxjd//UzB4gMlRTDVTEbH4UmKuTopJtdFJURCQkNOQiIhISCnQRkZBQoIuIhIQCXUQkJBToIiIhoUAXEQkJBbqISEgo0EVEQuL/A7DVbzJt0WmJAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "psipred=strat_test[strat_test[\"strat\"]==0].copy()\n",
    "psipred[[\"thd_pred_rmse\",\"thd_pred_accuracy\"]]=psipred[[\"psipred_rmse\",\"psipred_accuracy\"]]\n",
    "psipred[[\"thd_pred_rmse_signal\",\"thd_pred_accuracy_signal\"]]=psipred[[\"psipred_rmse_signal\",\"psipred_accuracy_signal\"]]\n",
    "psipred[\"strat\"]=\"psipred\"\n",
    "\n",
    "psipred\n",
    "\n",
    "strat_test2=strat_test.append(psipred)\n",
    "strat_test2\n",
    "\n",
    "strat_test2.pivot(\"strat\", \"structure\", \"thd_pred_accuracy\").plot(kind='bar')\n",
    "plt.legend(loc=\"lower left\")\n",
    "\n",
    "\n",
    "strat_test2.pivot(\"strat\", \"structure\", \"thd_pred_accuracy_signal\").plot(kind='bar')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
