{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy\n",
    "import math\n",
    "import imageio\n",
    "import json\n",
    "import time\n",
    "\n",
    "from collections import Counter\n",
    "from matplotlib.pyplot import figure\n",
    "from matplotlib import animation\n",
    "\n",
    "import concurrent.futures\n",
    "import multiprocessing\n",
    "\n",
    "available_cpu=multiprocessing.cpu_count()-1\n",
    "\n",
    "\n",
    "pdb_list, length_dict, input_features = np.load(\"datasets/sample-input-features.npy\",allow_pickle=True)\n",
    "pdb_list_y, distance_maps_cb = np.load(\"datasets/sample-distance-maps-cb.npy\",encoding=\"latin1\",allow_pickle=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "## aaron's code to get raw features from pdb\n",
    "\n",
    "import numpy as np\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "from matplotlib.pyplot import figure\n",
    "%matplotlib inline\n",
    "import time\n",
    "import requests\n",
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "# ! pip install biopython\n",
    "# ! pip install nglview\n",
    "# ! jupyter-nbextension enable nglview --py --sys-prefix\n",
    "\n",
    "from Bio.PDB import *\n",
    "import nglview as nv\n",
    "import math\n",
    "import warnings\n",
    "from Bio.PDB.StructureBuilder import PDBConstructionWarning\n",
    "warnings.filterwarnings(\"ignore\", category=PDBConstructionWarning)\n",
    "\n",
    "# this is dangerous of course but uncomment it when you want to run the code\n",
    "# warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# aaron's code to make sure the folders are all in their proper place\n",
    "\n",
    "dirlocal = os.path.curdir\n",
    "data_path = os.path.join(dirlocal, 'Data/')\n",
    "protein_path = os.path.join(data_path, 'Proteins/')\n",
    "pdb_path = os.path.join(data_path, 'PDB/')\n",
    "fasta_path = os.path.join(data_path, 'fasta/')\n",
    "dssp_path = os.path.join(data_path, 'DSSP/')\n",
    "\n",
    "if not os.path.exists(data_path):\n",
    "    os.makedirs(data_path)\n",
    "if not os.path.exists(protein_path):\n",
    "    os.makedirs(protein_path)\n",
    "if not os.path.exists(pdb_path):\n",
    "    os.makedirs(pdb_path)\n",
    "ss_path = os.path.join(data_path, 'ss.txt')\n",
    "feature_path = os.path.join(data_path, 'sample-input-features.npy')\n",
    "distance_path = os.path.join(data_path, 'sample-distance-maps-cb.npy')\n",
    "full_feature_path = os.path.join(data_path, 'full-input-features.npy')\n",
    "full_distance_path = os.path.join(data_path, 'full-distance-maps-cb.npy')\n",
    "test_feature_path = os.path.join(data_path, 'testset-input-features.npy')\n",
    "test_distance_path = os.path.join(data_path, 'testset-distance-maps-cb.npy')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "###  aaron's functions: \n",
    "\n",
    "# gets dataframe containing torsion angles, peptides, chain, etc\n",
    "# from pdb file\n",
    "def get_torsion_angles(pdb_id, degrees=False):\n",
    "    pdb_struct = get_pdb_structure(pdb_id)\n",
    "    torsion_angles = []\n",
    "    for model in pdb_struct:\n",
    "        for chain in model:\n",
    "            polypeptides = PPBuilder().build_peptides(chain)\n",
    "            for poly_index, poly in enumerate(polypeptides):\n",
    "                #print(\"Model %s Chain %s\" % (str(model.id), str(chain.id)))\n",
    "                #print(\"(part %i of %i)\" % (poly_index+1, len(polypeptides)))\n",
    "                #print(\"length %i\" % (len(poly)))\n",
    "                #print(\"from %s%i\" % (poly[0].resname, poly[0].id[1]))\n",
    "                #print(\"to %s%i\" % (poly[-1].resname, poly[-1].id[1]))\n",
    "                phi_psi = poly.get_phi_psi_list()\n",
    "                for res_index, residue in enumerate(poly):\n",
    "                    res_name = \"%s%i\" % (residue.resname, residue.id[1])\n",
    "                    #print(res_name, tuple(math.degrees(b) for b in phi_psi[res_index] if b))\n",
    "                    deg = phi_psi[res_index]\n",
    "                    if degrees:\n",
    "                        deg = tuple(math.degrees(b) if b else None for b in deg)\n",
    "                    phi, psi = deg\n",
    "                    model_name, model_id = model.full_id\n",
    "                    torsion_angles.append([model_name, model_id, chain.id, residue.resname, residue.id[1], phi, psi])\n",
    "    return pd.DataFrame(torsion_angles, columns=['Model_Name', 'Model_ID', 'Chain', 'Residue_Name', 'Residue_ID', 'Phi', 'Psi'])\n",
    "\n",
    "\n",
    "# generates ramachandran plot from a given pdb file\n",
    "def ramachandran_plot(pdb_id, degrees=True):\n",
    "    df = get_torsion_angles(pdb_id, degrees)\n",
    "    x = df['Phi']\n",
    "    y = df['Psi']\n",
    "    # Generate plot\n",
    "    plt.plot(x, y, \".\")\n",
    "    plt.title('Ramachandran Plot')\n",
    "    if degrees:\n",
    "        plt.xlabel(f'$\\Phi$ Angle (Degrees)')\n",
    "        plt.xlim(-180, 180)\n",
    "        plt.ylabel(f'$\\Psi$ Angle (Radians)')\n",
    "        plt.ylim(-180, 180)\n",
    "        plt.show()\n",
    "    else:\n",
    "        plt.xlabel(f'$\\Phi$ Angle (Radians)')\n",
    "        plt.xlim(-math.pi, math.pi)\n",
    "        plt.ylabel(f'$\\Psi$ Angle (Radians)')\n",
    "        plt.ylim(-math.pi, math.pi)\n",
    "        plt.show()\n",
    "\n",
    "# helper function which generates a biopython structure\n",
    "# for display in display_protein()\n",
    "def get_pdb_structure(pdb_id):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    parser = PDBParser()\n",
    "    file_path = os.path.join(pdb_path, f\"{pdb_id}.pdb\")\n",
    "    try:\n",
    "        struct = parser.get_structure(pdb_id, file_path)\n",
    "    except Exception as e:\n",
    "        print(e)\n",
    "        return None\n",
    "    return struct\n",
    "\n",
    "\n",
    "#  downloads the protein in question from pdb\n",
    "def get_pdb_file(pdb_id):\n",
    "    if pdb_id+\".pdb\" in os.listdir(pdb_path):\n",
    "        return True\n",
    "\n",
    "    pdb_id = pdb_id.upper()\n",
    "    parser = PDBParser()\n",
    "    url = f'https://files.rcsb.org/download/{pdb_id}.pdb'\n",
    "    #print(url)\n",
    "    resp = requests.get(url)\n",
    "    try:\n",
    "        file_path = os.path.join(pdb_path, f'{pdb_id}.pdb')\n",
    "        if os.path.isfile(file_path) and not replace:\n",
    "            return True\n",
    "        #print(file_path)\n",
    "        file = open(file_path, \"wb\")\n",
    "        file.write(resp.content)\n",
    "        file.close()\n",
    "    except Exception as e: \n",
    "        print(e)\n",
    "        return False\n",
    "    return True\n",
    "\n",
    "# neato 3d viewer of a protein's structure\n",
    "def display_protein(pdb_id):\n",
    "    pdb_struct = get_pdb_structure(pdb_id)\n",
    "    view = nv.show_biopython(pdb_struct)\n",
    "    return view\n",
    "\n",
    "\n",
    "#  returns a dataframe with secondary structure, amino acid code, chain id, etc\n",
    "# deprecated\n",
    "def get_secondary_structure(pdb_id):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    indexes = get_pdb_ss_seq(pdb_id)\n",
    "    chain_df = []\n",
    "    for i, chain_ss in enumerate(indexes):\n",
    "        chain_id, seq, ss = chain_ss\n",
    "        #unknown_list = [i for i, c in enumerate(seq) if c == 'X']\n",
    "        #print(unknown_list)\n",
    "        seq_list = [c for c in seq]\n",
    "        # [f(x) if condition else g(x) for x in sequence]\n",
    "        ss_list = ['L' if c is ' ' else c for c in ss]\n",
    "        #print(seq_list + ss_list)\n",
    "        chain_df.append(pd.DataFrame({'Chain': chain_id, 'Amino_Acid':seq_list, 'Secondary_Structure':ss_list}))\n",
    "    return pd.concat(chain_df)\n",
    "\n",
    "# get matrix of pairwise carbon beta distances\n",
    "def get_cb_distances(full_pdb_id, partial=False):\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    get_pdb_file(pdb_id)\n",
    "    pdb_struct = get_pdb_structure(pdb_id)  # Returns the BioPython structure of the PDB file\n",
    "    ppb = PPBuilder()\n",
    "    chain_dist_matrix = {}\n",
    "    for model in pdb_struct:\n",
    "        for chain in model:\n",
    "            residues = []\n",
    "            for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                for residue in pp:\n",
    "                    residues.append(residue)\n",
    "            lng = len(residues)\n",
    "            dist_matrix = np.zeros((lng, lng), np.float)\n",
    "            for i in range(0, lng):\n",
    "                for j in range(i, lng):\n",
    "                    try:\n",
    "                        diff_vector  = residues[i]['C'].coord - residues[j]['C'].coord\n",
    "                    except Exception:\n",
    "                        count = 0\n",
    "                        sums = 0\n",
    "                        if j - 1 >= 0:\n",
    "                            count += 1\n",
    "                            sums += chain_dist_matrix[i][j - 1]\n",
    "                        if i - 1 >= 0:\n",
    "                            count += 1\n",
    "                            sums += chain_dist_matrix[i - 1][j]\n",
    "                        if count > 0:\n",
    "                            diff_vector = sums / count\n",
    "                        else:\n",
    "                            diff_vector = 0\n",
    "                    dist = np.sqrt(np.sum(diff_vector * diff_vector))\n",
    "                    dist_matrix[i][j] = dist\n",
    "                    dist_matrix[j][i] = dist\n",
    "                chain_dist_matrix[chain.id] = dist_matrix\n",
    "        \n",
    "    seq_bool = get_seq_alignment(full_pdb_id)\n",
    "    if partial:    \n",
    "        seq_bool = get_seq_alignment(full_pdb_id)\n",
    "        dist_matrix = chain_dist_matrix[full_pdb_id[4].upper()]\n",
    "        return pd.DataFrame(dist_matrix[seq_bool]).iloc[:, seq_bool]\n",
    "    else:\n",
    "        return chain_dist_matrix\n",
    "\n",
    "# gets alignment of given sequence on top of the original sequence\n",
    "# from a fasta file downloaded from the cloud repository\n",
    "def get_seq_alignment(full_pdb_id):\n",
    "    seq = ''\n",
    "    for record in SeqIO.parse(os.path.join(fasta_path, f'{full_pdb_id}.fasta'), \"fasta\"):\n",
    "        seq = record.seq\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    #print(pdb_id)\n",
    "    get_pdb_file(pdb_id)\n",
    "    struct = get_pdb_structure(pdb_id)\n",
    "    ppb = PPBuilder()\n",
    "    for model in struct:\n",
    "        for chain in model:\n",
    "            if chain.id == full_pdb_id[4]:\n",
    "                ground_seq = ''\n",
    "                for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                    ground_seq += pp.get_sequence()\n",
    "                alignments = pairwise2.align.globalxx(ground_seq, str(seq))\n",
    "                al_seq = alignments[0][1]\n",
    "                al_seq_bool = [False if a is '-' else True for a in al_seq]\n",
    "                return al_seq_bool\n",
    "\n",
    "# calls dssp api to get dssp file\n",
    "def pdb_id_to_dssp_file(pdb_id, replace=False):\n",
    "    pdb_id = pdb_id.upper()\n",
    "    rest_url = 'http://www.cmbi.umcn.nl/xssp/'\n",
    "    # Read the pdb id data into a variable\n",
    "    data = {'data': pdb_id}\n",
    "\n",
    "    # Send a request to the server to create hssp data from the pdb file data.\n",
    "    # If an error occurs, an exception is raised and the program exits. If the\n",
    "    # request is successful, the id of the job running on the server is\n",
    "    # returned.\n",
    "    url_create = f'{rest_url}api/create/pdb_id/dssp/'\n",
    "    r = requests.post(url_create, data=data)\n",
    "    r.raise_for_status()\n",
    "\n",
    "    job_id = json.loads(r.text)['id']\n",
    "    #print(f'Job submitted successfully. Id is: {job_id}')\n",
    "\n",
    "    # Loop until the job running on the server has finished, either successfully\n",
    "    # or due to an error.\n",
    "    ready = False\n",
    "    while not ready:\n",
    "        # Check the status of the running job. If an error occurs an exception\n",
    "        # is raised and the program exits. If the request is successful, the\n",
    "        # status is returned.\n",
    "        url_status = f'{rest_url}api/status/pdb_id/dssp/{job_id}/'\n",
    "        r = requests.get(url_status)\n",
    "        r.raise_for_status()\n",
    "\n",
    "        status = json.loads(r.text)['status']\n",
    "        #print(f'Job status is: {status}')\n",
    "\n",
    "        # If the status equals SUCCESS, exit out of the loop by changing the\n",
    "        # condition ready. This causes the code to drop into the `else` block\n",
    "        # below.\n",
    "        #\n",
    "        # If the status equals either FAILURE or REVOKED, an exception is raised\n",
    "        # containing the error message. The program exits.\n",
    "        #\n",
    "        # Otherwise, wait for five seconds and start at the beginning of the\n",
    "        # loop again.\n",
    "        if status == 'SUCCESS':\n",
    "            ready = True\n",
    "        elif status in ['FAILURE', 'REVOKED']:\n",
    "            raise Exception(json.loads(r.text)['message'])\n",
    "        else:\n",
    "            time.sleep(5)\n",
    "    else:\n",
    "        # Requests the result of the job. If an error occurs an exception is\n",
    "        # raised and the program exits. If the request is successful, the result\n",
    "        # is returned.\n",
    "        url_result = f'{rest_url}api/result/pdb_id/dssp/{job_id}/'\n",
    "        r = requests.get(url_result)\n",
    "        r.raise_for_status()\n",
    "        result = json.loads(r.text)['result']\n",
    "        try:\n",
    "            file_path = os.path.join(dssp_path, f'{pdb_id}.dssp')\n",
    "            if os.path.isfile(file_path) and not replace:\n",
    "                return True\n",
    "            #print(file_path)\n",
    "            file = open(file_path, \"w\")\n",
    "            file.write(result)\n",
    "            file.close()\n",
    "        except Exception as e: \n",
    "            print(e)\n",
    "            return False\n",
    "        # Return the result to the caller, which prints it to the screen.\n",
    "        return True\n",
    "\n",
    "# uses api call file to generate secondary structure, phi/psi, solvent-accessibility stuff\n",
    "def get_ground_truth_api(full_pdb_id, partial=True):\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    get_pdb_file(pdb_id)\n",
    "    struct = get_pdb_structure(pdb_id)\n",
    "    ppb = PPBuilder()\n",
    "    struct_info = []\n",
    "    seq_count = 0\n",
    "    for model in struct:\n",
    "        file = os.path.join(pdb_path, f'{pdb_id}.pdb')\n",
    "        pdb_id_to_dssp_file(pdb_id)\n",
    "        dssp = DSSP(model=model, in_file=os.path.join(dssp_path, f'{pdb_id}.dssp'), file_type='DSSP')\n",
    "        #dssp = DSSP(model=model, in_file=file)\n",
    "        for i, chain in enumerate(model):\n",
    "            for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                seq = pp.get_sequence()\n",
    "                for j, residue in enumerate(pp):\n",
    "                    #print(residue.id[1], seq[i])\n",
    "                    #print(dssp.keys()[j % (i + 1)])\n",
    "                    key = list(dssp.keys())[seq_count]\n",
    "                    seq_count += 1\n",
    "                    dssp_info = dssp[key]\n",
    "                    #print(dssp_info)\n",
    "                    #dssp_id = dssp_info[0]\n",
    "                    amino_acid = dssp_info[1]\n",
    "                    sec_struct = dssp_info[2]\n",
    "                    solv_acc = dssp_info[3]\n",
    "                    phi = dssp_info[4]\n",
    "                    psi = dssp_info[5]\n",
    "                    # Keys 6 through 13 is bonding energy / relidx (no clue what this is)\n",
    "                    #print(dssp[dssp.keys()[i]])\n",
    "                    struct_info.append([model.full_id[0], model.full_id[1], chain.id,\n",
    "                    residue.resname, residue.id[1], amino_acid, sec_struct, solv_acc, phi, psi])\n",
    "    info_df = pd.DataFrame(struct_info, \n",
    "        columns=['Model_Name', 'Model_ID', 'Chain', 'Residue_Name',\n",
    "        'Residue_ID', 'Amino_Acid', 'Secondary_Structure', 'Solvent_Accessability', \n",
    "        'Phi', 'Psi'])\n",
    "    if partial:    \n",
    "        seq_bool = get_seq_alignment(full_pdb_id)\n",
    "        fin_df = info_df[info_df['Chain'] == full_pdb_id[4].upper()]\n",
    "        return fin_df[seq_bool]\n",
    "    else:\n",
    "        return info_df\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this generates the secondary structure ground truth proportions\n",
    "# for a protein in a certain range(start,stop) of amino acids\n",
    "\n",
    "def ss_ground_truth(protein_id, aa_range,secs):\n",
    "    sstructs=\"HBEGITS\"\n",
    "\n",
    "    k=secs[secs['Chain']==protein_id[4]]\n",
    "    k=k.iloc[aa_range,:]\n",
    "\n",
    "    ss_count=[]\n",
    "\n",
    "    for ss in sstructs:\n",
    "        ss_count.append(len(k[k['Secondary_Structure']==ss])/len(k))\n",
    "\n",
    "    return ss_count\n",
    "\n",
    "# H = alpha helix\n",
    "# B = residue in isolated beta-bridge\n",
    "# E = extended strand, participates in beta ladder\n",
    "# G = 3-helix (3/10 helix)\n",
    "# I = 5 helix (pi helix)\n",
    "# T = hydrogen bonded turn\n",
    "# S = bend\n",
    "\n",
    "\n",
    "# DON'T USE THIS\n",
    "# its too intensive to call it over and over again\n",
    "def format_torsion_angles_to_dataframe(pdb_id, start, chip_size):\n",
    "    \n",
    "    get_pdb_file(pdb_id)\n",
    "    \n",
    "    angles=[]\n",
    "    ta=get_torsion_angles(pdb_id)\n",
    "\n",
    "    angles+=list(ta.iloc[start:start+chip_size,:][\"Phi\"])\n",
    "    angles+=list(ta.iloc[start:start+chip_size,:][\"Psi\"])\n",
    "    \n",
    "    return angles\n",
    "\n",
    "def get_torsion_for_df(pdb_id):\n",
    "    get_pdb_file(pdb_id)\n",
    "    return get_torsion_angles(pdb_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define functions\n",
    "\n",
    "# mean function\n",
    "# just for utility purposes\n",
    "def mean(numlist):\n",
    "    return sum(numlist)/len(numlist)\n",
    "\n",
    "# this creates triangular chips along (but not including)\n",
    "# the diagonal within the comparison matrix\n",
    "def chip_diagonal(chip_size, step_size, aa_length):\n",
    "    tri=[]\n",
    "    for diag in range(0,aa_length-chip_size,step_size):\n",
    "        tri.append(\n",
    "            [(y,x) for y in range(diag,diag+chip_size) for x in range(y+1,chip_size+diag+1)]\n",
    "        )\n",
    "        # NB: y already has diag added to it\n",
    "    return tri\n",
    "\n",
    "\n",
    "# create a gif showing where the triangular chipping window is located\n",
    "def gif_from_tri(protein_id,tri,filename):\n",
    "    \n",
    "    protein = np.array(distance_maps_cb[protein_id])\n",
    "    l=len(protein)\n",
    "\n",
    "    gif=[]\n",
    "\n",
    "    for t in tri:\n",
    "        new_test=np.ones((l,l))*0\n",
    "        for (y,x) in t:\n",
    "            new_test[y,x]=protein[y,x]\n",
    "\n",
    "        im=plt.imshow(new_test)\n",
    "\n",
    "        x=im.make_image(\"AGG\")[0]\n",
    "        x=np.flipud(x)\n",
    "\n",
    "        gif.append(np.array(x))\n",
    "\n",
    "    imageio.mimsave(filename+'.gif', gif, fps=5)\n",
    "\n",
    "\n",
    "# create a dataframe from a given protein id, chip size, and step size\n",
    "def diag_chips_to_df(protein_id, chip_size, step_size):\n",
    "    \n",
    "    # arrange all the feature matrices into proper form so we can \n",
    "    # reference them properly later\n",
    "    \n",
    "    chain_id=protein_id[4]\n",
    "    \n",
    "    protein_dist = np.array(get_cb_distances(protein_id)[chain_id])\n",
    "    aa_length=len(protein_dist[0])\n",
    "    \n",
    "    protein_feat = np.array(input_features[protein_id])\n",
    "\n",
    "    feat_len=len(protein_feat[0])\n",
    "    pad_len=aa_length-feat_len\n",
    "    \n",
    "    ccmpred=protein_feat[5].astype(np.float)\n",
    "    ccmpred.shape=(feat_len,feat_len)\n",
    "    ccmpred=np.pad(ccmpred,pad_len,'constant', constant_values=(0))\n",
    "    \n",
    "    freecontact=protein_feat[6].astype(np.float)\n",
    "    freecontact.shape=(feat_len,feat_len)\n",
    "    freecontact=np.pad(freecontact,pad_len,'constant', constant_values=(0))\n",
    "    \n",
    "    pstat_pots=protein_feat[7].astype(np.float)\n",
    "    pstat_pots.shape=(feat_len,feat_len)\n",
    "    pstat_pots=np.pad(pstat_pots,pad_len,'constant', constant_values=(0))\n",
    "    \n",
    "    chips=chip_diagonal(chip_size, step_size, aa_length)\n",
    "    \n",
    "    gt=get_ground_truth_api(protein_id[0:4])\n",
    "    gt_chain=gt[gt['Chain']==protein_id[4]]\n",
    "    \n",
    "    # create column labels\n",
    "    cols=[\"protein_id\",\"chip_id\"]\n",
    "    cols+=[\"start\",\"stop\"]\n",
    "    cols+=[\"dist_\"+str(n) for n in range(0,len(chips[0]))]\n",
    "    cols+=[\"psipred_helix\",\"psipred_sheet\",\"psipred_coil\"]\n",
    "    cols+=[\"psisolv\",\"shannon_entropy\"]\n",
    "    cols+=[\"ccmpred\",\"freecontact\",\"pstat_pots\"]\n",
    "    cols+=[\"ground_truth_\"+x for x in \"HBEGITS\"]\n",
    "    cols+=[\"phi_\"+str(x) for x in range(0,chip_size)]\n",
    "    cols+=[\"psi_\"+str(x) for x in range(0,chip_size)]\n",
    "\n",
    "    \n",
    "    chiplist=[]\n",
    "#     get_ground_truth_api(pdb_list_y[0][0:4])\n",
    "    t_phi=list(gt_chain['Phi'])\n",
    "    t_psi=list(gt_chain['Psi'])\n",
    "    \n",
    "    # loop through all the chips\n",
    "    for i in range(0,len(chips)):\n",
    "        \n",
    "        # row identifiers\n",
    "        row=[protein_id,\"chip_\"+str(i)]\n",
    "        row+=[i,i+chip_size]\n",
    "        \n",
    "        # inclusion range\n",
    "        # between the first amino acid being compared \n",
    "        # and the last amino acid being compared\n",
    "        # NB this assumes that \n",
    "        incl_range=range(min(min(chips[i])),max(max(chips[i])))\n",
    "        \n",
    "        # 1d features\n",
    "        base_helix=protein_feat[0].astype(np.float)\n",
    "        base_sheet=protein_feat[1].astype(np.float)\n",
    "        base_coil=protein_feat[2].astype(np.float)\n",
    "        base_solv=protein_feat[3].astype(np.float)\n",
    "        base_shan=protein_feat[4].astype(np.float)\n",
    "        \n",
    "        base_helix=np.pad(base_helix,pad_len,'constant', constant_values=(0))\n",
    "        base_sheet=np.pad(base_sheet,pad_len,'constant', constant_values=(0))\n",
    "        base_coil=np.pad(base_coil,pad_len,'constant', constant_values=(0))\n",
    "        base_solv=np.pad(base_solv,pad_len,'constant', constant_values=(0))\n",
    "        base_shan=np.pad(base_shan,pad_len,'constant', constant_values=(0))\n",
    "        \n",
    "        psipred_helix=mean(base_helix[incl_range])\n",
    "        psipred_sheet=mean(base_sheet[incl_range])\n",
    "        psipred_coil=mean(base_coil[incl_range])\n",
    "        psisolv=mean(base_solv[incl_range])\n",
    "        shannon_entropy=mean(base_shan[incl_range])\n",
    "        \n",
    "        \n",
    "        \n",
    "        # 2d features\n",
    "        ccmpred_pool=[]\n",
    "        freecontact_pool=[]\n",
    "        pstat_pool=[]\n",
    "        \n",
    "        # ss ground truth\n",
    "        ssgt=ss_ground_truth(protein_id,incl_range,gt_chain)\n",
    "        \n",
    "        # loop through each pixel in the chip\n",
    "        for n in range(0,len(chips[i])):\n",
    "            \n",
    "#             make sure we're not on the diagonal\n",
    "#             print([chips[i][n][0],chips[i][n][1]])\n",
    "            \n",
    "            row.append(\n",
    "                protein_dist[chips[i][n][0],chips[i][n][1]]\n",
    "            )\n",
    "            ccmpred_pool.append(\n",
    "                ccmpred[chips[i][n][0],chips[i][n][1]]\n",
    "            )\n",
    "            freecontact_pool.append(\n",
    "                freecontact[chips[i][n][0],chips[i][n][1]]\n",
    "            )\n",
    "            pstat_pool.append(\n",
    "                pstat_pots[chips[i][n][0],chips[i][n][1]]\n",
    "            )\n",
    "            \n",
    "        # 1d features\n",
    "        row.append(psipred_helix)\n",
    "        row.append(psipred_sheet)\n",
    "        row.append(psipred_coil)\n",
    "        row.append(psisolv)\n",
    "        row.append(shannon_entropy)\n",
    "   \n",
    "        # 2d features\n",
    "        row.append(max(ccmpred_pool))\n",
    "        row.append(max(freecontact_pool))\n",
    "        row.append(max(pstat_pool))\n",
    "        \n",
    "        # ground truth\n",
    "        [row.append(x) for x in ssgt]\n",
    "        \n",
    "        # torsional angles\n",
    "        row+=t_phi[i:i+chip_size]\n",
    "        row+=t_psi[i:i+chip_size]\n",
    "        \n",
    "#         print(row)\n",
    "        \n",
    "        # add to df\n",
    "        chiplist.append(row)\n",
    "\n",
    "#     AssertionError: 94 columns passed, passed data had 75 columns\n",
    "# almost correct number of columns, but torsion angles not being passed correctly?\n",
    "    chip_df=pandas.DataFrame(chiplist,columns=cols)\n",
    "\n",
    "    return chip_df\n",
    "\n",
    "\n",
    "# compute the ground truth secondary structure\n",
    "# of which the majority of amino acids in a window\n",
    "# are a part of\n",
    "# i.e. that whose percentage within the window is above 50%\n",
    "def compute_majority_ss(df):\n",
    "    for ss in \"HBEGITS\":\n",
    "        df[\"maj_\"+ss]=df['ground_truth_'+ss]>.5\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# some variable sizes\n",
    "\n",
    "chip_size=10\n",
    "step_size=1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# demonstrate the gif function\n",
    "\n",
    "# gif_from_tri(\"1hzfA0\",chip_diagonal(50,1,256),\"chiptri_diag\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "D:\\Programs\\lib\\site-packages\\Bio\\PDB\\Polypeptide.py:328: UserWarning: Assuming residue  CA is an unknown modified amino acid\n",
      "  \"amino acid\" % residue.get_resname())\n",
      "D:\\Programs\\lib\\site-packages\\Bio\\PDB\\Polypeptide.py:328: UserWarning: Assuming residue  CA is an unknown modified amino acid\n",
      "  \"amino acid\" % residue.get_resname())\n",
      "D:\\Programs\\lib\\site-packages\\Bio\\PDB\\Polypeptide.py:328: UserWarning: Assuming residue  CA is an unknown modified amino acid\n",
      "  \"amino acid\" % residue.get_resname())\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-74-4251ae15bf95>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      8\u001b[0m \u001b[1;32mwith\u001b[0m \u001b[0mconcurrent\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfutures\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mThreadPoolExecutor\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mmax_workers\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mavailable_cpu\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mexecutor\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 9\u001b[1;33m     \u001b[0mdataset2\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mpandas\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mconcat\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mexecutor\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmap\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mchip_parallel\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mpdb_list_y\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     10\u001b[0m \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"parallel completed in \"\u001b[0m\u001b[1;33m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtime\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mclock\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m-\u001b[0m\u001b[0mstart_time\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[0mcompute_majority_ss\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdataset\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Programs\\lib\\site-packages\\pandas\\core\\reshape\\concat.py\u001b[0m in \u001b[0;36mconcat\u001b[1;34m(objs, axis, join, join_axes, ignore_index, keys, levels, names, verify_integrity, sort, copy)\u001b[0m\n\u001b[0;32m    226\u001b[0m                        \u001b[0mkeys\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mkeys\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlevels\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mlevels\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mnames\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mnames\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    227\u001b[0m                        \u001b[0mverify_integrity\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mverify_integrity\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 228\u001b[1;33m                        copy=copy, sort=sort)\n\u001b[0m\u001b[0;32m    229\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0mop\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mget_result\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    230\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Programs\\lib\\site-packages\\pandas\\core\\reshape\\concat.py\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, objs, axis, join, join_axes, keys, levels, names, ignore_index, verify_integrity, copy, sort)\u001b[0m\n\u001b[0;32m    257\u001b[0m             \u001b[0mobjs\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mobjs\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mk\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mk\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mkeys\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    258\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 259\u001b[1;33m             \u001b[0mobjs\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mobjs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    260\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    261\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mobjs\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Programs\\lib\\concurrent\\futures\\_base.py\u001b[0m in \u001b[0;36mresult_iterator\u001b[1;34m()\u001b[0m\n\u001b[0;32m    584\u001b[0m                     \u001b[1;31m# Careful not to keep a reference to the popped future\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    585\u001b[0m                     \u001b[1;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 586\u001b[1;33m                         \u001b[1;32myield\u001b[0m \u001b[0mfs\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpop\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mresult\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    587\u001b[0m                     \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    588\u001b[0m                         \u001b[1;32myield\u001b[0m \u001b[0mfs\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mpop\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mresult\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mend_time\u001b[0m \u001b[1;33m-\u001b[0m \u001b[0mtime\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmonotonic\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Programs\\lib\\concurrent\\futures\\_base.py\u001b[0m in \u001b[0;36mresult\u001b[1;34m(self, timeout)\u001b[0m\n\u001b[0;32m    423\u001b[0m                 \u001b[1;32mraise\u001b[0m \u001b[0mCancelledError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    424\u001b[0m             \u001b[1;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_state\u001b[0m \u001b[1;33m==\u001b[0m \u001b[0mFINISHED\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 425\u001b[1;33m                 \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__get_result\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    426\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    427\u001b[0m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_condition\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Programs\\lib\\concurrent\\futures\\_base.py\u001b[0m in \u001b[0;36m__get_result\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    382\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m__get_result\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    383\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_exception\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 384\u001b[1;33m             \u001b[1;32mraise\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_exception\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    385\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    386\u001b[0m             \u001b[1;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_result\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32mD:\\Programs\\lib\\concurrent\\futures\\thread.py\u001b[0m in \u001b[0;36mrun\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m     54\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     55\u001b[0m         \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 56\u001b[1;33m             \u001b[0mresult\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfn\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0margs\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     57\u001b[0m         \u001b[1;32mexcept\u001b[0m \u001b[0mBaseException\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mexc\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     58\u001b[0m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mfuture\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mset_exception\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mexc\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-74-4251ae15bf95>\u001b[0m in \u001b[0;36mchip_parallel\u001b[1;34m(x)\u001b[0m\n\u001b[0;32m      2\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mchip_parallel\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 4\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0mdiag_chips_to_df\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mx\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m10\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      5\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[0mstart_time\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mtime\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mclock\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-71-9efe53fbd4d3>\u001b[0m in \u001b[0;36mdiag_chips_to_df\u001b[1;34m(protein_id, chip_size, step_size)\u001b[0m\n\u001b[0;32m     71\u001b[0m     \u001b[0mchips\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mchip_diagonal\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mchip_size\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mstep_size\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maa_length\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     72\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 73\u001b[1;33m     \u001b[0mgt\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mget_ground_truth_api\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mprotein_id\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;36m4\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     74\u001b[0m     \u001b[0mgt_chain\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mgt\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mgt\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'Chain'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m==\u001b[0m\u001b[0mprotein_id\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m4\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     75\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-4-5d51b7b4fe95>\u001b[0m in \u001b[0;36mget_ground_truth_api\u001b[1;34m(full_pdb_id, partial)\u001b[0m\n\u001b[0;32m    264\u001b[0m                     \u001b[1;31m#print(residue.id[1], seq[i])\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    265\u001b[0m                     \u001b[1;31m#print(dssp.keys()[j % (i + 1)])\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 266\u001b[1;33m                     \u001b[0mkey\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mdssp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mseq_count\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    267\u001b[0m                     \u001b[0mseq_count\u001b[0m \u001b[1;33m+=\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    268\u001b[0m                     \u001b[0mdssp_info\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mdssp\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "# demonstrate how to create a dataset from it\n",
    "\n",
    "def chip_parallel(x):\n",
    "    return diag_chips_to_df(x,10,1)\n",
    "\n",
    "start_time=time.clock()\n",
    "\n",
    "with concurrent.futures.ThreadPoolExecutor(max_workers=available_cpu) as executor:\n",
    "    dataset2=pandas.concat(executor.map(chip_parallel, pdb_list_y))\n",
    "print(\"parallel completed in \"+ str(time.clock()-start_time))\n",
    "compute_majority_ss(dataset)\n",
    "\n",
    "dataset\n",
    "dataset.to_csv(\"datasets/protein_test_torsion.csv\")\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dataset=pandas.concat([(diag_chips_to_df(pid,10,1)) for pid in pdb_list_y[0:1]])\n",
    "df=diag_chips_to_df(pdb_list_y[0],10,1)\n",
    "\n",
    "# get_torsion_for_df(pdb_list_y[0][0:4])\n",
    "# gt=get_ground_truth_api(pdb_list_y[0][0:4])\n",
    "\n",
    "# gt_chain=gt[gt['Chain']==pdb_list_y[0][4]]\n",
    "\n",
    "# aa_range=list(range(240,250))\n",
    "\n",
    "# sstructs=\"HBEGITS\"\n",
    "\n",
    "# k=gt_chain[gt_chain['Chain']==pdb_list_y[0][4]]\n",
    "# k=k.iloc[aa_range,:]\n",
    "\n",
    "# ss_count=[]\n",
    "\n",
    "# for ss in sstructs:\n",
    "#     ss_count.append(len(k[k['Secondary_Structure']==ss])/len(k))\n",
    "\n",
    "# ss_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>chip_id</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>dist_0</th>\n",
       "      <th>dist_1</th>\n",
       "      <th>dist_2</th>\n",
       "      <th>dist_3</th>\n",
       "      <th>dist_4</th>\n",
       "      <th>dist_5</th>\n",
       "      <th>...</th>\n",
       "      <th>psi_0</th>\n",
       "      <th>psi_1</th>\n",
       "      <th>psi_2</th>\n",
       "      <th>psi_3</th>\n",
       "      <th>psi_4</th>\n",
       "      <th>psi_5</th>\n",
       "      <th>psi_6</th>\n",
       "      <th>psi_7</th>\n",
       "      <th>psi_8</th>\n",
       "      <th>psi_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>2.931498</td>\n",
       "      <td>4.466173</td>\n",
       "      <td>4.830971</td>\n",
       "      <td>5.929355</td>\n",
       "      <td>8.004140</td>\n",
       "      <td>9.493499</td>\n",
       "      <td>...</td>\n",
       "      <td>157.9</td>\n",
       "      <td>-56.8</td>\n",
       "      <td>-51.1</td>\n",
       "      <td>-58.3</td>\n",
       "      <td>-39.3</td>\n",
       "      <td>-33.5</td>\n",
       "      <td>-32.3</td>\n",
       "      <td>-57.3</td>\n",
       "      <td>-50.5</td>\n",
       "      <td>-41.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_1</td>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>2.977764</td>\n",
       "      <td>4.556571</td>\n",
       "      <td>4.825561</td>\n",
       "      <td>5.993463</td>\n",
       "      <td>8.187191</td>\n",
       "      <td>9.461169</td>\n",
       "      <td>...</td>\n",
       "      <td>-56.8</td>\n",
       "      <td>-51.1</td>\n",
       "      <td>-58.3</td>\n",
       "      <td>-39.3</td>\n",
       "      <td>-33.5</td>\n",
       "      <td>-32.3</td>\n",
       "      <td>-57.3</td>\n",
       "      <td>-50.5</td>\n",
       "      <td>-41.7</td>\n",
       "      <td>-44.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_2</td>\n",
       "      <td>2</td>\n",
       "      <td>12</td>\n",
       "      <td>3.011051</td>\n",
       "      <td>4.573128</td>\n",
       "      <td>4.944533</td>\n",
       "      <td>6.271376</td>\n",
       "      <td>8.255353</td>\n",
       "      <td>9.777748</td>\n",
       "      <td>...</td>\n",
       "      <td>-51.1</td>\n",
       "      <td>-58.3</td>\n",
       "      <td>-39.3</td>\n",
       "      <td>-33.5</td>\n",
       "      <td>-32.3</td>\n",
       "      <td>-57.3</td>\n",
       "      <td>-50.5</td>\n",
       "      <td>-41.7</td>\n",
       "      <td>-44.9</td>\n",
       "      <td>-54.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_3</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>3.055943</td>\n",
       "      <td>4.643566</td>\n",
       "      <td>4.900433</td>\n",
       "      <td>6.196664</td>\n",
       "      <td>8.368240</td>\n",
       "      <td>9.765861</td>\n",
       "      <td>...</td>\n",
       "      <td>-58.3</td>\n",
       "      <td>-39.3</td>\n",
       "      <td>-33.5</td>\n",
       "      <td>-32.3</td>\n",
       "      <td>-57.3</td>\n",
       "      <td>-50.5</td>\n",
       "      <td>-41.7</td>\n",
       "      <td>-44.9</td>\n",
       "      <td>-54.2</td>\n",
       "      <td>-29.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_4</td>\n",
       "      <td>4</td>\n",
       "      <td>14</td>\n",
       "      <td>3.033617</td>\n",
       "      <td>4.514310</td>\n",
       "      <td>4.787584</td>\n",
       "      <td>6.232026</td>\n",
       "      <td>8.312329</td>\n",
       "      <td>9.773744</td>\n",
       "      <td>...</td>\n",
       "      <td>-39.3</td>\n",
       "      <td>-33.5</td>\n",
       "      <td>-32.3</td>\n",
       "      <td>-57.3</td>\n",
       "      <td>-50.5</td>\n",
       "      <td>-41.7</td>\n",
       "      <td>-44.9</td>\n",
       "      <td>-54.2</td>\n",
       "      <td>-29.9</td>\n",
       "      <td>-53.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_5</td>\n",
       "      <td>5</td>\n",
       "      <td>15</td>\n",
       "      <td>3.223323</td>\n",
       "      <td>4.594050</td>\n",
       "      <td>5.066024</td>\n",
       "      <td>6.347551</td>\n",
       "      <td>8.505913</td>\n",
       "      <td>9.938494</td>\n",
       "      <td>...</td>\n",
       "      <td>-33.5</td>\n",
       "      <td>-32.3</td>\n",
       "      <td>-57.3</td>\n",
       "      <td>-50.5</td>\n",
       "      <td>-41.7</td>\n",
       "      <td>-44.9</td>\n",
       "      <td>-54.2</td>\n",
       "      <td>-29.9</td>\n",
       "      <td>-53.6</td>\n",
       "      <td>-50.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_6</td>\n",
       "      <td>6</td>\n",
       "      <td>16</td>\n",
       "      <td>2.976352</td>\n",
       "      <td>4.657997</td>\n",
       "      <td>5.042452</td>\n",
       "      <td>6.315280</td>\n",
       "      <td>8.412719</td>\n",
       "      <td>9.836761</td>\n",
       "      <td>...</td>\n",
       "      <td>-32.3</td>\n",
       "      <td>-57.3</td>\n",
       "      <td>-50.5</td>\n",
       "      <td>-41.7</td>\n",
       "      <td>-44.9</td>\n",
       "      <td>-54.2</td>\n",
       "      <td>-29.9</td>\n",
       "      <td>-53.6</td>\n",
       "      <td>-50.3</td>\n",
       "      <td>-28.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_7</td>\n",
       "      <td>7</td>\n",
       "      <td>17</td>\n",
       "      <td>2.983589</td>\n",
       "      <td>4.640706</td>\n",
       "      <td>5.149854</td>\n",
       "      <td>6.408232</td>\n",
       "      <td>8.367692</td>\n",
       "      <td>9.668061</td>\n",
       "      <td>...</td>\n",
       "      <td>-57.3</td>\n",
       "      <td>-50.5</td>\n",
       "      <td>-41.7</td>\n",
       "      <td>-44.9</td>\n",
       "      <td>-54.2</td>\n",
       "      <td>-29.9</td>\n",
       "      <td>-53.6</td>\n",
       "      <td>-50.3</td>\n",
       "      <td>-28.8</td>\n",
       "      <td>-42.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_8</td>\n",
       "      <td>8</td>\n",
       "      <td>18</td>\n",
       "      <td>3.081327</td>\n",
       "      <td>4.675354</td>\n",
       "      <td>5.066069</td>\n",
       "      <td>6.178646</td>\n",
       "      <td>8.111391</td>\n",
       "      <td>9.779182</td>\n",
       "      <td>...</td>\n",
       "      <td>-50.5</td>\n",
       "      <td>-41.7</td>\n",
       "      <td>-44.9</td>\n",
       "      <td>-54.2</td>\n",
       "      <td>-29.9</td>\n",
       "      <td>-53.6</td>\n",
       "      <td>-50.3</td>\n",
       "      <td>-28.8</td>\n",
       "      <td>-42.9</td>\n",
       "      <td>-32.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_9</td>\n",
       "      <td>9</td>\n",
       "      <td>19</td>\n",
       "      <td>3.076244</td>\n",
       "      <td>4.648451</td>\n",
       "      <td>5.031323</td>\n",
       "      <td>5.955137</td>\n",
       "      <td>8.097498</td>\n",
       "      <td>9.558247</td>\n",
       "      <td>...</td>\n",
       "      <td>-41.7</td>\n",
       "      <td>-44.9</td>\n",
       "      <td>-54.2</td>\n",
       "      <td>-29.9</td>\n",
       "      <td>-53.6</td>\n",
       "      <td>-50.3</td>\n",
       "      <td>-28.8</td>\n",
       "      <td>-42.9</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-34.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_10</td>\n",
       "      <td>10</td>\n",
       "      <td>20</td>\n",
       "      <td>3.038332</td>\n",
       "      <td>4.728816</td>\n",
       "      <td>4.798806</td>\n",
       "      <td>5.939553</td>\n",
       "      <td>7.934079</td>\n",
       "      <td>9.386290</td>\n",
       "      <td>...</td>\n",
       "      <td>-44.9</td>\n",
       "      <td>-54.2</td>\n",
       "      <td>-29.9</td>\n",
       "      <td>-53.6</td>\n",
       "      <td>-50.3</td>\n",
       "      <td>-28.8</td>\n",
       "      <td>-42.9</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-34.3</td>\n",
       "      <td>-35.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_11</td>\n",
       "      <td>11</td>\n",
       "      <td>21</td>\n",
       "      <td>3.010424</td>\n",
       "      <td>4.472245</td>\n",
       "      <td>5.071977</td>\n",
       "      <td>5.966181</td>\n",
       "      <td>7.969033</td>\n",
       "      <td>9.478682</td>\n",
       "      <td>...</td>\n",
       "      <td>-54.2</td>\n",
       "      <td>-29.9</td>\n",
       "      <td>-53.6</td>\n",
       "      <td>-50.3</td>\n",
       "      <td>-28.8</td>\n",
       "      <td>-42.9</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-34.3</td>\n",
       "      <td>-35.7</td>\n",
       "      <td>-37.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_12</td>\n",
       "      <td>12</td>\n",
       "      <td>22</td>\n",
       "      <td>3.148242</td>\n",
       "      <td>4.942146</td>\n",
       "      <td>5.011508</td>\n",
       "      <td>6.109199</td>\n",
       "      <td>8.375867</td>\n",
       "      <td>9.568907</td>\n",
       "      <td>...</td>\n",
       "      <td>-29.9</td>\n",
       "      <td>-53.6</td>\n",
       "      <td>-50.3</td>\n",
       "      <td>-28.8</td>\n",
       "      <td>-42.9</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-34.3</td>\n",
       "      <td>-35.7</td>\n",
       "      <td>-37.6</td>\n",
       "      <td>-40.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_13</td>\n",
       "      <td>13</td>\n",
       "      <td>23</td>\n",
       "      <td>2.977663</td>\n",
       "      <td>4.486660</td>\n",
       "      <td>4.839495</td>\n",
       "      <td>6.330968</td>\n",
       "      <td>8.282268</td>\n",
       "      <td>9.775724</td>\n",
       "      <td>...</td>\n",
       "      <td>-53.6</td>\n",
       "      <td>-50.3</td>\n",
       "      <td>-28.8</td>\n",
       "      <td>-42.9</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-34.3</td>\n",
       "      <td>-35.7</td>\n",
       "      <td>-37.6</td>\n",
       "      <td>-40.5</td>\n",
       "      <td>-28.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_14</td>\n",
       "      <td>14</td>\n",
       "      <td>24</td>\n",
       "      <td>3.080211</td>\n",
       "      <td>4.458592</td>\n",
       "      <td>4.706429</td>\n",
       "      <td>6.142321</td>\n",
       "      <td>8.378738</td>\n",
       "      <td>9.686699</td>\n",
       "      <td>...</td>\n",
       "      <td>-50.3</td>\n",
       "      <td>-28.8</td>\n",
       "      <td>-42.9</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-34.3</td>\n",
       "      <td>-35.7</td>\n",
       "      <td>-37.6</td>\n",
       "      <td>-40.5</td>\n",
       "      <td>-28.9</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_15</td>\n",
       "      <td>15</td>\n",
       "      <td>25</td>\n",
       "      <td>2.932920</td>\n",
       "      <td>4.282657</td>\n",
       "      <td>4.616333</td>\n",
       "      <td>6.260442</td>\n",
       "      <td>8.352125</td>\n",
       "      <td>9.567734</td>\n",
       "      <td>...</td>\n",
       "      <td>-28.8</td>\n",
       "      <td>-42.9</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-34.3</td>\n",
       "      <td>-35.7</td>\n",
       "      <td>-37.6</td>\n",
       "      <td>-40.5</td>\n",
       "      <td>-28.9</td>\n",
       "      <td>0.3</td>\n",
       "      <td>44.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_16</td>\n",
       "      <td>16</td>\n",
       "      <td>26</td>\n",
       "      <td>3.113236</td>\n",
       "      <td>4.620432</td>\n",
       "      <td>5.094796</td>\n",
       "      <td>6.513730</td>\n",
       "      <td>8.500165</td>\n",
       "      <td>9.696307</td>\n",
       "      <td>...</td>\n",
       "      <td>-42.9</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-34.3</td>\n",
       "      <td>-35.7</td>\n",
       "      <td>-37.6</td>\n",
       "      <td>-40.5</td>\n",
       "      <td>-28.9</td>\n",
       "      <td>0.3</td>\n",
       "      <td>44.3</td>\n",
       "      <td>137.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_17</td>\n",
       "      <td>17</td>\n",
       "      <td>27</td>\n",
       "      <td>3.003047</td>\n",
       "      <td>4.556078</td>\n",
       "      <td>5.047412</td>\n",
       "      <td>6.118340</td>\n",
       "      <td>7.877437</td>\n",
       "      <td>9.032131</td>\n",
       "      <td>...</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-34.3</td>\n",
       "      <td>-35.7</td>\n",
       "      <td>-37.6</td>\n",
       "      <td>-40.5</td>\n",
       "      <td>-28.9</td>\n",
       "      <td>0.3</td>\n",
       "      <td>44.3</td>\n",
       "      <td>137.6</td>\n",
       "      <td>152.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_18</td>\n",
       "      <td>18</td>\n",
       "      <td>28</td>\n",
       "      <td>3.142842</td>\n",
       "      <td>4.863575</td>\n",
       "      <td>5.154842</td>\n",
       "      <td>5.744147</td>\n",
       "      <td>7.385431</td>\n",
       "      <td>8.627468</td>\n",
       "      <td>...</td>\n",
       "      <td>-34.3</td>\n",
       "      <td>-35.7</td>\n",
       "      <td>-37.6</td>\n",
       "      <td>-40.5</td>\n",
       "      <td>-28.9</td>\n",
       "      <td>0.3</td>\n",
       "      <td>44.3</td>\n",
       "      <td>137.6</td>\n",
       "      <td>152.9</td>\n",
       "      <td>134.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_19</td>\n",
       "      <td>19</td>\n",
       "      <td>29</td>\n",
       "      <td>3.128385</td>\n",
       "      <td>4.869174</td>\n",
       "      <td>5.030503</td>\n",
       "      <td>5.046117</td>\n",
       "      <td>5.781323</td>\n",
       "      <td>5.672482</td>\n",
       "      <td>...</td>\n",
       "      <td>-35.7</td>\n",
       "      <td>-37.6</td>\n",
       "      <td>-40.5</td>\n",
       "      <td>-28.9</td>\n",
       "      <td>0.3</td>\n",
       "      <td>44.3</td>\n",
       "      <td>137.6</td>\n",
       "      <td>152.9</td>\n",
       "      <td>134.1</td>\n",
       "      <td>155.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_20</td>\n",
       "      <td>20</td>\n",
       "      <td>30</td>\n",
       "      <td>3.220644</td>\n",
       "      <td>4.942284</td>\n",
       "      <td>4.694743</td>\n",
       "      <td>4.624504</td>\n",
       "      <td>5.171173</td>\n",
       "      <td>7.949272</td>\n",
       "      <td>...</td>\n",
       "      <td>-37.6</td>\n",
       "      <td>-40.5</td>\n",
       "      <td>-28.9</td>\n",
       "      <td>0.3</td>\n",
       "      <td>44.3</td>\n",
       "      <td>137.6</td>\n",
       "      <td>152.9</td>\n",
       "      <td>134.1</td>\n",
       "      <td>155.1</td>\n",
       "      <td>129.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_21</td>\n",
       "      <td>21</td>\n",
       "      <td>31</td>\n",
       "      <td>3.162164</td>\n",
       "      <td>4.995806</td>\n",
       "      <td>6.318177</td>\n",
       "      <td>8.039869</td>\n",
       "      <td>11.060770</td>\n",
       "      <td>13.959582</td>\n",
       "      <td>...</td>\n",
       "      <td>-40.5</td>\n",
       "      <td>-28.9</td>\n",
       "      <td>0.3</td>\n",
       "      <td>44.3</td>\n",
       "      <td>137.6</td>\n",
       "      <td>152.9</td>\n",
       "      <td>134.1</td>\n",
       "      <td>155.1</td>\n",
       "      <td>129.1</td>\n",
       "      <td>137.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_22</td>\n",
       "      <td>22</td>\n",
       "      <td>32</td>\n",
       "      <td>3.548883</td>\n",
       "      <td>6.188423</td>\n",
       "      <td>8.440619</td>\n",
       "      <td>11.621278</td>\n",
       "      <td>14.576628</td>\n",
       "      <td>17.376261</td>\n",
       "      <td>...</td>\n",
       "      <td>-28.9</td>\n",
       "      <td>0.3</td>\n",
       "      <td>44.3</td>\n",
       "      <td>137.6</td>\n",
       "      <td>152.9</td>\n",
       "      <td>134.1</td>\n",
       "      <td>155.1</td>\n",
       "      <td>129.1</td>\n",
       "      <td>137.6</td>\n",
       "      <td>164.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_23</td>\n",
       "      <td>23</td>\n",
       "      <td>33</td>\n",
       "      <td>3.039205</td>\n",
       "      <td>5.773140</td>\n",
       "      <td>9.032375</td>\n",
       "      <td>12.072133</td>\n",
       "      <td>14.798532</td>\n",
       "      <td>17.520407</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3</td>\n",
       "      <td>44.3</td>\n",
       "      <td>137.6</td>\n",
       "      <td>152.9</td>\n",
       "      <td>134.1</td>\n",
       "      <td>155.1</td>\n",
       "      <td>129.1</td>\n",
       "      <td>137.6</td>\n",
       "      <td>164.0</td>\n",
       "      <td>-27.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_24</td>\n",
       "      <td>24</td>\n",
       "      <td>34</td>\n",
       "      <td>3.400494</td>\n",
       "      <td>6.638870</td>\n",
       "      <td>9.738158</td>\n",
       "      <td>12.659022</td>\n",
       "      <td>15.550603</td>\n",
       "      <td>18.014523</td>\n",
       "      <td>...</td>\n",
       "      <td>44.3</td>\n",
       "      <td>137.6</td>\n",
       "      <td>152.9</td>\n",
       "      <td>134.1</td>\n",
       "      <td>155.1</td>\n",
       "      <td>129.1</td>\n",
       "      <td>137.6</td>\n",
       "      <td>164.0</td>\n",
       "      <td>-27.9</td>\n",
       "      <td>165.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_25</td>\n",
       "      <td>25</td>\n",
       "      <td>35</td>\n",
       "      <td>3.336457</td>\n",
       "      <td>6.442289</td>\n",
       "      <td>9.494033</td>\n",
       "      <td>12.483414</td>\n",
       "      <td>15.055290</td>\n",
       "      <td>17.628443</td>\n",
       "      <td>...</td>\n",
       "      <td>137.6</td>\n",
       "      <td>152.9</td>\n",
       "      <td>134.1</td>\n",
       "      <td>155.1</td>\n",
       "      <td>129.1</td>\n",
       "      <td>137.6</td>\n",
       "      <td>164.0</td>\n",
       "      <td>-27.9</td>\n",
       "      <td>165.0</td>\n",
       "      <td>151.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_26</td>\n",
       "      <td>26</td>\n",
       "      <td>36</td>\n",
       "      <td>3.114574</td>\n",
       "      <td>6.306520</td>\n",
       "      <td>9.410297</td>\n",
       "      <td>12.190327</td>\n",
       "      <td>14.953785</td>\n",
       "      <td>18.373499</td>\n",
       "      <td>...</td>\n",
       "      <td>152.9</td>\n",
       "      <td>134.1</td>\n",
       "      <td>155.1</td>\n",
       "      <td>129.1</td>\n",
       "      <td>137.6</td>\n",
       "      <td>164.0</td>\n",
       "      <td>-27.9</td>\n",
       "      <td>165.0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>137.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_27</td>\n",
       "      <td>27</td>\n",
       "      <td>37</td>\n",
       "      <td>3.484477</td>\n",
       "      <td>6.666149</td>\n",
       "      <td>9.641264</td>\n",
       "      <td>12.575046</td>\n",
       "      <td>16.057203</td>\n",
       "      <td>17.294529</td>\n",
       "      <td>...</td>\n",
       "      <td>134.1</td>\n",
       "      <td>155.1</td>\n",
       "      <td>129.1</td>\n",
       "      <td>137.6</td>\n",
       "      <td>164.0</td>\n",
       "      <td>-27.9</td>\n",
       "      <td>165.0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>132.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_28</td>\n",
       "      <td>28</td>\n",
       "      <td>38</td>\n",
       "      <td>3.187134</td>\n",
       "      <td>6.255016</td>\n",
       "      <td>9.264649</td>\n",
       "      <td>12.774187</td>\n",
       "      <td>13.949685</td>\n",
       "      <td>16.738075</td>\n",
       "      <td>...</td>\n",
       "      <td>155.1</td>\n",
       "      <td>129.1</td>\n",
       "      <td>137.6</td>\n",
       "      <td>164.0</td>\n",
       "      <td>-27.9</td>\n",
       "      <td>165.0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>132.4</td>\n",
       "      <td>-8.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_29</td>\n",
       "      <td>29</td>\n",
       "      <td>39</td>\n",
       "      <td>3.240278</td>\n",
       "      <td>6.314395</td>\n",
       "      <td>9.809673</td>\n",
       "      <td>10.960084</td>\n",
       "      <td>13.871143</td>\n",
       "      <td>16.619455</td>\n",
       "      <td>...</td>\n",
       "      <td>129.1</td>\n",
       "      <td>137.6</td>\n",
       "      <td>164.0</td>\n",
       "      <td>-27.9</td>\n",
       "      <td>165.0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>137.0</td>\n",
       "      <td>132.4</td>\n",
       "      <td>-8.5</td>\n",
       "      <td>14.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_287</td>\n",
       "      <td>287</td>\n",
       "      <td>297</td>\n",
       "      <td>3.658035</td>\n",
       "      <td>7.193765</td>\n",
       "      <td>10.493552</td>\n",
       "      <td>13.384801</td>\n",
       "      <td>16.451735</td>\n",
       "      <td>19.236774</td>\n",
       "      <td>...</td>\n",
       "      <td>137.5</td>\n",
       "      <td>170.4</td>\n",
       "      <td>176.8</td>\n",
       "      <td>122.8</td>\n",
       "      <td>128.7</td>\n",
       "      <td>124.4</td>\n",
       "      <td>-45.3</td>\n",
       "      <td>-36.4</td>\n",
       "      <td>-36.6</td>\n",
       "      <td>-42.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>288</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_288</td>\n",
       "      <td>288</td>\n",
       "      <td>298</td>\n",
       "      <td>3.731670</td>\n",
       "      <td>7.053927</td>\n",
       "      <td>10.170790</td>\n",
       "      <td>13.284044</td>\n",
       "      <td>16.207235</td>\n",
       "      <td>17.961845</td>\n",
       "      <td>...</td>\n",
       "      <td>170.4</td>\n",
       "      <td>176.8</td>\n",
       "      <td>122.8</td>\n",
       "      <td>128.7</td>\n",
       "      <td>124.4</td>\n",
       "      <td>-45.3</td>\n",
       "      <td>-36.4</td>\n",
       "      <td>-36.6</td>\n",
       "      <td>-42.1</td>\n",
       "      <td>-44.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>289</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_289</td>\n",
       "      <td>289</td>\n",
       "      <td>299</td>\n",
       "      <td>3.465735</td>\n",
       "      <td>6.600113</td>\n",
       "      <td>9.704737</td>\n",
       "      <td>12.637264</td>\n",
       "      <td>14.372196</td>\n",
       "      <td>13.549279</td>\n",
       "      <td>...</td>\n",
       "      <td>176.8</td>\n",
       "      <td>122.8</td>\n",
       "      <td>128.7</td>\n",
       "      <td>124.4</td>\n",
       "      <td>-45.3</td>\n",
       "      <td>-36.4</td>\n",
       "      <td>-36.6</td>\n",
       "      <td>-42.1</td>\n",
       "      <td>-44.6</td>\n",
       "      <td>-23.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>290</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_290</td>\n",
       "      <td>290</td>\n",
       "      <td>300</td>\n",
       "      <td>3.370344</td>\n",
       "      <td>6.425179</td>\n",
       "      <td>9.491796</td>\n",
       "      <td>10.977680</td>\n",
       "      <td>10.150867</td>\n",
       "      <td>11.514866</td>\n",
       "      <td>...</td>\n",
       "      <td>122.8</td>\n",
       "      <td>128.7</td>\n",
       "      <td>124.4</td>\n",
       "      <td>-45.3</td>\n",
       "      <td>-36.4</td>\n",
       "      <td>-36.6</td>\n",
       "      <td>-42.1</td>\n",
       "      <td>-44.6</td>\n",
       "      <td>-23.9</td>\n",
       "      <td>-38.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>291</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_291</td>\n",
       "      <td>291</td>\n",
       "      <td>301</td>\n",
       "      <td>3.115947</td>\n",
       "      <td>6.150721</td>\n",
       "      <td>7.835769</td>\n",
       "      <td>7.266976</td>\n",
       "      <td>8.417007</td>\n",
       "      <td>10.899413</td>\n",
       "      <td>...</td>\n",
       "      <td>128.7</td>\n",
       "      <td>124.4</td>\n",
       "      <td>-45.3</td>\n",
       "      <td>-36.4</td>\n",
       "      <td>-36.6</td>\n",
       "      <td>-42.1</td>\n",
       "      <td>-44.6</td>\n",
       "      <td>-23.9</td>\n",
       "      <td>-38.4</td>\n",
       "      <td>-23.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>292</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_292</td>\n",
       "      <td>292</td>\n",
       "      <td>302</td>\n",
       "      <td>3.171775</td>\n",
       "      <td>4.799801</td>\n",
       "      <td>4.795369</td>\n",
       "      <td>5.926816</td>\n",
       "      <td>8.030334</td>\n",
       "      <td>9.561464</td>\n",
       "      <td>...</td>\n",
       "      <td>124.4</td>\n",
       "      <td>-45.3</td>\n",
       "      <td>-36.4</td>\n",
       "      <td>-36.6</td>\n",
       "      <td>-42.1</td>\n",
       "      <td>-44.6</td>\n",
       "      <td>-23.9</td>\n",
       "      <td>-38.4</td>\n",
       "      <td>-23.5</td>\n",
       "      <td>-12.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>293</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_293</td>\n",
       "      <td>293</td>\n",
       "      <td>303</td>\n",
       "      <td>3.149548</td>\n",
       "      <td>4.696058</td>\n",
       "      <td>4.844082</td>\n",
       "      <td>5.889104</td>\n",
       "      <td>8.053491</td>\n",
       "      <td>9.505167</td>\n",
       "      <td>...</td>\n",
       "      <td>-45.3</td>\n",
       "      <td>-36.4</td>\n",
       "      <td>-36.6</td>\n",
       "      <td>-42.1</td>\n",
       "      <td>-44.6</td>\n",
       "      <td>-23.9</td>\n",
       "      <td>-38.4</td>\n",
       "      <td>-23.5</td>\n",
       "      <td>-12.4</td>\n",
       "      <td>53.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>294</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_294</td>\n",
       "      <td>294</td>\n",
       "      <td>304</td>\n",
       "      <td>3.070030</td>\n",
       "      <td>4.424085</td>\n",
       "      <td>4.700176</td>\n",
       "      <td>5.880878</td>\n",
       "      <td>8.035958</td>\n",
       "      <td>9.647817</td>\n",
       "      <td>...</td>\n",
       "      <td>-36.4</td>\n",
       "      <td>-36.6</td>\n",
       "      <td>-42.1</td>\n",
       "      <td>-44.6</td>\n",
       "      <td>-23.9</td>\n",
       "      <td>-38.4</td>\n",
       "      <td>-23.5</td>\n",
       "      <td>-12.4</td>\n",
       "      <td>53.3</td>\n",
       "      <td>144.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>295</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_295</td>\n",
       "      <td>295</td>\n",
       "      <td>305</td>\n",
       "      <td>3.126979</td>\n",
       "      <td>4.879608</td>\n",
       "      <td>5.168180</td>\n",
       "      <td>6.353289</td>\n",
       "      <td>8.702175</td>\n",
       "      <td>9.866136</td>\n",
       "      <td>...</td>\n",
       "      <td>-36.6</td>\n",
       "      <td>-42.1</td>\n",
       "      <td>-44.6</td>\n",
       "      <td>-23.9</td>\n",
       "      <td>-38.4</td>\n",
       "      <td>-23.5</td>\n",
       "      <td>-12.4</td>\n",
       "      <td>53.3</td>\n",
       "      <td>144.0</td>\n",
       "      <td>-5.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_296</td>\n",
       "      <td>296</td>\n",
       "      <td>306</td>\n",
       "      <td>3.044583</td>\n",
       "      <td>4.541418</td>\n",
       "      <td>4.851020</td>\n",
       "      <td>6.342572</td>\n",
       "      <td>8.138972</td>\n",
       "      <td>9.360295</td>\n",
       "      <td>...</td>\n",
       "      <td>-42.1</td>\n",
       "      <td>-44.6</td>\n",
       "      <td>-23.9</td>\n",
       "      <td>-38.4</td>\n",
       "      <td>-23.5</td>\n",
       "      <td>-12.4</td>\n",
       "      <td>53.3</td>\n",
       "      <td>144.0</td>\n",
       "      <td>-5.1</td>\n",
       "      <td>136.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_297</td>\n",
       "      <td>297</td>\n",
       "      <td>307</td>\n",
       "      <td>3.054238</td>\n",
       "      <td>4.616457</td>\n",
       "      <td>5.135664</td>\n",
       "      <td>5.902821</td>\n",
       "      <td>7.271670</td>\n",
       "      <td>7.515536</td>\n",
       "      <td>...</td>\n",
       "      <td>-44.6</td>\n",
       "      <td>-23.9</td>\n",
       "      <td>-38.4</td>\n",
       "      <td>-23.5</td>\n",
       "      <td>-12.4</td>\n",
       "      <td>53.3</td>\n",
       "      <td>144.0</td>\n",
       "      <td>-5.1</td>\n",
       "      <td>136.4</td>\n",
       "      <td>-25.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>298</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_298</td>\n",
       "      <td>298</td>\n",
       "      <td>308</td>\n",
       "      <td>3.164052</td>\n",
       "      <td>4.926609</td>\n",
       "      <td>4.933313</td>\n",
       "      <td>5.078867</td>\n",
       "      <td>5.119645</td>\n",
       "      <td>7.506670</td>\n",
       "      <td>...</td>\n",
       "      <td>-23.9</td>\n",
       "      <td>-38.4</td>\n",
       "      <td>-23.5</td>\n",
       "      <td>-12.4</td>\n",
       "      <td>53.3</td>\n",
       "      <td>144.0</td>\n",
       "      <td>-5.1</td>\n",
       "      <td>136.4</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>-11.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_299</td>\n",
       "      <td>299</td>\n",
       "      <td>309</td>\n",
       "      <td>3.260659</td>\n",
       "      <td>4.922228</td>\n",
       "      <td>5.736267</td>\n",
       "      <td>7.216970</td>\n",
       "      <td>9.910399</td>\n",
       "      <td>10.182113</td>\n",
       "      <td>...</td>\n",
       "      <td>-38.4</td>\n",
       "      <td>-23.5</td>\n",
       "      <td>-12.4</td>\n",
       "      <td>53.3</td>\n",
       "      <td>144.0</td>\n",
       "      <td>-5.1</td>\n",
       "      <td>136.4</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>-11.4</td>\n",
       "      <td>-12.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>300</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_300</td>\n",
       "      <td>300</td>\n",
       "      <td>310</td>\n",
       "      <td>3.260212</td>\n",
       "      <td>5.690866</td>\n",
       "      <td>8.180490</td>\n",
       "      <td>11.374063</td>\n",
       "      <td>11.506120</td>\n",
       "      <td>12.517155</td>\n",
       "      <td>...</td>\n",
       "      <td>-23.5</td>\n",
       "      <td>-12.4</td>\n",
       "      <td>53.3</td>\n",
       "      <td>144.0</td>\n",
       "      <td>-5.1</td>\n",
       "      <td>136.4</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>-11.4</td>\n",
       "      <td>-12.7</td>\n",
       "      <td>-4.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_301</td>\n",
       "      <td>301</td>\n",
       "      <td>311</td>\n",
       "      <td>3.106471</td>\n",
       "      <td>5.923578</td>\n",
       "      <td>9.293181</td>\n",
       "      <td>9.765096</td>\n",
       "      <td>11.203648</td>\n",
       "      <td>11.919573</td>\n",
       "      <td>...</td>\n",
       "      <td>-12.4</td>\n",
       "      <td>53.3</td>\n",
       "      <td>144.0</td>\n",
       "      <td>-5.1</td>\n",
       "      <td>136.4</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>-11.4</td>\n",
       "      <td>-12.7</td>\n",
       "      <td>-4.2</td>\n",
       "      <td>145.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>302</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_302</td>\n",
       "      <td>302</td>\n",
       "      <td>312</td>\n",
       "      <td>3.502701</td>\n",
       "      <td>6.807855</td>\n",
       "      <td>8.236918</td>\n",
       "      <td>10.385587</td>\n",
       "      <td>11.164476</td>\n",
       "      <td>9.383610</td>\n",
       "      <td>...</td>\n",
       "      <td>53.3</td>\n",
       "      <td>144.0</td>\n",
       "      <td>-5.1</td>\n",
       "      <td>136.4</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>-11.4</td>\n",
       "      <td>-12.7</td>\n",
       "      <td>-4.2</td>\n",
       "      <td>145.6</td>\n",
       "      <td>99.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>303</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_303</td>\n",
       "      <td>303</td>\n",
       "      <td>313</td>\n",
       "      <td>3.386218</td>\n",
       "      <td>4.988798</td>\n",
       "      <td>7.562518</td>\n",
       "      <td>8.496767</td>\n",
       "      <td>7.618032</td>\n",
       "      <td>9.873635</td>\n",
       "      <td>...</td>\n",
       "      <td>144.0</td>\n",
       "      <td>-5.1</td>\n",
       "      <td>136.4</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>-11.4</td>\n",
       "      <td>-12.7</td>\n",
       "      <td>-4.2</td>\n",
       "      <td>145.6</td>\n",
       "      <td>99.7</td>\n",
       "      <td>-169.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_304</td>\n",
       "      <td>304</td>\n",
       "      <td>314</td>\n",
       "      <td>3.599923</td>\n",
       "      <td>6.628588</td>\n",
       "      <td>7.617048</td>\n",
       "      <td>8.138657</td>\n",
       "      <td>10.433300</td>\n",
       "      <td>11.576438</td>\n",
       "      <td>...</td>\n",
       "      <td>-5.1</td>\n",
       "      <td>136.4</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>-11.4</td>\n",
       "      <td>-12.7</td>\n",
       "      <td>-4.2</td>\n",
       "      <td>145.6</td>\n",
       "      <td>99.7</td>\n",
       "      <td>-169.7</td>\n",
       "      <td>131.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_305</td>\n",
       "      <td>305</td>\n",
       "      <td>315</td>\n",
       "      <td>3.035848</td>\n",
       "      <td>4.464854</td>\n",
       "      <td>5.459412</td>\n",
       "      <td>7.198708</td>\n",
       "      <td>8.070085</td>\n",
       "      <td>10.463783</td>\n",
       "      <td>...</td>\n",
       "      <td>136.4</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>-11.4</td>\n",
       "      <td>-12.7</td>\n",
       "      <td>-4.2</td>\n",
       "      <td>145.6</td>\n",
       "      <td>99.7</td>\n",
       "      <td>-169.7</td>\n",
       "      <td>131.1</td>\n",
       "      <td>152.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_306</td>\n",
       "      <td>306</td>\n",
       "      <td>316</td>\n",
       "      <td>3.058073</td>\n",
       "      <td>4.659638</td>\n",
       "      <td>5.065727</td>\n",
       "      <td>5.345264</td>\n",
       "      <td>7.688030</td>\n",
       "      <td>8.412176</td>\n",
       "      <td>...</td>\n",
       "      <td>-25.9</td>\n",
       "      <td>-11.4</td>\n",
       "      <td>-12.7</td>\n",
       "      <td>-4.2</td>\n",
       "      <td>145.6</td>\n",
       "      <td>99.7</td>\n",
       "      <td>-169.7</td>\n",
       "      <td>131.1</td>\n",
       "      <td>152.4</td>\n",
       "      <td>146.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_307</td>\n",
       "      <td>307</td>\n",
       "      <td>317</td>\n",
       "      <td>3.330867</td>\n",
       "      <td>4.866722</td>\n",
       "      <td>4.390885</td>\n",
       "      <td>6.715438</td>\n",
       "      <td>7.529260</td>\n",
       "      <td>9.017134</td>\n",
       "      <td>...</td>\n",
       "      <td>-11.4</td>\n",
       "      <td>-12.7</td>\n",
       "      <td>-4.2</td>\n",
       "      <td>145.6</td>\n",
       "      <td>99.7</td>\n",
       "      <td>-169.7</td>\n",
       "      <td>131.1</td>\n",
       "      <td>152.4</td>\n",
       "      <td>146.6</td>\n",
       "      <td>-36.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_308</td>\n",
       "      <td>308</td>\n",
       "      <td>318</td>\n",
       "      <td>3.447547</td>\n",
       "      <td>5.278381</td>\n",
       "      <td>8.418408</td>\n",
       "      <td>10.241159</td>\n",
       "      <td>12.185719</td>\n",
       "      <td>13.711534</td>\n",
       "      <td>...</td>\n",
       "      <td>-12.7</td>\n",
       "      <td>-4.2</td>\n",
       "      <td>145.6</td>\n",
       "      <td>99.7</td>\n",
       "      <td>-169.7</td>\n",
       "      <td>131.1</td>\n",
       "      <td>152.4</td>\n",
       "      <td>146.6</td>\n",
       "      <td>-36.3</td>\n",
       "      <td>-30.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_309</td>\n",
       "      <td>309</td>\n",
       "      <td>319</td>\n",
       "      <td>3.579916</td>\n",
       "      <td>6.860459</td>\n",
       "      <td>9.529537</td>\n",
       "      <td>12.067354</td>\n",
       "      <td>14.377881</td>\n",
       "      <td>16.408253</td>\n",
       "      <td>...</td>\n",
       "      <td>-4.2</td>\n",
       "      <td>145.6</td>\n",
       "      <td>99.7</td>\n",
       "      <td>-169.7</td>\n",
       "      <td>131.1</td>\n",
       "      <td>152.4</td>\n",
       "      <td>146.6</td>\n",
       "      <td>-36.3</td>\n",
       "      <td>-30.7</td>\n",
       "      <td>-32.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>310</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_310</td>\n",
       "      <td>310</td>\n",
       "      <td>320</td>\n",
       "      <td>3.340582</td>\n",
       "      <td>6.083349</td>\n",
       "      <td>8.927070</td>\n",
       "      <td>11.533739</td>\n",
       "      <td>13.824341</td>\n",
       "      <td>15.526217</td>\n",
       "      <td>...</td>\n",
       "      <td>145.6</td>\n",
       "      <td>99.7</td>\n",
       "      <td>-169.7</td>\n",
       "      <td>131.1</td>\n",
       "      <td>152.4</td>\n",
       "      <td>146.6</td>\n",
       "      <td>-36.3</td>\n",
       "      <td>-30.7</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-40.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_311</td>\n",
       "      <td>311</td>\n",
       "      <td>321</td>\n",
       "      <td>3.628062</td>\n",
       "      <td>6.973111</td>\n",
       "      <td>10.041386</td>\n",
       "      <td>12.652610</td>\n",
       "      <td>14.663662</td>\n",
       "      <td>15.269337</td>\n",
       "      <td>...</td>\n",
       "      <td>99.7</td>\n",
       "      <td>-169.7</td>\n",
       "      <td>131.1</td>\n",
       "      <td>152.4</td>\n",
       "      <td>146.6</td>\n",
       "      <td>-36.3</td>\n",
       "      <td>-30.7</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-40.1</td>\n",
       "      <td>-55.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>312</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_312</td>\n",
       "      <td>312</td>\n",
       "      <td>322</td>\n",
       "      <td>3.395794</td>\n",
       "      <td>6.610985</td>\n",
       "      <td>9.404448</td>\n",
       "      <td>11.707784</td>\n",
       "      <td>12.567925</td>\n",
       "      <td>10.352145</td>\n",
       "      <td>...</td>\n",
       "      <td>-169.7</td>\n",
       "      <td>131.1</td>\n",
       "      <td>152.4</td>\n",
       "      <td>146.6</td>\n",
       "      <td>-36.3</td>\n",
       "      <td>-30.7</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-40.1</td>\n",
       "      <td>-55.7</td>\n",
       "      <td>-18.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>313</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_313</td>\n",
       "      <td>313</td>\n",
       "      <td>323</td>\n",
       "      <td>3.558899</td>\n",
       "      <td>6.561012</td>\n",
       "      <td>9.231917</td>\n",
       "      <td>10.175973</td>\n",
       "      <td>8.621835</td>\n",
       "      <td>10.027635</td>\n",
       "      <td>...</td>\n",
       "      <td>131.1</td>\n",
       "      <td>152.4</td>\n",
       "      <td>146.6</td>\n",
       "      <td>-36.3</td>\n",
       "      <td>-30.7</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-40.1</td>\n",
       "      <td>-55.7</td>\n",
       "      <td>-18.7</td>\n",
       "      <td>122.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_314</td>\n",
       "      <td>314</td>\n",
       "      <td>324</td>\n",
       "      <td>3.025128</td>\n",
       "      <td>5.838418</td>\n",
       "      <td>7.025460</td>\n",
       "      <td>6.238036</td>\n",
       "      <td>7.632911</td>\n",
       "      <td>10.101269</td>\n",
       "      <td>...</td>\n",
       "      <td>152.4</td>\n",
       "      <td>146.6</td>\n",
       "      <td>-36.3</td>\n",
       "      <td>-30.7</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-40.1</td>\n",
       "      <td>-55.7</td>\n",
       "      <td>-18.7</td>\n",
       "      <td>122.7</td>\n",
       "      <td>155.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_315</td>\n",
       "      <td>315</td>\n",
       "      <td>325</td>\n",
       "      <td>2.962844</td>\n",
       "      <td>4.578427</td>\n",
       "      <td>4.941282</td>\n",
       "      <td>6.013155</td>\n",
       "      <td>8.022788</td>\n",
       "      <td>9.559601</td>\n",
       "      <td>...</td>\n",
       "      <td>146.6</td>\n",
       "      <td>-36.3</td>\n",
       "      <td>-30.7</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-40.1</td>\n",
       "      <td>-55.7</td>\n",
       "      <td>-18.7</td>\n",
       "      <td>122.7</td>\n",
       "      <td>155.4</td>\n",
       "      <td>30.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>12asA0</td>\n",
       "      <td>chip_316</td>\n",
       "      <td>316</td>\n",
       "      <td>326</td>\n",
       "      <td>3.147943</td>\n",
       "      <td>4.477601</td>\n",
       "      <td>4.455284</td>\n",
       "      <td>5.833719</td>\n",
       "      <td>8.102034</td>\n",
       "      <td>9.226256</td>\n",
       "      <td>...</td>\n",
       "      <td>-36.3</td>\n",
       "      <td>-30.7</td>\n",
       "      <td>-32.9</td>\n",
       "      <td>-40.1</td>\n",
       "      <td>-55.7</td>\n",
       "      <td>-18.7</td>\n",
       "      <td>122.7</td>\n",
       "      <td>155.4</td>\n",
       "      <td>30.7</td>\n",
       "      <td>132.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>317 rows × 94 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    protein_id   chip_id  start  stop    dist_0    dist_1     dist_2  \\\n",
       "0       12asA0    chip_0      0    10  2.931498  4.466173   4.830971   \n",
       "1       12asA0    chip_1      1    11  2.977764  4.556571   4.825561   \n",
       "2       12asA0    chip_2      2    12  3.011051  4.573128   4.944533   \n",
       "3       12asA0    chip_3      3    13  3.055943  4.643566   4.900433   \n",
       "4       12asA0    chip_4      4    14  3.033617  4.514310   4.787584   \n",
       "5       12asA0    chip_5      5    15  3.223323  4.594050   5.066024   \n",
       "6       12asA0    chip_6      6    16  2.976352  4.657997   5.042452   \n",
       "7       12asA0    chip_7      7    17  2.983589  4.640706   5.149854   \n",
       "8       12asA0    chip_8      8    18  3.081327  4.675354   5.066069   \n",
       "9       12asA0    chip_9      9    19  3.076244  4.648451   5.031323   \n",
       "10      12asA0   chip_10     10    20  3.038332  4.728816   4.798806   \n",
       "11      12asA0   chip_11     11    21  3.010424  4.472245   5.071977   \n",
       "12      12asA0   chip_12     12    22  3.148242  4.942146   5.011508   \n",
       "13      12asA0   chip_13     13    23  2.977663  4.486660   4.839495   \n",
       "14      12asA0   chip_14     14    24  3.080211  4.458592   4.706429   \n",
       "15      12asA0   chip_15     15    25  2.932920  4.282657   4.616333   \n",
       "16      12asA0   chip_16     16    26  3.113236  4.620432   5.094796   \n",
       "17      12asA0   chip_17     17    27  3.003047  4.556078   5.047412   \n",
       "18      12asA0   chip_18     18    28  3.142842  4.863575   5.154842   \n",
       "19      12asA0   chip_19     19    29  3.128385  4.869174   5.030503   \n",
       "20      12asA0   chip_20     20    30  3.220644  4.942284   4.694743   \n",
       "21      12asA0   chip_21     21    31  3.162164  4.995806   6.318177   \n",
       "22      12asA0   chip_22     22    32  3.548883  6.188423   8.440619   \n",
       "23      12asA0   chip_23     23    33  3.039205  5.773140   9.032375   \n",
       "24      12asA0   chip_24     24    34  3.400494  6.638870   9.738158   \n",
       "25      12asA0   chip_25     25    35  3.336457  6.442289   9.494033   \n",
       "26      12asA0   chip_26     26    36  3.114574  6.306520   9.410297   \n",
       "27      12asA0   chip_27     27    37  3.484477  6.666149   9.641264   \n",
       "28      12asA0   chip_28     28    38  3.187134  6.255016   9.264649   \n",
       "29      12asA0   chip_29     29    39  3.240278  6.314395   9.809673   \n",
       "..         ...       ...    ...   ...       ...       ...        ...   \n",
       "287     12asA0  chip_287    287   297  3.658035  7.193765  10.493552   \n",
       "288     12asA0  chip_288    288   298  3.731670  7.053927  10.170790   \n",
       "289     12asA0  chip_289    289   299  3.465735  6.600113   9.704737   \n",
       "290     12asA0  chip_290    290   300  3.370344  6.425179   9.491796   \n",
       "291     12asA0  chip_291    291   301  3.115947  6.150721   7.835769   \n",
       "292     12asA0  chip_292    292   302  3.171775  4.799801   4.795369   \n",
       "293     12asA0  chip_293    293   303  3.149548  4.696058   4.844082   \n",
       "294     12asA0  chip_294    294   304  3.070030  4.424085   4.700176   \n",
       "295     12asA0  chip_295    295   305  3.126979  4.879608   5.168180   \n",
       "296     12asA0  chip_296    296   306  3.044583  4.541418   4.851020   \n",
       "297     12asA0  chip_297    297   307  3.054238  4.616457   5.135664   \n",
       "298     12asA0  chip_298    298   308  3.164052  4.926609   4.933313   \n",
       "299     12asA0  chip_299    299   309  3.260659  4.922228   5.736267   \n",
       "300     12asA0  chip_300    300   310  3.260212  5.690866   8.180490   \n",
       "301     12asA0  chip_301    301   311  3.106471  5.923578   9.293181   \n",
       "302     12asA0  chip_302    302   312  3.502701  6.807855   8.236918   \n",
       "303     12asA0  chip_303    303   313  3.386218  4.988798   7.562518   \n",
       "304     12asA0  chip_304    304   314  3.599923  6.628588   7.617048   \n",
       "305     12asA0  chip_305    305   315  3.035848  4.464854   5.459412   \n",
       "306     12asA0  chip_306    306   316  3.058073  4.659638   5.065727   \n",
       "307     12asA0  chip_307    307   317  3.330867  4.866722   4.390885   \n",
       "308     12asA0  chip_308    308   318  3.447547  5.278381   8.418408   \n",
       "309     12asA0  chip_309    309   319  3.579916  6.860459   9.529537   \n",
       "310     12asA0  chip_310    310   320  3.340582  6.083349   8.927070   \n",
       "311     12asA0  chip_311    311   321  3.628062  6.973111  10.041386   \n",
       "312     12asA0  chip_312    312   322  3.395794  6.610985   9.404448   \n",
       "313     12asA0  chip_313    313   323  3.558899  6.561012   9.231917   \n",
       "314     12asA0  chip_314    314   324  3.025128  5.838418   7.025460   \n",
       "315     12asA0  chip_315    315   325  2.962844  4.578427   4.941282   \n",
       "316     12asA0  chip_316    316   326  3.147943  4.477601   4.455284   \n",
       "\n",
       "        dist_3     dist_4     dist_5  ...  psi_0  psi_1  psi_2  psi_3  psi_4  \\\n",
       "0     5.929355   8.004140   9.493499  ...  157.9  -56.8  -51.1  -58.3  -39.3   \n",
       "1     5.993463   8.187191   9.461169  ...  -56.8  -51.1  -58.3  -39.3  -33.5   \n",
       "2     6.271376   8.255353   9.777748  ...  -51.1  -58.3  -39.3  -33.5  -32.3   \n",
       "3     6.196664   8.368240   9.765861  ...  -58.3  -39.3  -33.5  -32.3  -57.3   \n",
       "4     6.232026   8.312329   9.773744  ...  -39.3  -33.5  -32.3  -57.3  -50.5   \n",
       "5     6.347551   8.505913   9.938494  ...  -33.5  -32.3  -57.3  -50.5  -41.7   \n",
       "6     6.315280   8.412719   9.836761  ...  -32.3  -57.3  -50.5  -41.7  -44.9   \n",
       "7     6.408232   8.367692   9.668061  ...  -57.3  -50.5  -41.7  -44.9  -54.2   \n",
       "8     6.178646   8.111391   9.779182  ...  -50.5  -41.7  -44.9  -54.2  -29.9   \n",
       "9     5.955137   8.097498   9.558247  ...  -41.7  -44.9  -54.2  -29.9  -53.6   \n",
       "10    5.939553   7.934079   9.386290  ...  -44.9  -54.2  -29.9  -53.6  -50.3   \n",
       "11    5.966181   7.969033   9.478682  ...  -54.2  -29.9  -53.6  -50.3  -28.8   \n",
       "12    6.109199   8.375867   9.568907  ...  -29.9  -53.6  -50.3  -28.8  -42.9   \n",
       "13    6.330968   8.282268   9.775724  ...  -53.6  -50.3  -28.8  -42.9  -32.9   \n",
       "14    6.142321   8.378738   9.686699  ...  -50.3  -28.8  -42.9  -32.9  -34.3   \n",
       "15    6.260442   8.352125   9.567734  ...  -28.8  -42.9  -32.9  -34.3  -35.7   \n",
       "16    6.513730   8.500165   9.696307  ...  -42.9  -32.9  -34.3  -35.7  -37.6   \n",
       "17    6.118340   7.877437   9.032131  ...  -32.9  -34.3  -35.7  -37.6  -40.5   \n",
       "18    5.744147   7.385431   8.627468  ...  -34.3  -35.7  -37.6  -40.5  -28.9   \n",
       "19    5.046117   5.781323   5.672482  ...  -35.7  -37.6  -40.5  -28.9    0.3   \n",
       "20    4.624504   5.171173   7.949272  ...  -37.6  -40.5  -28.9    0.3   44.3   \n",
       "21    8.039869  11.060770  13.959582  ...  -40.5  -28.9    0.3   44.3  137.6   \n",
       "22   11.621278  14.576628  17.376261  ...  -28.9    0.3   44.3  137.6  152.9   \n",
       "23   12.072133  14.798532  17.520407  ...    0.3   44.3  137.6  152.9  134.1   \n",
       "24   12.659022  15.550603  18.014523  ...   44.3  137.6  152.9  134.1  155.1   \n",
       "25   12.483414  15.055290  17.628443  ...  137.6  152.9  134.1  155.1  129.1   \n",
       "26   12.190327  14.953785  18.373499  ...  152.9  134.1  155.1  129.1  137.6   \n",
       "27   12.575046  16.057203  17.294529  ...  134.1  155.1  129.1  137.6  164.0   \n",
       "28   12.774187  13.949685  16.738075  ...  155.1  129.1  137.6  164.0  -27.9   \n",
       "29   10.960084  13.871143  16.619455  ...  129.1  137.6  164.0  -27.9  165.0   \n",
       "..         ...        ...        ...  ...    ...    ...    ...    ...    ...   \n",
       "287  13.384801  16.451735  19.236774  ...  137.5  170.4  176.8  122.8  128.7   \n",
       "288  13.284044  16.207235  17.961845  ...  170.4  176.8  122.8  128.7  124.4   \n",
       "289  12.637264  14.372196  13.549279  ...  176.8  122.8  128.7  124.4  -45.3   \n",
       "290  10.977680  10.150867  11.514866  ...  122.8  128.7  124.4  -45.3  -36.4   \n",
       "291   7.266976   8.417007  10.899413  ...  128.7  124.4  -45.3  -36.4  -36.6   \n",
       "292   5.926816   8.030334   9.561464  ...  124.4  -45.3  -36.4  -36.6  -42.1   \n",
       "293   5.889104   8.053491   9.505167  ...  -45.3  -36.4  -36.6  -42.1  -44.6   \n",
       "294   5.880878   8.035958   9.647817  ...  -36.4  -36.6  -42.1  -44.6  -23.9   \n",
       "295   6.353289   8.702175   9.866136  ...  -36.6  -42.1  -44.6  -23.9  -38.4   \n",
       "296   6.342572   8.138972   9.360295  ...  -42.1  -44.6  -23.9  -38.4  -23.5   \n",
       "297   5.902821   7.271670   7.515536  ...  -44.6  -23.9  -38.4  -23.5  -12.4   \n",
       "298   5.078867   5.119645   7.506670  ...  -23.9  -38.4  -23.5  -12.4   53.3   \n",
       "299   7.216970   9.910399  10.182113  ...  -38.4  -23.5  -12.4   53.3  144.0   \n",
       "300  11.374063  11.506120  12.517155  ...  -23.5  -12.4   53.3  144.0   -5.1   \n",
       "301   9.765096  11.203648  11.919573  ...  -12.4   53.3  144.0   -5.1  136.4   \n",
       "302  10.385587  11.164476   9.383610  ...   53.3  144.0   -5.1  136.4  -25.9   \n",
       "303   8.496767   7.618032   9.873635  ...  144.0   -5.1  136.4  -25.9  -11.4   \n",
       "304   8.138657  10.433300  11.576438  ...   -5.1  136.4  -25.9  -11.4  -12.7   \n",
       "305   7.198708   8.070085  10.463783  ...  136.4  -25.9  -11.4  -12.7   -4.2   \n",
       "306   5.345264   7.688030   8.412176  ...  -25.9  -11.4  -12.7   -4.2  145.6   \n",
       "307   6.715438   7.529260   9.017134  ...  -11.4  -12.7   -4.2  145.6   99.7   \n",
       "308  10.241159  12.185719  13.711534  ...  -12.7   -4.2  145.6   99.7 -169.7   \n",
       "309  12.067354  14.377881  16.408253  ...   -4.2  145.6   99.7 -169.7  131.1   \n",
       "310  11.533739  13.824341  15.526217  ...  145.6   99.7 -169.7  131.1  152.4   \n",
       "311  12.652610  14.663662  15.269337  ...   99.7 -169.7  131.1  152.4  146.6   \n",
       "312  11.707784  12.567925  10.352145  ... -169.7  131.1  152.4  146.6  -36.3   \n",
       "313  10.175973   8.621835  10.027635  ...  131.1  152.4  146.6  -36.3  -30.7   \n",
       "314   6.238036   7.632911  10.101269  ...  152.4  146.6  -36.3  -30.7  -32.9   \n",
       "315   6.013155   8.022788   9.559601  ...  146.6  -36.3  -30.7  -32.9  -40.1   \n",
       "316   5.833719   8.102034   9.226256  ...  -36.3  -30.7  -32.9  -40.1  -55.7   \n",
       "\n",
       "     psi_5  psi_6  psi_7  psi_8  psi_9  \n",
       "0    -33.5  -32.3  -57.3  -50.5  -41.7  \n",
       "1    -32.3  -57.3  -50.5  -41.7  -44.9  \n",
       "2    -57.3  -50.5  -41.7  -44.9  -54.2  \n",
       "3    -50.5  -41.7  -44.9  -54.2  -29.9  \n",
       "4    -41.7  -44.9  -54.2  -29.9  -53.6  \n",
       "5    -44.9  -54.2  -29.9  -53.6  -50.3  \n",
       "6    -54.2  -29.9  -53.6  -50.3  -28.8  \n",
       "7    -29.9  -53.6  -50.3  -28.8  -42.9  \n",
       "8    -53.6  -50.3  -28.8  -42.9  -32.9  \n",
       "9    -50.3  -28.8  -42.9  -32.9  -34.3  \n",
       "10   -28.8  -42.9  -32.9  -34.3  -35.7  \n",
       "11   -42.9  -32.9  -34.3  -35.7  -37.6  \n",
       "12   -32.9  -34.3  -35.7  -37.6  -40.5  \n",
       "13   -34.3  -35.7  -37.6  -40.5  -28.9  \n",
       "14   -35.7  -37.6  -40.5  -28.9    0.3  \n",
       "15   -37.6  -40.5  -28.9    0.3   44.3  \n",
       "16   -40.5  -28.9    0.3   44.3  137.6  \n",
       "17   -28.9    0.3   44.3  137.6  152.9  \n",
       "18     0.3   44.3  137.6  152.9  134.1  \n",
       "19    44.3  137.6  152.9  134.1  155.1  \n",
       "20   137.6  152.9  134.1  155.1  129.1  \n",
       "21   152.9  134.1  155.1  129.1  137.6  \n",
       "22   134.1  155.1  129.1  137.6  164.0  \n",
       "23   155.1  129.1  137.6  164.0  -27.9  \n",
       "24   129.1  137.6  164.0  -27.9  165.0  \n",
       "25   137.6  164.0  -27.9  165.0  151.0  \n",
       "26   164.0  -27.9  165.0  151.0  137.0  \n",
       "27   -27.9  165.0  151.0  137.0  132.4  \n",
       "28   165.0  151.0  137.0  132.4   -8.5  \n",
       "29   151.0  137.0  132.4   -8.5   14.2  \n",
       "..     ...    ...    ...    ...    ...  \n",
       "287  124.4  -45.3  -36.4  -36.6  -42.1  \n",
       "288  -45.3  -36.4  -36.6  -42.1  -44.6  \n",
       "289  -36.4  -36.6  -42.1  -44.6  -23.9  \n",
       "290  -36.6  -42.1  -44.6  -23.9  -38.4  \n",
       "291  -42.1  -44.6  -23.9  -38.4  -23.5  \n",
       "292  -44.6  -23.9  -38.4  -23.5  -12.4  \n",
       "293  -23.9  -38.4  -23.5  -12.4   53.3  \n",
       "294  -38.4  -23.5  -12.4   53.3  144.0  \n",
       "295  -23.5  -12.4   53.3  144.0   -5.1  \n",
       "296  -12.4   53.3  144.0   -5.1  136.4  \n",
       "297   53.3  144.0   -5.1  136.4  -25.9  \n",
       "298  144.0   -5.1  136.4  -25.9  -11.4  \n",
       "299   -5.1  136.4  -25.9  -11.4  -12.7  \n",
       "300  136.4  -25.9  -11.4  -12.7   -4.2  \n",
       "301  -25.9  -11.4  -12.7   -4.2  145.6  \n",
       "302  -11.4  -12.7   -4.2  145.6   99.7  \n",
       "303  -12.7   -4.2  145.6   99.7 -169.7  \n",
       "304   -4.2  145.6   99.7 -169.7  131.1  \n",
       "305  145.6   99.7 -169.7  131.1  152.4  \n",
       "306   99.7 -169.7  131.1  152.4  146.6  \n",
       "307 -169.7  131.1  152.4  146.6  -36.3  \n",
       "308  131.1  152.4  146.6  -36.3  -30.7  \n",
       "309  152.4  146.6  -36.3  -30.7  -32.9  \n",
       "310  146.6  -36.3  -30.7  -32.9  -40.1  \n",
       "311  -36.3  -30.7  -32.9  -40.1  -55.7  \n",
       "312  -30.7  -32.9  -40.1  -55.7  -18.7  \n",
       "313  -32.9  -40.1  -55.7  -18.7  122.7  \n",
       "314  -40.1  -55.7  -18.7  122.7  155.4  \n",
       "315  -55.7  -18.7  122.7  155.4   30.7  \n",
       "316  -18.7  122.7  155.4   30.7  132.1  \n",
       "\n",
       "[317 rows x 94 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get_ground_truth_api(pdb_list_y[0][0:4])\n",
    "df\n",
    "# df.to_csv(\"datasets/protein_dssp_api.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this will add majority secondary structure to the dataframe\n",
    "# through loading it from file\n",
    "# i've dummied it out through the if statement but its still usefull\n",
    "if False:\n",
    "    dataset=pandas.read_csv(\"datasets/protein_full_gt.csv\")\n",
    "\n",
    "    compute_majority_ss(dataset)\n",
    "    dataset.to_csv(\"datasets/protein_full_gt_classed.csv\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# these generate triangular numbers from a row number\n",
    "# and row numbers from a triangular number\n",
    "# not sure if I use them anymore but here they are\n",
    "def tri_num(x):\n",
    "    return x*(x+1)/2\n",
    "    \n",
    "def inv_tri_num(x):\n",
    "    for i in range(0,x):\n",
    "        if tri_num(i)>x:\n",
    "            return None\n",
    "        if tri_num(i)==x:\n",
    "            return i    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this draws a gif from a number of chips l\n",
    "# which will iterate through the rows of a datafile\n",
    "# and make each row (and the chip represented by that row)\n",
    "# into a frame of the gif\n",
    "def gif_from_chips(dataset,l,filename):    \n",
    "\n",
    "    ncol=len([c for c in dataset.columns if c[0:5]==\"dist_\"])\n",
    "    \n",
    "    df=dataset[[\"dist_\"+str(x) for x in range(0,ncol)]]\n",
    "    \n",
    "    gif=[]\n",
    "\n",
    "    for n in range(0,l):\n",
    "        \n",
    "        chpx=list(df.iloc[n,:])\n",
    "\n",
    "        chip_l=inv_tri_num(ncol)+1\n",
    "        chp_im=np.zeros((chip_l,chip_l))\n",
    "        triu=np.triu_indices(chip_l,1)\n",
    "        tri_ind=[(triu[0][i],triu[1][i]) for i in range(0,len(triu[0]))]\n",
    "\n",
    "        for i in range(0,len(tri_ind)):\n",
    "             chp_im[tri_ind[i][0],tri_ind[i][1]]=chpx[i]\n",
    "\n",
    "        im=plt.imshow(chp_im)    \n",
    "\n",
    "        \n",
    "        x=im.make_image(\"AGG\")[0]\n",
    "        x=np.flipud(x)\n",
    "\n",
    "        gif.append(np.array(x))\n",
    "\n",
    "    imageio.mimsave(filename+'.gif', gif, fps=60)\n",
    "\n",
    "    \n",
    "    \n",
    "# this is a function which will draw the average chip\n",
    "# of a dataset\n",
    "# i.e. draw a chip whose pixels are each an average of the\n",
    "# pixels in that position on each chip\n",
    "def draw_avg_chip(dataset, name, max_dist):\n",
    "    \n",
    "    ncol=len([c for c in dataset.columns if c[0:5]==\"dist_\"])\n",
    "    \n",
    "    df=dataset[[\"dist_\"+str(x) for x in range(0,ncol)]]\n",
    "\n",
    "    chpx=list(df.mean(0))\n",
    "    \n",
    "#     print(ncol)\n",
    "    chip_l=inv_tri_num(ncol)+1\n",
    "    chp_im=np.zeros((chip_l,chip_l))\n",
    "    triu=np.triu_indices(chip_l,1)\n",
    "    tri_ind=[(triu[0][i],triu[1][i]) for i in range(0,len(triu[0]))]\n",
    "    \n",
    "    for i in range(0,len(tri_ind)):\n",
    "         chp_im[tri_ind[i][0],tri_ind[i][1]]=chpx[i]\n",
    "\n",
    "    im=plt.imshow(chp_im,vmax=max_dist)\n",
    "    \n",
    "    x=im.make_image(\"AGG\")[0]\n",
    "    x=np.flipud(x)\n",
    "    \n",
    "    imageio.imsave(\"plots/\"+name+\".png\", x)\n",
    "    \n",
    "# as above, but draws a lineplot instead of a chip\n",
    "def draw_avg_dist_lineplot(dataset, name,max_dist):\n",
    "    ncol=len([c for c in dataset.columns if c[0:5]==\"dist_\"])\n",
    "    \n",
    "    df=dataset[[\"dist_\"+str(x) for x in range(0,ncol)]]\n",
    "\n",
    "    chpx=list(df.mean(0))\n",
    "    \n",
    "    chip_l=inv_tri_num(ncol)+1\n",
    "    chp_im=np.empty((chip_l,chip_l))\n",
    "    chp_im[:]=np.nan\n",
    "    chp_im[:,0]=0\n",
    "    triu=np.triu_indices(chip_l,1)\n",
    "    tri_ind=[(triu[0][i],triu[1][i]) for i in range(0,len(triu[0]))]\n",
    "    \n",
    "    \n",
    "    for i in range(0,len(tri_ind)):\n",
    "         chp_im[tri_ind[i][0],tri_ind[i][1]-tri_ind[i][0]]=chpx[i]\n",
    "\n",
    "    plt.ylim((0, max_dist))\n",
    "    plt.plot(np.nanmean(chp_im,0))\n",
    "    \n",
    "    plt.savefig(\"plots/lineg/\"+name+\".png\")\n",
    "    plt.cla()\n",
    "\n",
    "# as above, but draws a densityplot instead of a lineplot\n",
    "def draw_avg_dist_densityplot(dataset, name,max_dist):\n",
    "    ncol=len([c for c in dataset.columns if c[0:5]==\"dist_\"])\n",
    "    \n",
    "    df=dataset[[\"dist_\"+str(x) for x in range(0,ncol)]]\n",
    "    \n",
    "    chpx=list(df.mean(0))\n",
    "    \n",
    "    chip_l=inv_tri_num(ncol)+1\n",
    "    chp_im=np.empty((chip_l,chip_l))\n",
    "    chp_im[:]=np.nan\n",
    "    chp_im[:,0]=0\n",
    "    triu=np.triu_indices(chip_l,1)\n",
    "    seq_dist=[triu[1][i]-triu[0][i] for i in range(0,len(triu[0]))]\n",
    "    \n",
    "    cx=[]\n",
    "    cy=[]\n",
    "\n",
    "    for index, row in df.iterrows():\n",
    "        for i in range(0,len(seq_dist)):\n",
    "            cx.append(seq_dist[i])\n",
    "            cy.append(row[i])\n",
    "\n",
    "    plt.ylim((0, max_dist))\n",
    "\n",
    "    plt.hist2d(cx,cy)\n",
    " \n",
    "    plt.savefig(\"plots/2dhist/\"+name+\".png\")\n",
    "    plt.cla()\n",
    "    \n",
    "    \n",
    "    # as above, but draws a densityplot which does not have an\n",
    "    # artificially brightened area for lower amino acid sequence distances\n",
    "    # because we average across sequence distance per chip\n",
    "    # before averaging across chips\n",
    "def draw_avg_dist_densityplot2(dataset, name,max_dist):\n",
    "    ncol=len([c for c in dataset.columns if c[0:5]==\"dist_\"])\n",
    "    \n",
    "    df=dataset[[\"dist_\"+str(x) for x in range(0,ncol)]]\n",
    "        \n",
    "    chip_l=inv_tri_num(ncol)+1\n",
    "    chp_im=np.empty((chip_l,chip_l))\n",
    "    chp_im[:]=np.nan\n",
    "    chp_im[:,0]=0\n",
    "    triu=np.triu_indices(chip_l,1)\n",
    "\n",
    "    tri_ind=[(triu[0][i],triu[1][i]) for i in range(0,len(triu[0]))]\n",
    "    \n",
    "    cx=[]\n",
    "    cy=[]\n",
    "    \n",
    "    for index, row in df.iterrows():\n",
    "        for i in range(0,chip_l):\n",
    "            chp_im[tri_ind[i][0],tri_ind[i][1]-tri_ind[i][0]]=row[i]\n",
    "        \n",
    "        cx+=(list(range(1,chip_l)))\n",
    "        cy+=list(np.nanmean(chp_im,0))[1:]\n",
    "\n",
    "    xbins=np.linspace(1,chip_l,chip_l)\n",
    "    ybins=np.linspace(1,max_dist,chip_l*2)\n",
    "    \n",
    "\n",
    "    plt.hist2d(cx,cy,bins = (xbins,ybins))\n",
    " \n",
    "    plt.savefig(\"plots/2dhist/\"+name+\".png\")\n",
    "    \n",
    "def get_avg_distances(dataset):\n",
    "    ncol=len([c for c in dataset.columns if c[0:5]==\"dist_\"])\n",
    "    \n",
    "    df=dataset[[\"dist_\"+str(x) for x in range(0,ncol)]]\n",
    "\n",
    "    return(list(df.mean(0)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ground_truth_api(full_pdb_id, partial=False):\n",
    "    pdb_id = full_pdb_id[0:4].upper()\n",
    "    get_pdb_file(pdb_id)\n",
    "    struct = get_pdb_structure(pdb_id)\n",
    "    ppb = PPBuilder()\n",
    "    struct_info = []\n",
    "    seq_count = 0\n",
    "    for model in struct:\n",
    "        file = os.path.join(pdb_path, f'{pdb_id}.pdb')\n",
    "        pdb_id_to_dssp_file(pdb_id)\n",
    "        dssp = DSSP(model=model, in_file=os.path.join(dssp_path, f'{pdb_id}.dssp'), file_type='DSSP')\n",
    "        #dssp = DSSP(model=model, in_file=file)\n",
    "        for i, chain in enumerate(model):\n",
    "            for pp in ppb.build_peptides(chain, aa_only=False):\n",
    "                seq = pp.get_sequence()\n",
    "                for j, residue in enumerate(pp):\n",
    "                    #print(residue.id[1], seq[i])\n",
    "                    #print(dssp.keys()[j % (i + 1)])\n",
    "                    key = list(dssp.keys())[seq_count]\n",
    "                    seq_count += 1\n",
    "                    dssp_info = dssp[key]\n",
    "                    #print(dssp_info)\n",
    "                    #dssp_id = dssp_info[0]\n",
    "                    amino_acid = dssp_info[1]\n",
    "                    sec_struct = dssp_info[2]\n",
    "                    solv_acc = dssp_info[3]\n",
    "                    phi = dssp_info[4]\n",
    "                    psi = dssp_info[5]\n",
    "                    # Keys 6 through 13 is bonding energy / relidx (no clue what this is)\n",
    "                    #print(dssp[dssp.keys()[i]])\n",
    "                    struct_info.append([model.full_id[0], model.full_id[1], chain.id,\n",
    "                    residue.resname, residue.id[1], amino_acid, sec_struct, solv_acc, phi, psi])\n",
    "    info_df = pd.DataFrame(struct_info, \n",
    "        columns=['Model_Name', 'Model_ID', 'Chain', 'Residue_Name',\n",
    "        'Residue_ID', 'Amino_Acid', 'Secondary_Structure', 'Solvent_Accessability', \n",
    "        'Phi', 'Psi'])\n",
    "    if partial:    \n",
    "        seq_bool = get_seq_alignment(full_pdb_id)\n",
    "        fin_df = info_df[info_df['Chain'] == full_pdb_id[4].upper()]\n",
    "        return fin_df[seq_bool]\n",
    "    else:\n",
    "        return info_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "dataset=pandas.read_csv(\"datasets/protein_test_gt.csv\")\n",
    "\n",
    "\n",
    "draw_avg_dist_densityplot2(dataset.iloc[0:100,:],\"test\",30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# demonstrate the gif_from_chips function\n",
    "# which draws an animated gif of a number of chips from a saved dataset\n",
    "\n",
    "# gif_from_chips(dataset,500,\"chip_draw\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this draws the avg distance for each pixel in all chips\n",
    "# within each point in a thd dendrogram\n",
    "\n",
    "def draw_thd_segmentations(folder,prefix,dataset):\n",
    "    tda_dir=\"datasets/\"+folder\n",
    "\n",
    "    thds=[n for n in os.listdir(tda_dir) if n[0:len(prefix)]==prefix]\n",
    "        \n",
    "    df_list=[]\n",
    "    names=[]\n",
    "    maxes=[]\n",
    "    \n",
    "    for model in thds:\n",
    "        \n",
    "        with open(tda_dir+model) as jsonf:\n",
    "            a=json.load(jsonf)\n",
    "            \n",
    "            df=dataset.iloc[[int(m) for m in a['rowList']],:]\n",
    "            \n",
    "#             print(model)\n",
    "#             print([int(m) for m in a['rowList']])\n",
    "            \n",
    "            maxes.append(max(get_avg_distances(df)))\n",
    "            df_list.append(df)\n",
    "            names.append(model[len(prefix):len(model)-5])\n",
    "            \n",
    "    max_dist=max(maxes)\n",
    "    for i in range(0,len(df_list)):\n",
    "#         draw_avg_chip(df_list[i],names[i],max_dist)\n",
    "#         draw_avg_dist_lineplot(df_list[i],names[i],max_dist)\n",
    "#         draw_avg_dist_densityplot(df_list[i],names[i],max_dist)\n",
    "        draw_avg_dist_densityplot2(df_list[i],names[i],max_dist)\n",
    "\n",
    "def thd_row_lengths(folder,prefix):\n",
    "    tda_dir=\"datasets/\"+folder\n",
    "\n",
    "    thds=[n for n in os.listdir(tda_dir) if n[0:len(prefix)]==prefix]\n",
    "    \n",
    "    d={}\n",
    "    \n",
    "    for model in thds:\n",
    "        \n",
    "        with open(tda_dir+model) as jsonf:\n",
    "            a=json.load(jsonf)\n",
    "\n",
    "            d[model]=a['meta']['row_count']\n",
    "    \n",
    "    for i in d:\n",
    "        print(i +\" \"+ str(round(d[i]/max(d.values()),2)))\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# demonstrate draw_thd_segmentations\n",
    "\n",
    "# read in the dataset to use below\n",
    "# dataset=pandas.read_csv(\"datasets/protein_test_gt.csv\")\n",
    "\n",
    "\n",
    "#     draw_thd_segmentations(\"thd_test__Absolute Correlation_protein_test_gt_majority.csv_2019.06.18 15.51.08/\",\"thd_test_ \",dataset)\n",
    "# thd_row_lengths(\"thd_test__Absolute Correlation_protein_test_gt_majority.csv_2019.06.18 15.51.08/\",\"thd_test_ \")\n",
    "\n",
    "#     draw_thd_segmentations(\"THD_test_Absolute Correlation_protein_test_fifteen_gt.csv_2019.06.19 14.20.55/\",\"THD_test \",dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# screwing around with showing secondary structures in chart\n",
    "def twod_secondary_struct(pdb_id):\n",
    "    dim=length_dict[pdb_id]\n",
    "    k=distance_maps_cb[pdb_id].astype(float).reshape(dim,dim)\n",
    "\n",
    "    plt.imshow(np.triu(k))\n",
    "    plt.show()\n",
    "\n",
    "    helix=np.matrix(input_features[pdb_id][0].astype(float))\n",
    "    k=np.repeat(helix,dim,0)\n",
    "\n",
    "    plt.imshow(np.triu(np.transpose(k)*k))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
